

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Scripts &mdash; trxtools 0.3.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=e6c5d56c" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=1e28cc32"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="trxtools package" href="trxtools.html" />
    <link rel="prev" title="trxtools" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            trxtools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#main-scripts">Main scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sam2profilesgenomic-py">SAM2profilesGenomic.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sam2profilestranscripts-py">SAM2profilesTranscripts.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genomenascentfolding-py">genomeNascentFolding.py</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#supporting-scripts">Supporting scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome2windowsgtf-py">genome2WindowsGTF.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv2pickle-py">csv2pickle.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mergesalmon-py">mergeSalmon.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fasta2slidingwindows-py">fasta2slidingWindows.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polyafraction-py">polyAfraction.py</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="trxtools.html">trxtools package</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="release.html">Release information</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">trxtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Scripts</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/scripts.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="scripts">
<h1>Scripts<a class="headerlink" href="#scripts" title="Link to this heading"></a></h1>
<section id="main-scripts">
<h2>Main scripts<a class="headerlink" href="#main-scripts" title="Link to this heading"></a></h2>
<section id="sam2profilesgenomic-py">
<h3>SAM2profilesGenomic.py<a class="headerlink" href="#sam2profilesgenomic-py" title="Link to this heading"></a></h3>
<p>This script generates genomic profiles from SAM files. It provides various options to customize the generation of profiles based on the input SAM file.</p>
<dl>
<dt>Usage:</dt><dd><p>SAM2profilesGenomic.py -f file.sam -u 3end -n -s polyA</p>
</dd>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span>, <span class="option">--sam_file <var>FILE</var></span></kbd></dt>
<dd><p>SAM file to be processed.</p>
</dd>
</dl>
<dl class="simple">
<dt>-u {read,3end,5end,del}, –use {read,3end,5end,del}</dt><dd><p>Generate profiles using reads (read) or the 3’ ends (3end), 5’ ends (5end), or deletions (del).</p>
</dd>
</dl>
<dl class="option-list">
<dt><kbd><span class="option">-n</span>, <span class="option">--noncoded</span></kbd></dt>
<dd><p>Save non-coded ends. Can be used ONLY with: -u 3end.</p>
</dd>
</dl>
<dl class="simple">
<dt>-s {polyA}, –noncoded_suffix {polyA}</dt><dd><p>Select non-coded ends. Can be used ONLY with: -u 3end.</p>
</dd>
</dl>
<dl class="option-list">
<dt><kbd><span class="option">-e <var>EXPAND</var></span>, <span class="option">--expand <var>EXPAND</var></span></kbd></dt>
<dd><p>Will expand position of each deletion by +/- the value. Works ONLY with: -u del.</p>
</dd>
<dt><kbd><span class="option">-c <var>TOCLEAR</var></span>, <span class="option">--toClear <var>TOCLEAR</var></span></kbd></dt>
<dd><p>String of signs to be cleared from the name of SAM file.</p>
</dd>
<dt><kbd><span class="option">--chunks <var>CHUNKS</var></span></kbd></dt>
<dd><p>Divide list of genes into chunks of given size.</p>
</dd>
</dl>
</dd>
<dt>Example:</dt><dd><p>SAM2profilesGenomic.py -f file.sam -u 3end -n -s polyA</p>
</dd>
</dl>
</section>
<section id="sam2profilestranscripts-py">
<h3>SAM2profilesTranscripts.py<a class="headerlink" href="#sam2profilestranscripts-py" title="Link to this heading"></a></h3>
<p>This script generates profiles from SAM files for a list of given transcripts.
It reads a SAM file, a list of genes, and a file with transcriptome details,
and processes the data to create profiles. The output can be saved as a pickle
DataFrame if specified.</p>
<dl>
<dt>Usage:</dt><dd><p>SAM2profilesTranscripts.py -f file.sam -l list_of_genes.txt -d transcriptome_details.tab</p>
</dd>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span>, <span class="option">--sam_file <var>FILE</var></span></kbd></dt>
<dd><p>SAM file to be processed.</p>
</dd>
<dt><kbd><span class="option">-l <var>FILE</var></span>, <span class="option">--list_file <var>FILE</var></span></kbd></dt>
<dd><p>File containing a list of genes.</p>
</dd>
<dt><kbd><span class="option">-d <var>FILE</var></span>, <span class="option">--details_file <var>FILE</var></span></kbd></dt>
<dd><p>File with transcriptome details. Default is ‘/homes/tollervey/COVID19/databases/transcriptome_details_biomart_tRNA_rRNA_UPDATEJan2021.tab’.</p>
</dd>
<dt><kbd><span class="option">--del</span></kbd></dt>
<dd><p>Generate additional profiles for deletions.</p>
</dd>
<dt><kbd><span class="option">-p</span>, <span class="option">--pickle</span></kbd></dt>
<dd><p>Save output as a pickle DataFrame.</p>
</dd>
<dt><kbd><span class="option">-e <var>INT</var></span>, <span class="option">--expand <var>INT</var></span></kbd></dt>
<dd><p>For deletions, position can be expanded by the value on each side (e=5 gives 10 nt long). Default is 5.</p>
</dd>
<dt><kbd><span class="option">-c <var>STRING</var></span>, <span class="option">--toClear <var>STRING</var></span></kbd></dt>
<dd><p>String of signs to be cleared from the name of SAM file. Default is ‘_comp_flexbar_STARAligned.out’.</p>
</dd>
<dt><kbd><span class="option">--chunks <var>INT</var></span></kbd></dt>
<dd><p>Divide list of genes into chunks of given size. Default is 0.</p>
</dd>
</dl>
</dd>
<dt>Example:</dt><dd><p>SAM2profilesTranscripts.py -f file.sam -l list_of_genes.txt -d transcriptome_details.tab</p>
</dd>
</dl>
</section>
<section id="genomenascentfolding-py">
<h3>genomeNascentFolding.py<a class="headerlink" href="#genomenascentfolding-py" title="Link to this heading"></a></h3>
<p>This script processes a given FASTA file to generate sliding windows of sequences and their reverse complements.
It then folds these sequences using RNAfold and parses the folding output to calculate the free energy (dG) of the folded sequences.
The results are saved in WIG and BigWig formats for visualization.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>python genomeNascentFolding.py -f genome.fasta -w 65 -t 37 -s both</p>
</dd>
</dl>
<p>Arguments:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-f</span>, <span class="option">--file</span></kbd></dt>
<dd><p>Path to the input FASTA file (required).</p>
</dd>
<dt><kbd><span class="option">-g</span>, <span class="option">--gzip</span></kbd></dt>
<dd><p>Enable gzip compression for the input file.</p>
</dd>
<dt><kbd><span class="option">-w</span>, <span class="option">--window</span></kbd></dt>
<dd><p>Size of the sliding window (default: 65).</p>
</dd>
<dt><kbd><span class="option">-t</span>, <span class="option">--temp</span></kbd></dt>
<dd><p>Temperature for RNA folding (default: 37).</p>
</dd>
<dt><kbd><span class="option">-s</span>, <span class="option">--strand</span></kbd></dt>
<dd><p>Strand to process (‘plus’, ‘minus’, or ‘both’; default: ‘both’).</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Output:</dt><dd><p>The script generates sliding window sequences and their reverse complements, folds them using RNAfold,
and saves the folding free energy (dG) values in WIG and BigWig formats. The output files are saved in the same
directory as the input file, with appropriate naming conventions.</p>
</dd>
</dl>
</section>
</section>
<section id="supporting-scripts">
<h2>Supporting scripts<a class="headerlink" href="#supporting-scripts" title="Link to this heading"></a></h2>
<section id="genome2windowsgtf-py">
<h3>genome2WindowsGTF.py<a class="headerlink" href="#genome2windowsgtf-py" title="Link to this heading"></a></h3>
<p>This script generates a GTF file for a given genome with fixed-size windows (default 100bp).</p>
<dl>
<dt>Usage:</dt><dd><p>python genome2WindowsGTF.py -f genome.len -w 1000</p>
</dd>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span></kbd></dt>
<dd><p>File with genome length</p>
</dd>
<dt><kbd><span class="option">-w <var>WINDOW</var></span></kbd></dt>
<dd><p>Window size (default: 100)</p>
</dd>
</dl>
</dd>
<dt>Description:</dt><dd><p>The script reads a genome length file and generates a GTF file with fixed-size windows.
Each window is represented as an exon feature in the GTF file. The script creates windows
for both the positive and negative strands of the genome.</p>
</dd>
<dt>Output:</dt><dd><p>The script outputs a GTF file with the specified window size for the given genome.</p>
</dd>
</dl>
</section>
<section id="csv2pickle-py">
<h3>csv2pickle.py<a class="headerlink" href="#csv2pickle-py" title="Link to this heading"></a></h3>
<p>This script reads a CSV file and saves it as a pickle file. It supports optional gzip compression.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>python csv2pickle.py -f &lt;filename.csv&gt; –gzip</p>
</dd>
</dl>
<p>Arguments:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span>, <span class="option">--csv_file <var>FILE</var></span></kbd></dt>
<dd><p>Path to the input CSV file.</p>
</dd>
</dl>
<p class="attribution">—gzip
Enable gzip compression for the output pickle file.</p>
</div></blockquote>
<dl class="simple">
<dt>Notes:</dt><dd><p>The script reads the CSV file from the current working directory and saves the pickle file in the same directory.</p>
</dd>
</dl>
</section>
<section id="mergesalmon-py">
<h3>mergeSalmon.py<a class="headerlink" href="#mergesalmon-py" title="Link to this heading"></a></h3>
<p>This script merges multiple Salmon quantification files into a single tab-delimited file. It takes the name of the Salmon output directory and the column to be used (TPM or NumReads). The script can also filter the files to be loaded, clear some strings from the experiment name, and add some strings to the experiment name. The output is saved in the current directory with the name ‘merged.tab’.</p>
<dl>
<dt>Usage:</dt><dd><p>python mergeSalmon.py -i ‘Salmon_output’ -u TPM -f ‘sample’ -a ‘RNAseq’</p>
</dd>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-i</span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory</p>
</dd>
<dt><kbd><span class="option">-o</span></kbd></dt>
<dd><p>Name of output file (default: ‘merged’)</p>
</dd>
<dt><kbd><span class="option">-u</span></kbd></dt>
<dd><p>Column to be used {TPM, NumReads} (default: ‘NumReads’)</p>
</dd>
<dt><kbd><span class="option">-f</span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory. Optional as additional filter (default: None)</p>
</dd>
<dt><kbd><span class="option">-c</span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory, will be cleared (default: None)</p>
</dd>
<dt><kbd><span class="option">-a</span></kbd></dt>
<dd><p>String of signs to be added to experiment name (default: None)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="fasta2slidingwindows-py">
<h3>fasta2slidingWindows.py<a class="headerlink" href="#fasta2slidingwindows-py" title="Link to this heading"></a></h3>
<p>This script converts a FASTA file to sliding windows as sequences and their reverse complements (RC).
The output is saved in a folder named with the prefix ‘window’ followed by the window size. Each chromosome
is saved in a separate file. If the tab_output option is selected, the output is saved as a tab-separated file;
otherwise, it is saved as FASTA files.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>python fasta2slidingWindows.py -f genome.fasta -w 65 -t 37 -s both –tab</p>
</dd>
</dl>
<p>Arguments:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-f</span>, <span class="option">--file</span></kbd></dt>
<dd><p>Path to the input FASTA file (required).</p>
</dd>
<dt><kbd><span class="option">-g</span>, <span class="option">--gzip</span></kbd></dt>
<dd><p>Enable gzip compression for the input file.</p>
</dd>
<dt><kbd><span class="option">-w</span>, <span class="option">--window</span></kbd></dt>
<dd><p>Sliding window size (default: 65).</p>
</dd>
<dt><kbd><span class="option">-t</span>, <span class="option">--temp</span></kbd></dt>
<dd><p>Temperature (default: 37).</p>
</dd>
<dt><kbd><span class="option">-s</span>, <span class="option">--strand</span></kbd></dt>
<dd><p>Strand to process, options are “plus”, “minus”, or “both” (default: “both”).</p>
</dd>
</dl>
<p class="attribution">—tab
Save output as tab-separated file.</p>
</div></blockquote>
</section>
<section id="polyafraction-py">
<h3>polyAfraction.py<a class="headerlink" href="#polyafraction-py" title="Link to this heading"></a></h3>
<p>This script generates fraction (eg. polyA) for genomic profiles from BigWig files. Bot fwd and rev files will be processed.</p>
<dl>
<dt>Usage:</dt><dd><p>polyAfraction.py -f file_polyA_fwd.bw</p>
</dd>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span>, <span class="option">--sam_file <var>FILE</var></span></kbd></dt>
<dd><p>BigWig file to be processed. Requires file with suffix “_fwd.bw” or “_rev.bw”. Either can be used.
The script will automatically detect the suffix and process the file accordingly.
The script will also look for the corresponding file with the other suffix (if not provided).
For example, if you provide “file_polyA_fwd.bw”, it will look for “file_polyA_rev.bw” in the same directory.</p>
</dd>
<dt><kbd><span class="option">-s <var>SUFFIX</var></span>, <span class="option">--suffix <var>SUFFIX</var></span></kbd></dt>
<dd><p>Suffix for the input file (default: polyA).</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>Show this help message and exit.</p>
</dd>
</dl>
</dd>
<dt>Example:</dt><dd><p>polyAfraction.py -f file_polyA_fwd.bw</p>
</dd>
</dl>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="trxtools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="trxtools.html" class="btn btn-neutral float-right" title="trxtools package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Turowski Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>