Search.setIndex({"alltitles": {"Authors": [[1, "authors"]], "Contents": [[0, "contents"]], "Indices and tables": [[0, "indices-and-tables"]], "Main scripts": [[2, "main-scripts"]], "Release information": [[1, null]], "Reporting bugs": [[0, "reporting-bugs"], [1, "reporting-bugs"]], "SAM2profilesGenomic.py": [[2, "sam2profilesgenomic-py"]], "SAM2profilesTranscripts.py": [[2, "sam2profilestranscripts-py"]], "Scripts": [[2, null]], "Supporting scripts": [[2, "supporting-scripts"]], "Versions": [[1, "versions"]], "csv2pickle.py": [[2, "csv2pickle-py"]], "fasta2slidingWindows.py": [[2, "fasta2slidingwindows-py"]], "genome2WindowsGTF.py": [[2, "genome2windowsgtf-py"]], "genomeNascentFolding.py": [[2, "genomenascentfolding-py"]], "mergeSalmon.py": [[2, "mergesalmon-py"]], "trxtools": [[0, null]], "trxtools package": [[3, null]], "trxtools.BigWig module": [[3, "module-trxtools.BigWig"]], "trxtools.assays module": [[3, "module-trxtools.assays"]], "trxtools.go_enrichment module": [[3, "module-trxtools.go_enrichment"]], "trxtools.metaprofiles module": [[3, "module-trxtools.metaprofiles"]], "trxtools.methods module": [[3, "module-trxtools.methods"]], "trxtools.nascent module": [[3, "module-trxtools.nascent"]], "trxtools.plotting module": [[3, "module-trxtools.plotting"]], "trxtools.profiles module": [[3, "module-trxtools.profiles"]], "trxtools.sam module": [[3, "trxtools-sam-module"]], "trxtools.sam.SAMgeneral submodule": [[3, "module-trxtools.sam.SAMgeneral"]], "trxtools.sam.SAMgenome submodule": [[3, "module-trxtools.sam.SAMgenome"]], "trxtools.sam.SAMtranscripts submodule": [[3, "module-trxtools.sam.SAMtranscripts"]], "trxtools.secondary module": [[3, "module-trxtools.secondary"]], "v0.2.1": [[1, "v0-2-1"]], "v0.2.2": [[1, "v0-2-2"]], "v0.3.0": [[1, "v0-3-0"]]}, "docnames": ["index", "release", "scripts", "trxtools"], "envversion": {"sphinx": 64, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filenames": ["index.rst", "release.rst", "scripts.rst", "trxtools.rst"], "indexentries": {"addcluster() (in module trxtools.methods)": [[3, "trxtools.methods.addCluster", false]], "bashcommand() (in module trxtools.methods)": [[3, "trxtools.methods.bashCommand", false]], "bashfolding() (trxtools.nascent.fold method)": [[3, "trxtools.nascent.Fold.bashFolding", false]], "bashhybrid() (trxtools.nascent.hybrid method)": [[3, "trxtools.nascent.Hybrid.bashHybrid", false]], "bed2len() (in module trxtools.methods)": [[3, "trxtools.methods.bed2len", false]], "bed_split_strands() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.bed_split_strands", false]], "bincollect3() (in module trxtools.profiles)": [[3, "trxtools.profiles.binCollect3", false]], "binmultiplematrices() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.binMultipleMatrices", false]], "boxplot1() (in module trxtools.plotting)": [[3, "trxtools.plotting.boxplot1", false]], "bulkinputidt() (in module trxtools.assays)": [[3, "trxtools.assays.bulkInputIDT", false]], "calculatefdr() (in module trxtools.profiles)": [[3, "trxtools.profiles.calculateFDR", false]], "calgc() (in module trxtools.methods)": [[3, "trxtools.methods.calGC", false]], "checkvienna() (in module trxtools.secondary)": [[3, "trxtools.secondary.checkVienna", false]], "chromosome2profile3end() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.chromosome2profile3end", false]], "cleannames() (in module trxtools.methods)": [[3, "trxtools.methods.cleanNames", false]], "clusterclustermap() (in module trxtools.plotting)": [[3, "trxtools.plotting.clusterClusterMap", false]], "compare1toref() (in module trxtools.profiles)": [[3, "trxtools.profiles.compare1toRef", false]], "comparemoretoref() (in module trxtools.profiles)": [[3, "trxtools.profiles.compareMoretoRef", false]], "countdeletion() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.countDeletion", false]], "countmiddle() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.countMiddle", false]], "countread() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.countRead", false]], "cumulativedifference() (in module trxtools.plotting)": [[3, "trxtools.plotting.cumulativeDifference", false]], "cumulativepeaks() (in module trxtools.plotting)": [[3, "trxtools.plotting.cumulativePeaks", false]], "define_experiments() (in module trxtools.methods)": [[3, "trxtools.methods.define_experiments", false]], "dictbigwig() (in module trxtools.profiles)": [[3, "trxtools.profiles.dictBigWig", false]], "dna_stretch_positions() (in module trxtools.methods)": [[3, "trxtools.methods.DNA_stretch_positions", false]], "dna_string_positions() (in module trxtools.methods)": [[3, "trxtools.methods.DNA_string_positions", false]], "enhancedvolcano() (in module trxtools.plotting)": [[3, "trxtools.plotting.enhancedVolcano", false]], "enriched() (in module trxtools.methods)": [[3, "trxtools.methods.enriched", false]], "expnameparser() (in module trxtools.methods)": [[3, "trxtools.methods.expNameParser", false]], "expstats() (in module trxtools.methods)": [[3, "trxtools.methods.expStats", false]], "extendingwindow() (in module trxtools.nascent)": [[3, "trxtools.nascent.extendingWindow", false]], "extruded() (in module trxtools.assays)": [[3, "trxtools.assays.extruded", false]], "filterexp() (in module trxtools.methods)": [[3, "trxtools.methods.filterExp", false]], "find_pol3_terminators() (in module trxtools.methods)": [[3, "trxtools.methods.find_pol3_terminators", false]], "findpeaks() (in module trxtools.profiles)": [[3, "trxtools.profiles.findPeaks", false]], "findprimer() (in module trxtools.assays)": [[3, "trxtools.assays.findPrimer", false]], "findtroughs() (in module trxtools.profiles)": [[3, "trxtools.profiles.findTroughs", false]], "fold (class in trxtools.nascent)": [[3, "trxtools.nascent.Fold", false]], "foldingfrombigwig() (in module trxtools.bigwig)": [[3, "trxtools.BigWig.foldingFromBigWig", false]], "foldingfrombigwig() (in module trxtools.profiles)": [[3, "trxtools.profiles.FoldingFromBigWig", false]], "foldnascentelem() (in module trxtools.nascent)": [[3, "trxtools.nascent.foldNascentElem", false]], "genefrombigwig() (in module trxtools.bigwig)": [[3, "trxtools.BigWig.geneFromBigWig", false]], "genefrombigwig() (in module trxtools.profiles)": [[3, "trxtools.profiles.geneFromBigWig", false]], "generatesubplots() (in module trxtools.plotting)": [[3, "trxtools.plotting.generateSubplots", false]], "get_bw_data() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.get_bw_data", false]], "get_enrichment() (in module trxtools.go_enrichment)": [[3, "trxtools.go_enrichment.get_enrichment", false]], "get_multiple_matrices() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.get_multiple_matrices", false]], "getmultiplematrices() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.getMultipleMatrices", false]], "getmultiplematricesfrompeak() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.getMultipleMatricesFromPeak", false]], "getseqdata() (in module trxtools.bigwig)": [[3, "trxtools.BigWig.getSeqData", false]], "goterm() (in module trxtools.plotting)": [[3, "trxtools.plotting.GOterm", false]], "groupcigar() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.groupCIGAR", false]], "groupcracsamples() (in module trxtools.methods)": [[3, "trxtools.methods.groupCRACsamples", false]], "handleinput() (in module trxtools.nascent)": [[3, "trxtools.nascent.handleInput", false]], "hplotstarstats_readlen() (in module trxtools.plotting)": [[3, "trxtools.plotting.hplotSTARstats_readLen", false]], "hplotstarstats_reads() (in module trxtools.plotting)": [[3, "trxtools.plotting.hplotSTARstats_reads", false]], "hybrid (class in trxtools.nascent)": [[3, "trxtools.nascent.Hybrid", false]], "indexorder() (in module trxtools.methods)": [[3, "trxtools.methods.indexOrder", false]], "is_inside() (in module trxtools.methods)": [[3, "trxtools.methods.is_inside", false]], "join2d() (in module trxtools.nascent)": [[3, "trxtools.nascent.join2d", false]], "join_strand_matrices() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.join_strand_matrices", false]], "lettercontent() (in module trxtools.methods)": [[3, "trxtools.methods.letterContent", false]], "listpaths() (in module trxtools.methods)": [[3, "trxtools.methods.listPaths", false]], "loadgtf() (in module trxtools.methods)": [[3, "trxtools.methods.loadGTF", false]], "loops() (in module trxtools.secondary)": [[3, "trxtools.secondary.loops", false]], "loopstems() (in module trxtools.secondary)": [[3, "trxtools.secondary.loopStems", false]], "lstem() (in module trxtools.secondary)": [[3, "trxtools.secondary.Lstem", false]], "matrix_from_bigwig() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.matrix_from_bigwig", false]], "matrixfrombigwig() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.matrixFromBigWig", false]], "merge2d() (in module trxtools.nascent)": [[3, "trxtools.nascent.merge2d", false]], "metaprofile() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.metaprofile", false]], "metaprofileandheatmap() (in module trxtools.plotting)": [[3, "trxtools.plotting.metaprofileAndHeatmap", false]], "module": [[3, "module-trxtools.BigWig", false], [3, "module-trxtools.assays", false], [3, "module-trxtools.go_enrichment", false], [3, "module-trxtools.metaprofiles", false], [3, "module-trxtools.methods", false], [3, "module-trxtools.nascent", false], [3, "module-trxtools.plotting", false], [3, "module-trxtools.profiles", false], [3, "module-trxtools.sam.SAMgeneral", false], [3, "module-trxtools.sam.SAMgenome", false], [3, "module-trxtools.sam.SAMtranscripts", false], [3, "module-trxtools.secondary", false]], "name2index() (in module trxtools.nascent)": [[3, "trxtools.nascent.name2index", false]], "nascentelems() (in module trxtools.nascent)": [[3, "trxtools.nascent.nascentElems", false]], "nascentelemsdataframe() (in module trxtools.nascent)": [[3, "trxtools.nascent.nascentElemsDataFrame", false]], "nascentfolding() (in module trxtools.nascent)": [[3, "trxtools.nascent.nascentFolding", false]], "nested_region_cleanup() (in module trxtools.methods)": [[3, "trxtools.methods.nested_region_cleanup", false]], "noncoded2profile() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.noncoded2profile", false]], "noncoded2profile1() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.noncoded2profile1", false]], "nontemplatedna() (in module trxtools.assays)": [[3, "trxtools.assays.nonTemplateDNA", false]], "normalize() (in module trxtools.methods)": [[3, "trxtools.methods.normalize", false]], "ntotal() (in module trxtools.profiles)": [[3, "trxtools.profiles.ntotal", false]], "parseconcatfile() (in module trxtools.profiles)": [[3, "trxtools.profiles.parseConcatFile", false]], "parsecracname() (in module trxtools.methods)": [[3, "trxtools.methods.parseCRACname", false]], "parsefoldingname() (in module trxtools.nascent)": [[3, "trxtools.nascent.parseFoldingName", false]], "parseheader() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.parseHeader", false]], "parsenoncoded() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.parseNoncoded", false]], "parsenoncodedlist() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.parseNoncodedList", false]], "pca() (in module trxtools.plotting)": [[3, "trxtools.plotting.PCA", false]], "peak2matrice() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.peak2matrice", false]], "plot_as_box_plot() (in module trxtools.plotting)": [[3, "trxtools.plotting.plot_as_box_plot", false]], "plot_diff() (in module trxtools.plotting)": [[3, "trxtools.plotting.plot_diff", false]], "plot_heatmap() (in module trxtools.plotting)": [[3, "trxtools.plotting.plot_heatmap", false]], "plot_to_compare() (in module trxtools.plotting)": [[3, "trxtools.plotting.plot_to_compare", false]], "plotandfolding() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotAndFolding", false]], "plotstarstats() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats", false]], "plotstarstats_chimeric() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats_chimeric", false]], "plotstarstats_mapping() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats_mapping", false]], "plotstarstats_mistmatches() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats_mistmatches", false]], "plotstarstats_readlen() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats_readLen", false]], "plotstarstats_reads() (in module trxtools.plotting)": [[3, "trxtools.plotting.plotSTARstats_reads", false]], "preparenascent() (in module trxtools.nascent)": [[3, "trxtools.nascent.prepareNascent", false]], "preprocess() (in module trxtools.profiles)": [[3, "trxtools.profiles.preprocess", false]], "pseudocounts() (in module trxtools.profiles)": [[3, "trxtools.profiles.pseudocounts", false]], "quantilecategory() (in module trxtools.methods)": [[3, "trxtools.methods.quantileCategory", false]], "randomdnaall() (in module trxtools.methods)": [[3, "trxtools.methods.randomDNAall", false]], "randomdnasingle() (in module trxtools.methods)": [[3, "trxtools.methods.randomDNAsingle", false]], "read_bed() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.read_bed", false]], "read_deseq() (in module trxtools.methods)": [[3, "trxtools.methods.read_DEseq", false]], "read_featurecount() (in module trxtools.methods)": [[3, "trxtools.methods.read_featureCount", false]], "read_htseq_output() (in module trxtools.methods)": [[3, "trxtools.methods.read_HTSeq_output", false]], "read_list() (in module trxtools.methods)": [[3, "trxtools.methods.read_list", false]], "read_starstats() (in module trxtools.methods)": [[3, "trxtools.methods.read_STARstats", false]], "read_tabfile() (in module trxtools.methods)": [[3, "trxtools.methods.read_tabFile", false]], "readbed() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.readBED", false]], "reads2genome() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.reads2genome", false]], "reads2genome3end() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.reads2genome3end", false]], "reads2genome5end() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.reads2genome5end", false]], "reads2genomedeletions() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.reads2genomeDeletions", false]], "reads2profile() (in module trxtools.sam.samtranscripts)": [[3, "trxtools.sam.SAMtranscripts.reads2profile", false]], "reads2profiledeletions() (in module trxtools.sam.samtranscripts)": [[3, "trxtools.sam.SAMtranscripts.reads2profileDeletions", false]], "readsalmon() (in module trxtools.methods)": [[3, "trxtools.methods.readSalmon", false]], "regionscore() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.regionScore", false]], "reverse_complement() (in module trxtools.methods)": [[3, "trxtools.methods.reverse_complement", false]], "reverse_complement_dna() (in module trxtools.methods)": [[3, "trxtools.methods.reverse_complement_DNA", false]], "reverse_complement_rna() (in module trxtools.methods)": [[3, "trxtools.methods.reverse_complement_RNA", false]], "rnafold() (trxtools.nascent.fold method)": [[3, "trxtools.nascent.Fold.RNAfold", false]], "rnahybrid() (trxtools.nascent.hybrid method)": [[3, "trxtools.nascent.Hybrid.RNAhybrid", false]], "rnainvert() (trxtools.nascent.fold method)": [[3, "trxtools.nascent.Fold.RNAinvert", false]], "rnainvertstall() (trxtools.nascent.fold method)": [[3, "trxtools.nascent.Fold.RNAinvertStall", false]], "rollinggc() (in module trxtools.methods)": [[3, "trxtools.methods.rollingGC", false]], "rstem() (in module trxtools.secondary)": [[3, "trxtools.secondary.Rstem", false]], "runpca() (in module trxtools.methods)": [[3, "trxtools.methods.runPCA", false]], "sam2genome() (in module trxtools.sam.samgenome)": [[3, "trxtools.sam.SAMgenome.sam2genome", false]], "sam2profiles() (in module trxtools.sam.samtranscripts)": [[3, "trxtools.sam.SAMtranscripts.sam2profiles", false]], "save_csv() (in module trxtools.profiles)": [[3, "trxtools.profiles.save_csv", false]], "savebigwig() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.saveBigWig", false]], "select_colors() (in module trxtools.plotting)": [[3, "trxtools.plotting.select_colors", false]], "selectends() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.selectEnds", false]], "selectpolya() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.selectPolyA", false]], "selectsortpaths() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.selectSortPaths", false]], "selectsubsetmm() (in module trxtools.metaprofiles)": [[3, "trxtools.metaprofiles.selectSubsetMM", false]], "sequenceconstrain() (in module trxtools.assays)": [[3, "trxtools.assays.sequenceConstrain", false]], "slidingwindow() (in module trxtools.nascent)": [[3, "trxtools.nascent.slidingWindow", false]], "stalled() (in module trxtools.assays)": [[3, "trxtools.assays.stalled", false]], "strip_bigwig_names() (in module trxtools.bigwig)": [[3, "trxtools.BigWig.strip_BigWig_names", false]], "stripbigwignames() (in module trxtools.profiles)": [[3, "trxtools.profiles.stripBigWignames", false]], "stripcigar() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.stripCIGAR", false]], "stripsubstitutions() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.stripSubstitutions", false]], "structurefile() (in module trxtools.assays)": [[3, "trxtools.assays.structureFile", false]], "substructures() (in module trxtools.secondary)": [[3, "trxtools.secondary.substructures", false]], "templatedna() (in module trxtools.assays)": [[3, "trxtools.assays.templateDNA", false]], "test() (in module trxtools.secondary)": [[3, "trxtools.secondary.test", false]], "testscaffold() (in module trxtools.assays)": [[3, "trxtools.assays.testScaffold", false]], "timestamp() (in module trxtools.methods)": [[3, "trxtools.methods.timestamp", false]], "timestamprandomint() (in module trxtools.methods)": [[3, "trxtools.methods.timestampRandomInt", false]], "toorder() (in module trxtools.assays)": [[3, "trxtools.assays.toOrder", false]], "tostripcigarfive() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.tostripCIGARfive", false]], "tostripcigarthree() (in module trxtools.sam.samgeneral)": [[3, "trxtools.sam.SAMgeneral.tostripCIGARthree", false]], "transcript2profile() (in module trxtools.sam.samtranscripts)": [[3, "trxtools.sam.SAMtranscripts.transcript2profile", false]], "transcript2profiledeletions() (in module trxtools.sam.samtranscripts)": [[3, "trxtools.sam.SAMtranscripts.transcript2profileDeletions", false]], "trxtools.assays": [[3, "module-trxtools.assays", false]], "trxtools.bigwig": [[3, "module-trxtools.BigWig", false]], "trxtools.go_enrichment": [[3, "module-trxtools.go_enrichment", false]], "trxtools.metaprofiles": [[3, "module-trxtools.metaprofiles", false]], "trxtools.methods": [[3, "module-trxtools.methods", false]], "trxtools.nascent": [[3, "module-trxtools.nascent", false]], "trxtools.plotting": [[3, "module-trxtools.plotting", false]], "trxtools.profiles": [[3, "module-trxtools.profiles", false]], "trxtools.sam.samgeneral": [[3, "module-trxtools.sam.SAMgeneral", false]], "trxtools.sam.samgenome": [[3, "module-trxtools.sam.SAMgenome", false]], "trxtools.sam.samtranscripts": [[3, "module-trxtools.sam.SAMtranscripts", false]], "trxtools.secondary": [[3, "module-trxtools.secondary", false]], "unafold() (trxtools.nascent.fold method)": [[3, "trxtools.nascent.Fold.UNAfold", false]], "venndiagram() (in module trxtools.plotting)": [[3, "trxtools.plotting.vennDiagram", false]], "vienna2format() (in module trxtools.secondary)": [[3, "trxtools.secondary.vienna2format", false]]}, "objects": {"trxtools": [[3, 0, 0, "-", "BigWig"], [3, 0, 0, "-", "assays"], [3, 0, 0, "-", "go_enrichment"], [3, 0, 0, "-", "metaprofiles"], [3, 0, 0, "-", "methods"], [3, 0, 0, "-", "nascent"], [3, 0, 0, "-", "plotting"], [3, 0, 0, "-", "profiles"], [3, 0, 0, "-", "secondary"]], "trxtools.BigWig": [[3, 1, 1, "", "foldingFromBigWig"], [3, 1, 1, "", "geneFromBigWig"], [3, 1, 1, "", "getSeqData"], [3, 1, 1, "", "strip_BigWig_names"]], "trxtools.assays": [[3, 1, 1, "", "bulkInputIDT"], [3, 1, 1, "", "extruded"], [3, 1, 1, "", "findPrimer"], [3, 1, 1, "", "nonTemplateDNA"], [3, 1, 1, "", "sequenceConstrain"], [3, 1, 1, "", "stalled"], [3, 1, 1, "", "structureFile"], [3, 1, 1, "", "templateDNA"], [3, 1, 1, "", "testScaffold"], [3, 1, 1, "", "toOrder"]], "trxtools.go_enrichment": [[3, 1, 1, "", "get_enrichment"]], "trxtools.metaprofiles": [[3, 1, 1, "", "bed_split_strands"], [3, 1, 1, "", "binMultipleMatrices"], [3, 1, 1, "", "getMultipleMatrices"], [3, 1, 1, "", "getMultipleMatricesFromPeak"], [3, 1, 1, "", "get_bw_data"], [3, 1, 1, "", "get_multiple_matrices"], [3, 1, 1, "", "join_strand_matrices"], [3, 1, 1, "", "matrixFromBigWig"], [3, 1, 1, "", "matrix_from_bigwig"], [3, 1, 1, "", "metaprofile"], [3, 1, 1, "", "peak2matrice"], [3, 1, 1, "", "readBED"], [3, 1, 1, "", "read_bed"], [3, 1, 1, "", "regionScore"], [3, 1, 1, "", "selectSubsetMM"]], "trxtools.methods": [[3, 1, 1, "", "DNA_stretch_positions"], [3, 1, 1, "", "DNA_string_positions"], [3, 1, 1, "", "addCluster"], [3, 1, 1, "", "bashCommand"], [3, 1, 1, "", "bed2len"], [3, 1, 1, "", "calGC"], [3, 1, 1, "", "cleanNames"], [3, 1, 1, "", "define_experiments"], [3, 1, 1, "", "enriched"], [3, 1, 1, "", "expNameParser"], [3, 1, 1, "", "expStats"], [3, 1, 1, "", "filterExp"], [3, 1, 1, "", "find_pol3_terminators"], [3, 1, 1, "", "groupCRACsamples"], [3, 1, 1, "", "indexOrder"], [3, 1, 1, "", "is_inside"], [3, 1, 1, "", "letterContent"], [3, 1, 1, "", "listPaths"], [3, 1, 1, "", "loadGTF"], [3, 1, 1, "", "nested_region_cleanup"], [3, 1, 1, "", "normalize"], [3, 1, 1, "", "parseCRACname"], [3, 1, 1, "", "quantileCategory"], [3, 1, 1, "", "randomDNAall"], [3, 1, 1, "", "randomDNAsingle"], [3, 1, 1, "", "readSalmon"], [3, 1, 1, "", "read_DEseq"], [3, 1, 1, "", "read_HTSeq_output"], [3, 1, 1, "", "read_STARstats"], [3, 1, 1, "", "read_featureCount"], [3, 1, 1, "", "read_list"], [3, 1, 1, "", "read_tabFile"], [3, 1, 1, "", "reverse_complement"], [3, 1, 1, "", "reverse_complement_DNA"], [3, 1, 1, "", "reverse_complement_RNA"], [3, 1, 1, "", "rollingGC"], [3, 1, 1, "", "runPCA"], [3, 1, 1, "", "timestamp"], [3, 1, 1, "", "timestampRandomInt"]], "trxtools.nascent": [[3, 2, 1, "", "Fold"], [3, 2, 1, "", "Hybrid"], [3, 1, 1, "", "extendingWindow"], [3, 1, 1, "", "foldNascentElem"], [3, 1, 1, "", "handleInput"], [3, 1, 1, "", "join2d"], [3, 1, 1, "", "merge2d"], [3, 1, 1, "", "name2index"], [3, 1, 1, "", "nascentElems"], [3, 1, 1, "", "nascentElemsDataFrame"], [3, 1, 1, "", "nascentFolding"], [3, 1, 1, "", "parseFoldingName"], [3, 1, 1, "", "prepareNascent"], [3, 1, 1, "", "slidingWindow"]], "trxtools.nascent.Fold": [[3, 3, 1, "", "RNAfold"], [3, 3, 1, "", "RNAinvert"], [3, 3, 1, "", "RNAinvertStall"], [3, 3, 1, "", "UNAfold"], [3, 3, 1, "", "bashFolding"]], "trxtools.nascent.Hybrid": [[3, 3, 1, "", "RNAhybrid"], [3, 3, 1, "", "bashHybrid"]], "trxtools.plotting": [[3, 1, 1, "", "GOterm"], [3, 1, 1, "", "PCA"], [3, 1, 1, "", "boxplot1"], [3, 1, 1, "", "clusterClusterMap"], [3, 1, 1, "", "cumulativeDifference"], [3, 1, 1, "", "cumulativePeaks"], [3, 1, 1, "", "enhancedVolcano"], [3, 1, 1, "", "generateSubplots"], [3, 1, 1, "", "hplotSTARstats_readLen"], [3, 1, 1, "", "hplotSTARstats_reads"], [3, 1, 1, "", "metaprofileAndHeatmap"], [3, 1, 1, "", "plotAndFolding"], [3, 1, 1, "", "plotSTARstats"], [3, 1, 1, "", "plotSTARstats_chimeric"], [3, 1, 1, "", "plotSTARstats_mapping"], [3, 1, 1, "", "plotSTARstats_mistmatches"], [3, 1, 1, "", "plotSTARstats_readLen"], [3, 1, 1, "", "plotSTARstats_reads"], [3, 1, 1, "", "plot_as_box_plot"], [3, 1, 1, "", "plot_diff"], [3, 1, 1, "", "plot_heatmap"], [3, 1, 1, "", "plot_to_compare"], [3, 1, 1, "", "select_colors"], [3, 1, 1, "", "vennDiagram"]], "trxtools.profiles": [[3, 1, 1, "", "FoldingFromBigWig"], [3, 1, 1, "", "binCollect3"], [3, 1, 1, "", "calculateFDR"], [3, 1, 1, "", "compare1toRef"], [3, 1, 1, "", "compareMoretoRef"], [3, 1, 1, "", "dictBigWig"], [3, 1, 1, "", "findPeaks"], [3, 1, 1, "", "findTroughs"], [3, 1, 1, "", "geneFromBigWig"], [3, 1, 1, "", "ntotal"], [3, 1, 1, "", "parseConcatFile"], [3, 1, 1, "", "preprocess"], [3, 1, 1, "", "pseudocounts"], [3, 1, 1, "", "save_csv"], [3, 1, 1, "", "stripBigWignames"]], "trxtools.sam": [[3, 0, 0, "-", "SAMgeneral"], [3, 0, 0, "-", "SAMgenome"], [3, 0, 0, "-", "SAMtranscripts"]], "trxtools.sam.SAMgeneral": [[3, 1, 1, "", "countDeletion"], [3, 1, 1, "", "countMiddle"], [3, 1, 1, "", "countRead"], [3, 1, 1, "", "groupCIGAR"], [3, 1, 1, "", "noncoded2profile"], [3, 1, 1, "", "noncoded2profile1"], [3, 1, 1, "", "parseNoncoded"], [3, 1, 1, "", "parseNoncodedList"], [3, 1, 1, "", "saveBigWig"], [3, 1, 1, "", "selectEnds"], [3, 1, 1, "", "selectPolyA"], [3, 1, 1, "", "selectSortPaths"], [3, 1, 1, "", "stripCIGAR"], [3, 1, 1, "", "stripSubstitutions"], [3, 1, 1, "", "tostripCIGARfive"], [3, 1, 1, "", "tostripCIGARthree"]], "trxtools.sam.SAMgenome": [[3, 1, 1, "", "chromosome2profile3end"], [3, 1, 1, "", "parseHeader"], [3, 1, 1, "", "reads2genome"], [3, 1, 1, "", "reads2genome3end"], [3, 1, 1, "", "reads2genome5end"], [3, 1, 1, "", "reads2genomeDeletions"], [3, 1, 1, "", "sam2genome"]], "trxtools.sam.SAMtranscripts": [[3, 1, 1, "", "reads2profile"], [3, 1, 1, "", "reads2profileDeletions"], [3, 1, 1, "", "sam2profiles"], [3, 1, 1, "", "transcript2profile"], [3, 1, 1, "", "transcript2profileDeletions"]], "trxtools.secondary": [[3, 1, 1, "", "Lstem"], [3, 1, 1, "", "Rstem"], [3, 1, 1, "", "checkVienna"], [3, 1, 1, "", "loopStems"], [3, 1, 1, "", "loops"], [3, 1, 1, "", "substructures"], [3, 1, 1, "", "test"], [3, 1, 1, "", "vienna2format"]]}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "method", "Python method"]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:method"}, "terms": {"": [2, 3], "0": [0, 2, 3], "000000": 3, "01": 3, "020420": 0, "05": 3, "1": [0, 3], "10": [2, 3], "100": [2, 3], "1000": [2, 3], "100bp": 2, "11": 3, "12": 3, "13": 3, "15": 3, "150": 3, "2": [0, 3], "20": 3, "200": 3, "242641": 3, "25": 3, "250": 3, "2d": 3, "2nd": 3, "3": [0, 2, 3], "30": 3, "300": 3, "30c": 3, "34": 3, "35": 3, "36": 3, "37": 2, "39": 3, "3end": [2, 3], "3rd": 3, "3s15m1d9m2": 3, "3s44m1s1h": 3, "4": 3, "40": 3, "400": 3, "401": 3, "402": 3, "403": 3, "404": 3, "405": 3, "406": 3, "407": 3, "408": 3, "409": 3, "41": 3, "410": 3, "411": 3, "412": 3, "413": 3, "414": 3, "415": 3, "416": 3, "417": 3, "418": 3, "419": 3, "420": 3, "421": 3, "422": 3, "423": 3, "424": 3, "44": 3, "5": [2, 3], "50": 3, "500": 3, "5end": [2, 3], "5prime": 3, "6": 3, "60": 3, "65": 2, "6th": 3, "7": 3, "72": 3, "75": 3, "8": 3, "9": 3, "90": 3, "9606": 3, "A": 3, "AT": 3, "For": 2, "If": [0, 1, 2, 3], "It": [2, 3], "Not": 3, "The": [2, 3], "To": 3, "Will": 2, "_": 3, "__o_o": 3, "_comp_flexbar_staralign": 2, "_description_": 3, "_fwd": 3, "_hittable_read": 3, "_rev": 3, "_type_": 3, "aa": 3, "aaa": 3, "aacaa": 3, "ab123456": 3, "abov": 3, "ac": 3, "accord": 3, "ad": [1, 2, 3], "add": [2, 3], "addclust": [0, 3], "addit": [2, 3], "additional_tag": 3, "adjust": 3, "after": [0, 3], "afterstal": 3, "ag": 3, "agcuaguca": 3, "agg_typ": 3, "aggccgaaa": 3, "aggreg": 3, "algorithm": 3, "align": 3, "align_3end": 3, "all": 3, "allow": 3, "also": [2, 3], "an": [0, 2, 3], "analysi": [0, 3], "annot": 3, "anoth": 3, "api": 3, "append": 3, "appli": 3, "appropri": 2, "ar": [2, 3], "area": 3, "argrelextrema": 3, "argument": 2, "around": 3, "arrai": 3, "arsenite30min": 0, "assai": 0, "assign": 3, "associ": 3, "atcg": 3, "atcgatcg": 3, "aucg": 3, "author": 0, "avail": 3, "averag": 3, "avoid": 3, "awar": 3, "axi": 3, "b": 3, "b_d_e_p": 3, "background": 3, "bait": 3, "balanc": 3, "bar": 3, "base": [2, 3], "bash": 3, "bashcommand": [0, 3], "bashfold": [0, 3], "bashhybrid": [0, 3], "becom": 3, "bed": 3, "bed2len": [0, 3], "bed_df": 3, "bed_minu": 3, "bed_path": 3, "bed_plu": 3, "bed_row": 3, "bed_split_strand": [0, 3], "befor": 3, "being": 3, "between": 3, "bigwig": [0, 2], "bin": 3, "bincollect3": [0, 3], "binmultiplematric": [0, 3], "binomi": 3, "bioinformat": 0, "biological_process": 3, "black": 3, "blackman": 3, "blue": 3, "bonferroni": 3, "bool": 3, "boolean": 3, "both": [2, 3], "box": 3, "box_plot": 3, "boxcar": 3, "boxplot1": [0, 3], "brca1": 3, "broadpeak": 3, "bulk": 3, "bulkinput": 3, "bulkinputidt": [0, 3], "buri": 3, "bw": 3, "bw_name": 3, "bw_path": 3, "bw_paths_minu": 3, "bw_paths_plu": 3, "bwfwd": 3, "bwfwd_dict": 3, "bwrev": 3, "bwrev_dict": 3, "by4741": 3, "c": [2, 3], "c123": 0, "c123_jk050420_polr2a": 0, "ca": 3, "calcul": [2, 3], "calculatefdr": [0, 3], "calfor": 3, "calgc": [0, 3], "call": 3, "can": [2, 3], "categori": 3, "cc": 3, "cell": 0, "center": 3, "cg": 3, "cgat": 3, "cgau": 3, "cgaucguag": 3, "cgta": 3, "cgtacgtaac": 3, "chang": [1, 3], "char": 3, "charact": 3, "character": 3, "check": 3, "checkvienna": [0, 3], "chimer": 3, "chip": 0, "choos": 3, "chosen": 3, "chr": 3, "chr1": 3, "chri": 3, "chrii": 3, "chrom": 3, "chromosm": 3, "chromosom": [2, 3], "chromosome2profile3end": [0, 3], "chromsom": 3, "chunk": [2, 3], "cigar": 3, "cigar_str": 3, "cigarstr": 3, "class": 3, "clean": 3, "cleannam": [0, 3], "clear": 2, "clip": [0, 3], "cluster": 3, "clusterclustermap": [0, 3], "clustermap": 3, "code": [2, 3], "collect": 3, "color": 3, "color1": 3, "color2": 3, "colormap": 3, "column": [2, 3], "combin": 3, "command": 3, "common": 3, "compar": 3, "compare1toref": [0, 3], "comparemoretoref": [0, 3], "comparison": 3, "comparison_plot": 3, "compat": 3, "complement": [2, 3], "compon": 3, "compress": 2, "comput": 3, "concat": 3, "condition1": [0, 3], "condition2": [0, 3], "condition3": [0, 3], "conduct": 0, "consid": 3, "consist": 3, "constrain": 3, "constraint": 3, "contain": [2, 3], "containinig": 3, "content": 3, "continu": 1, "convent": [2, 3], "convert": [2, 3], "coordin": 3, "core": 3, "correct": 3, "correpond": 3, "correspond": 3, "count": 3, "countdelet": [0, 3], "countmiddl": [0, 3], "countread": [0, 3], "coverag": 3, "covid19": 2, "cpm": 3, "crac": [0, 3], "creat": [2, 3], "criteria": 3, "csv": [2, 3], "csv2pickl": 0, "csv_file": 2, "ct": 3, "cumul": 3, "cumulativediffer": [0, 3], "cumulativepeak": [0, 3], "current": [2, 3], "custom": [2, 3], "cutoff": 3, "d": [2, 3], "data": [0, 2, 3], "data_fil": 3, "data_heatmap": 3, "data_metaplot": 3, "data_path": 3, "data_ref": 3, "databas": 2, "datafram": [2, 3], "datapoint": 3, "dataset": 3, "date": 3, "decim": 3, "default": [2, 3], "define_experi": [0, 3], "del": [2, 3], "delet": [2, 3], "delimit": 2, "delta": 3, "deprec": 3, "describ": 3, "descript": [2, 3], "deseq2": 3, "detail": [2, 3], "details_fil": 2, "develop": 1, "df": 3, "df1": 3, "df2": 3, "df3": 3, "df_detail": 3, "df_input": 3, "dg": [2, 3], "dg65nt": 3, "diagram": 3, "dict": 3, "dictbigwig": [0, 3], "dictionari": 3, "differ": 3, "difference_plot": 3, "differenti": 3, "differential_heatmap": 3, "directori": [2, 3], "dirpath": 3, "discard": 3, "discoveri": 3, "discret": 3, "displai": 3, "disqualifi": 3, "distanc": 3, "divid": [2, 3], "dna": 3, "dna_stretch_posit": [0, 3], "dna_string_posit": [0, 3], "do": 3, "document": 1, "dot": 3, "dpi": 3, "drop": 3, "dtype": 3, "dure": 3, "e": [0, 2, 3], "each": [2, 3], "either": 3, "element": 3, "empti": 3, "enabl": 2, "encount": [0, 1], "end": [2, 3], "end_col": 3, "energi": [2, 3], "enhanc": 3, "enhancedvolcano": [0, 3], "enrich": [0, 3], "entir": 3, "entri": 3, "equal": 3, "equal_weight": 3, "error": 3, "exampl": [2, 3], "except": 3, "exclud": 3, "execut": 3, "exon": 2, "exp": 3, "exp1": 3, "exp2": 3, "expand": [2, 3], "expans": 3, "expdat": [0, 3], "expect": 3, "experi": [0, 2, 3], "experiment": 3, "expid": [0, 3], "explain": 3, "expnamepars": [0, 3], "express": 3, "expstat": [0, 3], "extend": 3, "extendingwindow": [0, 3], "extens": 3, "extract": 3, "extrema": 3, "extrud": [0, 3], "f": [2, 3], "facilit": 0, "fals": 3, "fasta": 2, "fasta2slidingwindow": 0, "fc": 3, "fc_cutoff": 3, "fdr": 3, "featur": [2, 3], "featurecount": 3, "figsiz": 3, "figur": 3, "file": [0, 2, 3], "file1": 3, "file1_fwd": 3, "file1_rev": 3, "file2": 3, "file2_fwd": 3, "file2_rev": 3, "filenam": [2, 3], "filepath": 3, "fill": 3, "fill_between": 3, "fill_na": 3, "fillna": 3, "filter": [2, 3], "filter_df": 3, "filterexp": [0, 3], "final": 3, "find": 3, "find_pol3_termin": [0, 3], "findpeak": [0, 3], "findprim": [0, 3], "findtrough": [0, 3], "first": 3, "fisher": 3, "fix": [1, 2], "flag": 0, "flag_wt_arsenite30min": 0, "flank": 3, "flank_3": 3, "flank_5": 3, "float": 3, "float64": 3, "fmp": 3, "fname": 3, "focu": 0, "fold": [0, 2, 3], "folder": [2, 3], "foldingfrombigwig": [0, 3], "foldnascentelem": [0, 3], "follow": [2, 3], "followin": 0, "form": 3, "format": [2, 3], "forward": 3, "found": [2, 3], "fraction": 3, "frame": 3, "free": 2, "from": [0, 2, 3], "fulli": 1, "function": [0, 1, 3], "futur": 3, "fwd": 3, "g": [2, 3], "ga": 3, "gc": 3, "gcau": 3, "gene": [2, 3], "gene1": 3, "gene_nam": 3, "genefrombigwig": [0, 3], "genelist": 3, "gener": [2, 3], "generatesubplot": [0, 3], "genom": [2, 3], "genome2windowsgtf": 0, "genomenascentfold": 0, "get": 3, "get_bw_data": [0, 3], "get_enrich": [0, 3], "get_multiple_matric": [0, 3], "getmultiplematric": [0, 3], "getmultiplematricesfrompeak": [0, 3], "getseqdata": [0, 3], "gg": 3, "github": [0, 1], "give": 2, "given": [2, 3], "go": 3, "go_dataset": 3, "go_enrich": 0, "goterm": [0, 3], "grch38": 3, "green": 3, "grid": 3, "group": 3, "groupcigar": [0, 3], "groupcracsampl": [0, 3], "gt": 3, "gtf": [2, 3], "gtf2": 3, "gtf_file": 3, "gtf_object": 3, "gtf_path": 3, "gucuguuuugugg": 3, "gzip": [2, 3], "h": 3, "h_line": 3, "hairpin": 3, "handl": 3, "handleinput": [0, 3], "head": 3, "heatmap": 3, "helper": 3, "high": 0, "highlight": 3, "home": 2, "horizont": 3, "hplotstarstats_read": [0, 3], "hplotstarstats_readlen": [0, 3], "htp": 3, "htseq": 3, "hybrid": [0, 3], "i": [0, 2, 3], "id": 3, "identifi": 3, "inch": 3, "includ": 3, "index": [0, 3], "indexord": [0, 3], "indic": 3, "individu": 3, "inf": 3, "inform": [0, 3], "initi": 1, "inner": 3, "inner_end": 3, "inner_start": 3, "input": [2, 3], "input_df": 3, "input_valu": 3, "insert": 3, "insid": 3, "instead": 3, "int": [2, 3], "int64": 3, "integ": 3, "integr": 3, "intern": 3, "is_insid": [0, 3], "issu": [0, 1], "item": 3, "item1": 3, "item2": 3, "iter": 3, "its": 3, "jan": 1, "jk": 0, "join": 3, "join2d": [0, 3], "join_strand_matric": [0, 3], "keep": 3, "keepnam": 3, "kei": 3, "kept": 3, "kmean": 3, "l": [2, 3], "label": 3, "label1": 3, "label2": 3, "last": 3, "lc": 3, "legend": 3, "len": [2, 3], "length": [2, 3], "length_col": 3, "lengthmax": 3, "let_in": 3, "let_out": 3, "letter": 3, "lettercont": [0, 3], "librari": 3, "lightgreen": 3, "lightpink": 3, "lim": 3, "limit": 3, "line": 3, "linker": 3, "list": [2, 3], "list_fil": 2, "list_of_gen": 2, "listpath": [0, 3], "lit": 3, "load": [2, 3], "loadgtf": [0, 3], "local": 3, "local_po": 3, "locat": 3, "log": 3, "log2": 3, "log2foldchang": 3, "logarithm": 3, "lognam": 3, "long": [2, 3], "longer": 3, "look": 3, "loop": [0, 3], "loopslist": 3, "loopstem": [0, 3], "lstem": [0, 3], "m": 3, "magenta": 3, "mai": 3, "main": 0, "major": 0, "map": 3, "mark": 3, "match": 3, "matric": 3, "matrix": 3, "matrix_dict": 3, "matrix_from_bigwig": [0, 3], "matrixfrombigwig": [0, 3], "max": 3, "max_len": 3, "max_t": 3, "maximum": 3, "mean": 3, "median": 3, "memori": 3, "merg": [2, 3], "merge2d": [0, 3], "mergesalmon": 0, "metaplot": 3, "metaprofil": 0, "metaprofileandheatmap": [0, 3], "method": 0, "middl": 3, "miko\u0142ajczyk": 1, "million": 3, "min": 3, "min_len": 3, "min_t": 3, "minim": 3, "minima": 3, "minimum": 3, "minlen": 3, "minor": 1, "minu": [2, 3], "minus_dict": 3, "mismatch": 3, "mm": 3, "modifi": 3, "modul": 0, "molecular_funct": 3, "more": 3, "multiloop": 3, "multipl": [2, 3], "multistem": 3, "multistemslist": 3, "must": 3, "n": [2, 3], "n_compon": 3, "name": [0, 2, 3], "name2index": [0, 3], "name_cmap": 3, "name_col": 3, "nameelem": 3, "nan": 3, "narrowpeak": 3, "nascent": 0, "nascentelem": [0, 3], "nascentelemsdatafram": [0, 3], "nascentfold": [0, 3], "need": 3, "neg": 2, "nest": 3, "nested_region_cleanup": [0, 3], "net": 0, "new": [1, 3], "non": [2, 3], "noncod": [2, 3], "noncoded2profil": [0, 3], "noncoded2profile1": [0, 3], "noncoded_fwd": 3, "noncoded_pa": 3, "noncoded_raw": 3, "noncoded_rev": 3, "noncoded_suffix": [2, 3], "none": [2, 3], "nontemplatedna": [0, 3], "normal": [0, 3], "normalize_intern": 3, "normalize_libs": 3, "notat": 3, "note": 2, "np": 3, "nroadpeak": 3, "nt": [2, 3], "ntotal": [0, 3], "ntotal_param": 3, "nucleotid": 3, "number": 3, "numread": [2, 3], "o": [2, 3], "object": 3, "obtain": 3, "occur": 3, "offset": 3, "oligo": 3, "onc": 3, "one": 3, "ones": 3, "onli": [2, 3], "oper": 3, "option": [2, 3], "order": 3, "organ": 3, "origin": 0, "otherwis": [2, 3], "out": [2, 3], "outer": 3, "outer_end": 3, "outer_start": 3, "output": [2, 3], "output_df_fwd": 3, "output_df_rev": 3, "outsid": 3, "overhang": 3, "overhang5end": 3, "overwrit": 3, "p": [2, 3], "packag": [0, 1], "padj": 3, "panda": 3, "panther": 3, "param": 3, "paramet": 3, "parenthes": 3, "pars": [2, 3], "parseconcatfil": [0, 3], "parsecracnam": [0, 3], "parsefoldingnam": [0, 3], "parsehead": [0, 3], "parsenoncod": [0, 3], "parsenoncodedlist": [0, 3], "parser": 3, "part": 3, "pass": 3, "path": [2, 3], "paths_in": 3, "pc1": 3, "pc2": 3, "pca": [0, 3], "pclimit": 3, "pcval": 3, "pd": 3, "peak": 3, "peak2matric": [0, 3], "peak_file_path": 3, "peak_path": 3, "per": 3, "perform": 3, "person": 0, "pickl": [2, 3], "place": 3, "pleas": [0, 1], "plot": 0, "plot_as_box_plot": [0, 3], "plot_diff": [0, 3], "plot_heatmap": [0, 3], "plot_median": 3, "plot_n": 3, "plot_rang": 3, "plot_to_compar": [0, 3], "plotandfold": [0, 3], "plotstarstat": [0, 3], "plotstarstats_chimer": [0, 3], "plotstarstats_map": [0, 3], "plotstarstats_mistmatch": [0, 3], "plotstarstats_read": [0, 3], "plotstarstats_readlen": [0, 3], "plu": [2, 3], "plus_dict": 3, "png": 3, "point": 3, "pol3": 3, "polr2a": 0, "polya": [2, 3], "posit": [2, 3], "positionsof": 3, "possibl": 3, "postprocess": 0, "pre": 3, "prefix": 2, "preload": 3, "prepar": 3, "preparenasc": [0, 3], "preprocess": [0, 3], "presenc": 3, "present": 3, "primary_assembli": 3, "primer": 3, "princip": 3, "print": 3, "process": [2, 3], "profil": [0, 2], "properli": 3, "protein": [0, 3], "provid": [2, 3], "pseudocount": [0, 3], "pseudocounts_param": 3, "puldown": 0, "pull": 3, "put": 0, "pval_cutoff": 3, "pvalu": 3, "py": [0, 3], "pybigwig": 3, "pycrac": 3, "python": [2, 3], "q": 3, "q1": 3, "q3": 3, "qcut": 3, "qq": 3, "qualifi": 3, "quant": 3, "quantif": 2, "quantil": 3, "quantilecategori": [0, 3], "quartil": 3, "queri": 3, "query_gen": 3, "quick": 3, "rais": 3, "rand": 3, "random": 3, "randomdnaal": [0, 3], "randomdnasingl": [0, 3], "rang": 3, "range5end": 3, "rate": 3, "ratio": 3, "rc": 2, "rdn37": 3, "re": 3, "read": [2, 3], "read_b": [0, 3], "read_deseq": [0, 3], "read_featurecount": [0, 3], "read_htseq_output": [0, 3], "read_list": [0, 3], "read_starstat": [0, 3], "read_tabfil": [0, 3], "readb": [0, 3], "reads2genom": [0, 3], "reads2genome3end": [0, 3], "reads2genome5end": [0, 3], "reads2genomedelet": [0, 3], "reads2profil": [0, 3], "reads2profiledelet": [0, 3], "readsalmon": [0, 3], "recalcul": 3, "recogn": 3, "recommend": 3, "red": 3, "ref": 3, "ref_gen": 3, "ref_label": 3, "ref_organ": 3, "ref_speci": 3, "refer": 3, "region": 3, "region1": 3, "region2": 3, "region_col": 3, "regionscor": [0, 3], "reject": 3, "rel": 3, "releas": 0, "remov": 3, "reorder": 3, "replac": 3, "report": 3, "repres": [2, 3], "requir": [2, 3], "rerenc": 3, "resolut": 3, "result": [2, 3], "retriev": 3, "return": 3, "rev": 3, "revers": [2, 3], "reverse_compl": [0, 3], "reverse_complement_dna": [0, 3], "reverse_complement_rna": [0, 3], "rna": [0, 2, 3], "rna_prim": 3, "rnafold": [0, 2, 3], "rnahybrid": [0, 3], "rnainvert": [0, 3], "rnainvertstal": [0, 3], "rnap": 3, "rnaprim": 3, "rnaseq": 2, "rollinggc": [0, 3], "root": 3, "round": 3, "row": 3, "rpm": 3, "rstem": [0, 3], "rtype": 3, "run": 3, "runpca": [0, 3], "s1": 3, "salmon": [2, 3], "salmon_output": 2, "sam": [0, 2], "sam2genom": [0, 3], "sam2genome3end": 3, "sam2genome5end": 3, "sam2profil": [0, 3], "sam2profilesgenom": [0, 3], "sam2profilestranscript": 0, "sam_fil": 2, "same": 2, "samgener": 0, "samgenom": 0, "sampl": [2, 3], "sample1": 3, "sample1_fwd": 3, "sample1_rev": 3, "sample2": 3, "samplerep": 3, "samtranscript": 0, "sapien": 3, "save": [2, 3], "save_csv": [0, 3], "savebigwig": [0, 3], "savedata": 3, "scaffold": 3, "scale": 3, "scheme": 3, "scipi": 3, "score": 3, "script": [0, 3], "search": 3, "second": 3, "secondari": 0, "select": [2, 3], "select_color": [0, 3], "selectend": [0, 3], "selectpolya": [0, 3], "selectsortpath": [0, 3], "selectsubsetmm": [0, 3], "separ": [2, 3], "seq": [0, 3], "seq_col": 3, "sequenc": [0, 2, 3], "sequenceconstrain": [0, 3], "seri": 3, "set": [0, 3], "sf": 3, "shift": 3, "short": 3, "shorter": 3, "side": [2, 3], "sign": 2, "signific": 3, "similar": 3, "simpl": 3, "singl": [2, 3], "size": [2, 3], "skyblu": 3, "slice": 3, "slide": [2, 3], "slidingwindow": [0, 3], "smooth": 3, "soft": 3, "some": [0, 2, 3], "sort": 3, "space": 3, "speci": 3, "specif": [0, 3], "specifi": [2, 3], "spectral": 3, "splice": 3, "split": 3, "ssllss": 3, "stall": [0, 3], "stallful": 3, "star": 3, "start": 3, "start_col": 3, "stat": 3, "statist": 3, "std": 3, "stem": 3, "stem_1": 3, "stemslist": 3, "stop": 3, "str": 3, "strand": [2, 3], "stretch": 3, "string": [2, 3], "strip": 3, "strip_bigwig_nam": [0, 3], "stripbigwignam": [0, 3], "stripcigar": [0, 3], "stripsubstitut": [0, 3], "structur": 3, "structurefil": [0, 3], "sub": 3, "submodul": 0, "subplot": 3, "subprocess": 3, "subset": 3, "subset1": 3, "subset2": 3, "substitut": 3, "substr": 3, "substructur": [0, 3], "substutiotn": 3, "suffix": 3, "sum": 3, "support": 0, "system": 0, "t": [2, 3], "ta": 3, "tab": [2, 3], "tab_output": 2, "tabl": 3, "tag": 3, "take": [2, 3], "target": 3, "target_fdr": 3, "tatist": 3, "taxid": 3, "tc": 3, "temp": [2, 3], "tempdir": 3, "temperatur": [2, 3], "templat": 3, "templatedna": [0, 3], "term": 3, "termin": 3, "test": [0, 1, 3], "test_typ": 3, "testprint": 3, "testscaffold": [0, 3], "tg": 3, "tga": 3, "tgatcggtac": 3, "than": 3, "them": [2, 3], "thi": [0, 2, 3], "three": 3, "threshold": 3, "timestamp": [0, 3], "timestamprandomint": [0, 3], "titl": 3, "to_strip": 3, "toadd": 3, "toclear": [2, 3], "todrop": 3, "tollervei": 2, "toload": 3, "tomasz": 1, "tool": [0, 3], "toorder": [0, 3], "top": 3, "tostrip": 3, "tostripcigarf": [0, 3], "tostripcigarthre": [0, 3], "total": 3, "tpm": 2, "tracker": [0, 1], "transcript": [0, 2, 3], "transcript2profil": [0, 3], "transcript2profiledelet": [0, 3], "transcriptom": 2, "transcriptome_detail": 2, "transcriptome_details_biomart_trna_rrna_updatejan2021": 2, "transctipt": 3, "transform": 3, "tripl": 3, "trna": [0, 3], "trough": 3, "troughput": 0, "true": 3, "trxtool": 1, "tt": 3, "tttgcgcagtcgtgcggggcgcagcgccc": 3, "tupl": 3, "turn": 3, "turowski": 1, "two": 3, "txt": [2, 3], "type": 3, "u": [2, 3], "una": 3, "unafold": [0, 3], "undergo": 1, "unfilt": 3, "union": 3, "uniqu": 3, "updat": 1, "us": [0, 2, 3], "usag": 2, "use_reference_set": 3, "v0": 0, "valid": 3, "valu": [2, 3], "valueerror": 3, "varianc": 3, "variou": 2, "venn": 3, "venndiagram": [0, 3], "verbos": 3, "veri": 3, "version": [0, 3], "vert": 3, "vertic": 3, "vienna": 3, "vienna2format": [0, 3], "viennarna": 3, "visual": 2, "vmax": 3, "vmin": 3, "volcano": 3, "w": [1, 2], "wa": 3, "warn": 3, "warninig": 3, "weight": 3, "when": 3, "where": 3, "whether": 3, "which": 3, "whole": 3, "whole_nam": 3, "wig": 2, "win_typ": 3, "window": [2, 3], "wise": 3, "within": 3, "without": 3, "wont_find_this_str": 3, "word": 3, "work": [2, 3], "wrap": 3, "wrapper": 3, "wt": 0, "x": 3, "x_fc": 3, "xlim": 3, "y": 3, "y_pval": 3, "ylim": 3, "you": [0, 1], "yyyymmdd_hhmmss": 3, "yyyymmdd_hhmmss_randomint": 3, "zero": 3}, "titles": ["trxtools", "Release information", "Scripts", "trxtools package"], "titleterms": {"0": 1, "1": 1, "2": 1, "3": 1, "assai": 3, "author": 1, "bigwig": 3, "bug": [0, 1], "content": 0, "csv2pickl": 2, "fasta2slidingwindow": 2, "genome2windowsgtf": 2, "genomenascentfold": 2, "go_enrich": 3, "indic": 0, "inform": 1, "main": 2, "mergesalmon": 2, "metaprofil": 3, "method": 3, "modul": 3, "nascent": 3, "packag": 3, "plot": 3, "profil": 3, "py": 2, "releas": 1, "report": [0, 1], "sam": 3, "sam2profilesgenom": 2, "sam2profilestranscript": 2, "samgener": 3, "samgenom": 3, "samtranscript": 3, "script": 2, "secondari": 3, "submodul": 3, "support": 2, "tabl": 0, "trxtool": [0, 3], "v0": 1, "version": 1}})