

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>trxtools package &mdash; trxtools 0.3.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=e6c5d56c" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=1e28cc32"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Release information" href="release.html" />
    <link rel="prev" title="Scripts" href="scripts.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            trxtools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">trxtools package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#trxtools-sam-module">trxtools.sam module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.sam.SAMgeneral">trxtools.sam.SAMgeneral submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.countDeletion"><code class="docutils literal notranslate"><span class="pre">countDeletion()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.countMiddle"><code class="docutils literal notranslate"><span class="pre">countMiddle()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.countRead"><code class="docutils literal notranslate"><span class="pre">countRead()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.groupCIGAR"><code class="docutils literal notranslate"><span class="pre">groupCIGAR()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.noncoded2profile"><code class="docutils literal notranslate"><span class="pre">noncoded2profile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.noncoded2profile1"><code class="docutils literal notranslate"><span class="pre">noncoded2profile1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.saveBigWig"><code class="docutils literal notranslate"><span class="pre">saveBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.selectEnds"><code class="docutils literal notranslate"><span class="pre">selectEnds()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.selectNoncodedAndProfile"><code class="docutils literal notranslate"><span class="pre">selectNoncodedAndProfile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.selectSortPaths"><code class="docutils literal notranslate"><span class="pre">selectSortPaths()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.stripCIGAR"><code class="docutils literal notranslate"><span class="pre">stripCIGAR()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.stripSubstitutions"><code class="docutils literal notranslate"><span class="pre">stripSubstitutions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.tostripCIGARfive"><code class="docutils literal notranslate"><span class="pre">tostripCIGARfive()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgeneral.tostripCIGARthree"><code class="docutils literal notranslate"><span class="pre">tostripCIGARthree()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.sam.SAMgenome">trxtools.sam.SAMgenome submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.chromosome2profile3end"><code class="docutils literal notranslate"><span class="pre">chromosome2profile3end()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.parseHeader"><code class="docutils literal notranslate"><span class="pre">parseHeader()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.reads2genome"><code class="docutils literal notranslate"><span class="pre">reads2genome()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.reads2genome3end"><code class="docutils literal notranslate"><span class="pre">reads2genome3end()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.reads2genome5end"><code class="docutils literal notranslate"><span class="pre">reads2genome5end()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.reads2genomeDeletions"><code class="docutils literal notranslate"><span class="pre">reads2genomeDeletions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMgenome.sam2genome"><code class="docutils literal notranslate"><span class="pre">sam2genome()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.sam.SAMtranscripts">trxtools.sam.SAMtranscripts submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMtranscripts.reads2profile"><code class="docutils literal notranslate"><span class="pre">reads2profile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMtranscripts.reads2profileDeletions"><code class="docutils literal notranslate"><span class="pre">reads2profileDeletions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMtranscripts.sam2profiles"><code class="docutils literal notranslate"><span class="pre">sam2profiles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMtranscripts.transcript2profile"><code class="docutils literal notranslate"><span class="pre">transcript2profile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.sam.SAMtranscripts.transcript2profileDeletions"><code class="docutils literal notranslate"><span class="pre">transcript2profileDeletions()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trxtools-profiles-module">trxtools.profiles module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.profiles.BigWig">trxtools.profiles.BigWig module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.BigWig.foldingFromBigWig"><code class="docutils literal notranslate"><span class="pre">foldingFromBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.BigWig.geneFromBigWig"><code class="docutils literal notranslate"><span class="pre">geneFromBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.BigWig.geneFromBigWigTree"><code class="docutils literal notranslate"><span class="pre">geneFromBigWigTree()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.BigWig.getSeqData"><code class="docutils literal notranslate"><span class="pre">getSeqData()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.BigWig.strip_BigWig_names"><code class="docutils literal notranslate"><span class="pre">strip_BigWig_names()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.profiles.metaprofiles">trxtools.profiles.metaprofiles module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.bed_split_strands"><code class="docutils literal notranslate"><span class="pre">bed_split_strands()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.binMultipleMatrices"><code class="docutils literal notranslate"><span class="pre">binMultipleMatrices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.getMultipleMatrices"><code class="docutils literal notranslate"><span class="pre">getMultipleMatrices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.getMultipleMatricesFromPeak"><code class="docutils literal notranslate"><span class="pre">getMultipleMatricesFromPeak()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.getMultipleMatricesSparse"><code class="docutils literal notranslate"><span class="pre">getMultipleMatricesSparse()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.get_bw_data"><code class="docutils literal notranslate"><span class="pre">get_bw_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.get_chrom_lens"><code class="docutils literal notranslate"><span class="pre">get_chrom_lens()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.get_multiple_matrices"><code class="docutils literal notranslate"><span class="pre">get_multiple_matrices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.join_strand_matrices"><code class="docutils literal notranslate"><span class="pre">join_strand_matrices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.matrixFromBigWig"><code class="docutils literal notranslate"><span class="pre">matrixFromBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.matrixFromBigWigSparse"><code class="docutils literal notranslate"><span class="pre">matrixFromBigWigSparse()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.matrix_from_bigwig"><code class="docutils literal notranslate"><span class="pre">matrix_from_bigwig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.metaprofile"><code class="docutils literal notranslate"><span class="pre">metaprofile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.normalizeToLibrary"><code class="docutils literal notranslate"><span class="pre">normalizeToLibrary()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.peak2matrice"><code class="docutils literal notranslate"><span class="pre">peak2matrice()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.readBED"><code class="docutils literal notranslate"><span class="pre">readBED()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.read_bed"><code class="docutils literal notranslate"><span class="pre">read_bed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.regionScore"><code class="docutils literal notranslate"><span class="pre">regionScore()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.selectSubsetMM"><code class="docutils literal notranslate"><span class="pre">selectSubsetMM()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.metaprofiles.trim_bed"><code class="docutils literal notranslate"><span class="pre">trim_bed()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.profiles.profileTools">trxtools.profiles.profileTools module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.FoldingFromBigWig"><code class="docutils literal notranslate"><span class="pre">FoldingFromBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.binCollect3"><code class="docutils literal notranslate"><span class="pre">binCollect3()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.calculateFDR"><code class="docutils literal notranslate"><span class="pre">calculateFDR()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.compare1toRef"><code class="docutils literal notranslate"><span class="pre">compare1toRef()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.compareMoretoRef"><code class="docutils literal notranslate"><span class="pre">compareMoretoRef()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.dictBigWig"><code class="docutils literal notranslate"><span class="pre">dictBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.findPeaks"><code class="docutils literal notranslate"><span class="pre">findPeaks()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.findTroughs"><code class="docutils literal notranslate"><span class="pre">findTroughs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.geneFromBigWig"><code class="docutils literal notranslate"><span class="pre">geneFromBigWig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.ntotal"><code class="docutils literal notranslate"><span class="pre">ntotal()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.parseConcatFile"><code class="docutils literal notranslate"><span class="pre">parseConcatFile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.preprocess"><code class="docutils literal notranslate"><span class="pre">preprocess()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.pseudocounts"><code class="docutils literal notranslate"><span class="pre">pseudocounts()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.save_csv"><code class="docutils literal notranslate"><span class="pre">save_csv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.profiles.profileTools.stripBigWignames"><code class="docutils literal notranslate"><span class="pre">stripBigWignames()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trxtools-folding-module">trxtools.folding module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.folding.assays">trxtools.folding.assays submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.bulkInputIDT"><code class="docutils literal notranslate"><span class="pre">bulkInputIDT()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.extruded"><code class="docutils literal notranslate"><span class="pre">extruded()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.findPrimer"><code class="docutils literal notranslate"><span class="pre">findPrimer()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.nonTemplateDNA"><code class="docutils literal notranslate"><span class="pre">nonTemplateDNA()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.sequenceConstrain"><code class="docutils literal notranslate"><span class="pre">sequenceConstrain()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.stalled"><code class="docutils literal notranslate"><span class="pre">stalled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.structureFile"><code class="docutils literal notranslate"><span class="pre">structureFile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.templateDNA"><code class="docutils literal notranslate"><span class="pre">templateDNA()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.testScaffold"><code class="docutils literal notranslate"><span class="pre">testScaffold()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.assays.toOrder"><code class="docutils literal notranslate"><span class="pre">toOrder()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.folding.nascent">trxtools.folding.nascent submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.Fold"><code class="docutils literal notranslate"><span class="pre">Fold</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.Hybrid"><code class="docutils literal notranslate"><span class="pre">Hybrid</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.extendingWindow"><code class="docutils literal notranslate"><span class="pre">extendingWindow()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.foldNascentElem"><code class="docutils literal notranslate"><span class="pre">foldNascentElem()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.handleInput"><code class="docutils literal notranslate"><span class="pre">handleInput()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.join2d"><code class="docutils literal notranslate"><span class="pre">join2d()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.merge2d"><code class="docutils literal notranslate"><span class="pre">merge2d()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.name2index"><code class="docutils literal notranslate"><span class="pre">name2index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.nascentElems"><code class="docutils literal notranslate"><span class="pre">nascentElems()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.nascentElemsDataFrame"><code class="docutils literal notranslate"><span class="pre">nascentElemsDataFrame()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.nascentFolding"><code class="docutils literal notranslate"><span class="pre">nascentFolding()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.parseFoldingName"><code class="docutils literal notranslate"><span class="pre">parseFoldingName()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.prepareNascent"><code class="docutils literal notranslate"><span class="pre">prepareNascent()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.nascent.slidingWindow"><code class="docutils literal notranslate"><span class="pre">slidingWindow()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.folding.secondary">trxtools.folding.secondary submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.Lstem"><code class="docutils literal notranslate"><span class="pre">Lstem()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.Rstem"><code class="docutils literal notranslate"><span class="pre">Rstem()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.checkVienna"><code class="docutils literal notranslate"><span class="pre">checkVienna()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.loopStems"><code class="docutils literal notranslate"><span class="pre">loopStems()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.loops"><code class="docutils literal notranslate"><span class="pre">loops()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.substructures"><code class="docutils literal notranslate"><span class="pre">substructures()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.test"><code class="docutils literal notranslate"><span class="pre">test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.folding.secondary.vienna2format"><code class="docutils literal notranslate"><span class="pre">vienna2format()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trxtools-tables-module">trxtools.tables module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-trxtools.tables.goEnrichment">trxtools.tables.goEnrichment submodule</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trxtools.tables.goEnrichment.get_enrichment"><code class="docutils literal notranslate"><span class="pre">get_enrichment()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-trxtools.methods">trxtools.methods module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.DNA_stretch_positions"><code class="docutils literal notranslate"><span class="pre">DNA_stretch_positions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.DNA_string_positions"><code class="docutils literal notranslate"><span class="pre">DNA_string_positions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.addCluster"><code class="docutils literal notranslate"><span class="pre">addCluster()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.bashCommand"><code class="docutils literal notranslate"><span class="pre">bashCommand()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.bed2len"><code class="docutils literal notranslate"><span class="pre">bed2len()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.calGC"><code class="docutils literal notranslate"><span class="pre">calGC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.cleanNames"><code class="docutils literal notranslate"><span class="pre">cleanNames()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.define_experiments"><code class="docutils literal notranslate"><span class="pre">define_experiments()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.enriched"><code class="docutils literal notranslate"><span class="pre">enriched()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.expNameParser"><code class="docutils literal notranslate"><span class="pre">expNameParser()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.expStats"><code class="docutils literal notranslate"><span class="pre">expStats()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.filterExp"><code class="docutils literal notranslate"><span class="pre">filterExp()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.find_pol3_terminators"><code class="docutils literal notranslate"><span class="pre">find_pol3_terminators()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.groupCRACsamples"><code class="docutils literal notranslate"><span class="pre">groupCRACsamples()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.indexOrder"><code class="docutils literal notranslate"><span class="pre">indexOrder()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.is_inside"><code class="docutils literal notranslate"><span class="pre">is_inside()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.letterContent"><code class="docutils literal notranslate"><span class="pre">letterContent()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.listPaths"><code class="docutils literal notranslate"><span class="pre">listPaths()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.loadGTF"><code class="docutils literal notranslate"><span class="pre">loadGTF()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.nested_region_cleanup"><code class="docutils literal notranslate"><span class="pre">nested_region_cleanup()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.normalize"><code class="docutils literal notranslate"><span class="pre">normalize()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.parseCRACname"><code class="docutils literal notranslate"><span class="pre">parseCRACname()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.quantileCategory"><code class="docutils literal notranslate"><span class="pre">quantileCategory()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.randomDNAall"><code class="docutils literal notranslate"><span class="pre">randomDNAall()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.randomDNAsingle"><code class="docutils literal notranslate"><span class="pre">randomDNAsingle()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.readSalmon"><code class="docutils literal notranslate"><span class="pre">readSalmon()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_DEseq"><code class="docutils literal notranslate"><span class="pre">read_DEseq()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_HTSeq_output"><code class="docutils literal notranslate"><span class="pre">read_HTSeq_output()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_STARstats"><code class="docutils literal notranslate"><span class="pre">read_STARstats()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_featureCount"><code class="docutils literal notranslate"><span class="pre">read_featureCount()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_list"><code class="docutils literal notranslate"><span class="pre">read_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.read_tabFile"><code class="docutils literal notranslate"><span class="pre">read_tabFile()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.reverse_complement"><code class="docutils literal notranslate"><span class="pre">reverse_complement()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.reverse_complement_DNA"><code class="docutils literal notranslate"><span class="pre">reverse_complement_DNA()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.reverse_complement_RNA"><code class="docutils literal notranslate"><span class="pre">reverse_complement_RNA()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.rollingGC"><code class="docutils literal notranslate"><span class="pre">rollingGC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.runPCA"><code class="docutils literal notranslate"><span class="pre">runPCA()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.timestamp"><code class="docutils literal notranslate"><span class="pre">timestamp()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.methods.timestampRandomInt"><code class="docutils literal notranslate"><span class="pre">timestampRandomInt()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-trxtools.plotting">trxtools.plotting module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.GOterm"><code class="docutils literal notranslate"><span class="pre">GOterm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.PCA"><code class="docutils literal notranslate"><span class="pre">PCA()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.boxplot1"><code class="docutils literal notranslate"><span class="pre">boxplot1()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.clusterClusterMap"><code class="docutils literal notranslate"><span class="pre">clusterClusterMap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.cumulativeDifference"><code class="docutils literal notranslate"><span class="pre">cumulativeDifference()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.cumulativePeaks"><code class="docutils literal notranslate"><span class="pre">cumulativePeaks()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.enhancedVolcano"><code class="docutils literal notranslate"><span class="pre">enhancedVolcano()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.generateSubplots"><code class="docutils literal notranslate"><span class="pre">generateSubplots()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.hplotSTARstats_readLen"><code class="docutils literal notranslate"><span class="pre">hplotSTARstats_readLen()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.hplotSTARstats_reads"><code class="docutils literal notranslate"><span class="pre">hplotSTARstats_reads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.metaprofileAndHeatmap"><code class="docutils literal notranslate"><span class="pre">metaprofileAndHeatmap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotAndFolding"><code class="docutils literal notranslate"><span class="pre">plotAndFolding()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats"><code class="docutils literal notranslate"><span class="pre">plotSTARstats()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats_chimeric"><code class="docutils literal notranslate"><span class="pre">plotSTARstats_chimeric()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats_mapping"><code class="docutils literal notranslate"><span class="pre">plotSTARstats_mapping()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats_mistmatches"><code class="docutils literal notranslate"><span class="pre">plotSTARstats_mistmatches()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats_readLen"><code class="docutils literal notranslate"><span class="pre">plotSTARstats_readLen()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plotSTARstats_reads"><code class="docutils literal notranslate"><span class="pre">plotSTARstats_reads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plot_as_box_plot"><code class="docutils literal notranslate"><span class="pre">plot_as_box_plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plot_diff"><code class="docutils literal notranslate"><span class="pre">plot_diff()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plot_heatmap"><code class="docutils literal notranslate"><span class="pre">plot_heatmap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.plot_to_compare"><code class="docutils literal notranslate"><span class="pre">plot_to_compare()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.select_colors"><code class="docutils literal notranslate"><span class="pre">select_colors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#trxtools.plotting.vennDiagram"><code class="docutils literal notranslate"><span class="pre">vennDiagram()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="release.html">Release information</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">trxtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">trxtools package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/trxtools.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="trxtools-package">
<h1>trxtools package<a class="headerlink" href="#trxtools-package" title="Link to this heading"></a></h1>
<section id="trxtools-sam-module">
<h2>trxtools.sam module<a class="headerlink" href="#trxtools-sam-module" title="Link to this heading"></a></h2>
<section id="module-trxtools.sam.SAMgeneral">
<span id="trxtools-sam-samgeneral-submodule"></span><h3>trxtools.sam.SAMgeneral submodule<a class="headerlink" href="#module-trxtools.sam.SAMgeneral" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.countDeletion">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countDeletion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.countDeletion" title="Link to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list of mapped deletions. Each deletion is counted once, but expand parameter can merge some longer deletions (common expansion).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nucleotides to expand for each side, defaults to 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countDeletion</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([415])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">countDeletion</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">),</span><span class="n">expand</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([412, 413, 414, 415, 416, 417, 418])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.countMiddle">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countMiddle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.countMiddle" title="Link to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list with the middle of mapped read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nucleotides to expand for each side, defaults to 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countMiddle</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([412])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">countMiddle</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">),</span><span class="n">expand</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([409, 410, 411, 412, 413, 414])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.countRead">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countRead</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.countRead" title="Link to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list of mapped positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countRead</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412,</span>
<span class="go">   413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.groupCIGAR">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">groupCIGAR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cigar_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.groupCIGAR" title="Link to this definition"></a></dt>
<dd><p>Split CIGAR string to list of tuples</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cigar_string</strong> (<em>str</em>)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of tuples <code class="docutils literal notranslate"><span class="pre">[(</span> <span class="pre">),()]</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">groupCIGAR</span><span class="p">(</span><span class="s2">&quot;3S44M1S1H&quot;</span><span class="p">)</span>
<span class="go">[(&#39;3&#39;, &#39;S&#39;), (&#39;44&#39;, &#39;M&#39;), (&#39;1&#39;, &#39;S&#39;), (&#39;1&#39;, &#39;H&#39;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.noncoded2profile">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">noncoded2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_input=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.noncoded2profile" title="Link to this definition"></a></dt>
<dd><p>Turns non-coded ends into profile</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_input</strong> (<em>DataFrame</em>) – output of parseNoncoded function</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – chromosome lengths</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with profiles</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.noncoded2profile1">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">noncoded2profile1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.noncoded2profile1" title="Link to this definition"></a></dt>
<dd><p>Turns non-coded ends into profile</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_input</strong> (<em>DataFrame</em>) – output of parseNoncodedList function</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – chromosome lengths</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with profiles</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.saveBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">saveBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chroms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.saveBigWig" title="Link to this definition"></a></dt>
<dd><p>Save gzip pickle data to BigWig
This function saves data from gzip pickle files to a BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>paths</strong> (<em>dict</em>) – Dictionary with paths to gzip pickle files, where keys are file paths and values are file contents.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to be removed from chromosome names in the paths, defaults to an empty string.</p></li>
<li><p><strong>bw_name</strong> (<em>str</em>) – Name of the output BigWig file.</p></li>
<li><p><strong>chroms</strong> (<em>list</em><em> of </em><em>tuples</em>) – List of tuples with chromosome names and their lengths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Success message indicating the suffix and that the file was saved successfully.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">paths</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;/path/to/file_chr1.pkl.gz&#39;</span><span class="p">:</span> <span class="s1">&#39;data1&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/file_chr2.pkl.gz&#39;</span><span class="p">:</span> <span class="s1">&#39;data2&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;3end&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_name</span> <span class="o">=</span> <span class="s1">&#39;output.bw&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chroms</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1000000</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="mi">2000000</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">saveBigWig</span><span class="p">(</span><span class="n">paths</span><span class="p">,</span> <span class="n">suffix</span><span class="p">,</span> <span class="n">bw_name</span><span class="p">,</span> <span class="n">chroms</span><span class="p">)</span>
<span class="go">&#39;3end saved successfully&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.selectEnds">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">selectEnds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends='polyA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.selectEnds" title="Link to this definition"></a></dt>
<dd><p>Wrapper for functions selecting non-coded ends</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>DataFrame</em>) – output of parseNoncoded or parseNoncodedList</p></li>
<li><p><strong>ends</strong> (<em>str</em>) – type of ends, currently only “polyA” is availible, defaults to “polyA”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>runs selectPolyA</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.selectNoncodedAndProfile">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">selectNoncodedAndProfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minLen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxLen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tail</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'A'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">content</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_intermediate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.selectNoncodedAndProfile" title="Link to this definition"></a></dt>
<dd><p>Select sequences with non-coded ends and profile their occurrence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>list</em>) – list of tuples where each tuple contains a position of 3’ end of read match and a sequence string, defaults to []</p></li>
<li><p><strong>minLen</strong> (<em>int</em><em>, </em><em>optional</em>) – minimum length of the sequence to be considered, defaults to 3</p></li>
<li><p><strong>tail</strong> (<em>str</em><em>, </em><em>optional</em>) – the non-coded end sequence to look for, defaults to “AAA”</p></li>
<li><p><strong>letter</strong> (<em>str</em><em>, </em><em>optional</em>) – the letter to check the content of in the sequence, defaults to “A”</p></li>
<li><p><strong>content</strong> (<em>float</em><em>, </em><em>optional</em>) – the minimum content threshold of the letter in the sequence, defaults to &gt;=0.80</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a pandas Series profiling the occurrence of selected sequences</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">l</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;AAA&quot;</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;AACAA&quot;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selectNoncodedAndProfile</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">minLen</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">tail</span><span class="o">=</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="n">letter</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="n">content</span><span class="o">=</span><span class="mf">0.80</span><span class="p">)</span>
<span class="go">1    1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.selectSortPaths">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">selectSortPaths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chroms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.selectSortPaths" title="Link to this definition"></a></dt>
<dd><p>Select and sort paths based on chromosome names and suffix
This function filters and sorts the paths dictionary based on the provided chromosome names and suffix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>paths</strong> (<em>dict</em>) – Dictionary with paths to gzip pickle files, where keys are file paths and values are file contents.</p></li>
<li><p><strong>chroms</strong> (<em>list</em><em> of </em><em>tuples</em>) – List of tuples with chromosome names and their lengths.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to be matched in the file paths, defaults to an empty string.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with filtered and sorted paths.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">paths</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;/path/to/file_chr1_3end.pkl.gz&#39;</span><span class="p">:</span> <span class="s1">&#39;data1&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/file_chr2_3end.pkl.gz&#39;</span><span class="p">:</span> <span class="s1">&#39;data2&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chroms</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1000000</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="mi">2000000</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;3end&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selectSortPaths</span><span class="p">(</span><span class="n">paths</span><span class="p">,</span> <span class="n">chroms</span><span class="p">,</span> <span class="n">suffix</span><span class="p">)</span>
<span class="go">{&#39;/path/to/file_chr1_3end.pkl.gz&#39;: &#39;data1&#39;, &#39;/path/to/file_chr2_3end.pkl.gz&#39;: &#39;data2&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.stripCIGAR">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">stripCIGAR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_strip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'H'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.stripCIGAR" title="Link to this definition"></a></dt>
<dd><p>Removes H from output of groupCIGAR</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p></li>
<li><p><strong>to_strip</strong> (<em>str</em><em>, </em><em>optional</em>) – CIGAR mark to be stripped, defaults to “H”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>modified list of tuples</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.stripSubstitutions">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">stripSubstitutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.stripSubstitutions" title="Link to this definition"></a></dt>
<dd><p>Strip substutiotns on both ends</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of tuples</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.tostripCIGARfive">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">tostripCIGARfive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.tostripCIGARfive" title="Link to this definition"></a></dt>
<dd><p>Calculates length of soft-clipped nucleotides at the 5’ end of the read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>number of substituted nucleotides</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tostripCIGARfive</span><span class="p">([(</span><span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;44&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">)])</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgeneral.tostripCIGARthree">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">tostripCIGARthree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgeneral.tostripCIGARthree" title="Link to this definition"></a></dt>
<dd><p>Nucleotides without alignment at the 3’ end of the read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>number of substituted nucleotides</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tostripCIGARthree</span><span class="p">([(</span><span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;44&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">)])</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-trxtools.sam.SAMgenome">
<span id="trxtools-sam-samgenome-submodule"></span><h3>trxtools.sam.SAMgenome submodule<a class="headerlink" href="#module-trxtools.sam.SAMgenome" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.chromosome2profile3end">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">chromosome2profile3end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FWD'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.chromosome2profile3end" title="Link to this definition"></a></dt>
<dd><p>Generates profile for the 3’ ends of reads and saves position of non-coded end</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>list</em>) – list of triple tuples (position, cigar_string, sequence), defaults to []</p></li>
<li><p><strong>length</strong> (<em>int</em>) – length of chromosome</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – <code class="docutils literal notranslate"><span class="pre">'FWD'</span></code> or <code class="docutils literal notranslate"><span class="pre">'REV'</span></code>, defaults to ‘FWD’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>profile, noncoded</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array, list of tuples</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="p">[(</span><span class="mi">10</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">,</span><span class="s2">&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;</span><span class="p">)],</span><span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="s2">&quot;FWD&quot;</span><span class="p">)</span>
<span class="go">(0     0.0</span>
<span class="go">1     0.0</span>
<span class="go">...</span>
<span class="go">34    0.0</span>
<span class="go">35    1.0</span>
<span class="go">36    0.0</span>
<span class="go">...</span>
<span class="go">50    0.0</span>
<span class="go">dtype: float64,</span>
<span class="go">[(35, &#39;CC&#39;)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="p">[(</span><span class="mi">40</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">,</span><span class="s2">&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;</span><span class="p">)],</span><span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="s2">&quot;REV&quot;</span><span class="p">)</span>
<span class="go">(0     0.0</span>
<span class="go">1     0.0</span>
<span class="go">...</span>
<span class="go">39    0.0</span>
<span class="go">40    1.0</span>
<span class="go">41    0.0</span>
<span class="go">...</span>
<span class="go">50    0.0</span>
<span class="go">dtype: float64,</span>
<span class="go">[(40, &#39;AAA&#39;)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.parseHeader">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">parseHeader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.parseHeader" title="Link to this definition"></a></dt>
<dd><p>Parses the header of a SAM file to extract chromosome names and lengths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – Path to the SAM file</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Name of the experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – Directory path to save intermediate files</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with chromosome names and lengths</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parseHeader</span><span class="p">(</span><span class="s2">&quot;example.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;experiment1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/dir&quot;</span><span class="p">)</span>
<span class="go">chrName  length</span>
<span class="go">chr1     248956422</span>
<span class="go">chr2     242193529</span>
<span class="go">...</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.reads2genome">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='read'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logName=''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.reads2genome" title="Link to this definition"></a></dt>
<dd><p>Function used by sam2genome. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>)</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.reads2genome3end">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome3end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='3end'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends='polyA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logName=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncLen_min=3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncLen_max=30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_intermediate=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.reads2genome3end" title="Link to this definition"></a></dt>
<dd><p>Function used by sam2genome3end. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>)</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
<li><p><strong>noncoded</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True then will parse and save non-coded ends, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log, noncoded_fwd, noncoded_rev</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list, DataFrame, DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.reads2genome5end">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome5end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='5end'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logName=''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.reads2genome5end" title="Link to this definition"></a></dt>
<dd><p>Function used by sam2genome5end. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>)</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.reads2genomeDeletions">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genomeDeletions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='del'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logName=''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.reads2genomeDeletions" title="Link to this definition"></a></dt>
<dd><p>Function used by sam2genome. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>)</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMgenome.sam2genome">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">sam2genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'3end'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'polyA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncLen_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncLen_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_intermediate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMgenome.sam2genome" title="Link to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles. Executed using wrapping script SAM2profilesGenomic.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>)</p></li>
<li><p><strong>path</strong> (<em>str</em>)</p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – element of filename to be removed, defaults to ‘’</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Read SAM file in chunks, defaults to 0</p></li>
<li><p><strong>use</strong> (<em>str</em><em>, </em><em>optional</em>) – <code class="docutils literal notranslate"><span class="pre">'read'</span></code>, <code class="docutils literal notranslate"><span class="pre">'3end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'5end'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'del'</span></code>, defaults to ‘3end’</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – For deletions position can be expanded by the value on each side (e=5 gives 10 nt long), defaults to 0</p></li>
<li><p><strong>noncoded</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save non-coded ends, defaults to True</p></li>
<li><p><strong>noncoded_suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix for non-coded ends, defaults to ‘polyA’</p></li>
<li><p><strong>ncLen_min</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimal length of non-coded tail for noncoded_suffix=polyAmm, defaults to 5</p></li>
<li><p><strong>ncLen_max</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximal length of non-coded tail for noncoded_suffix=polyAmm, defaults to 10</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-trxtools.sam.SAMtranscripts">
<span id="trxtools-sam-samtranscripts-submodule"></span><h3>trxtools.sam.SAMtranscripts submodule<a class="headerlink" href="#module-trxtools.sam.SAMtranscripts" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMtranscripts.reads2profile">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">reads2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMtranscripts.reads2profile" title="Link to this definition"></a></dt>
<dd><p>Takes a list of aligned reads and transforms them into a profile for each transcript. Tested only for PLUS strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – The name of the file (without extension) containing the aligned reads.</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – The directory path where the file is located.</p></li>
<li><p><strong>df_details</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing details about the transcripts, including their lengths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the output DataFrame with profiles and a log list with status messages.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple (pandas.DataFrame, list)</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">output_df</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="n">reads2profile</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;aligned_reads&#39;</span><span class="p">,</span> <span class="n">dirPath</span><span class="o">=</span><span class="s1">&#39;/path/to/files&#39;</span><span class="p">,</span> <span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMtranscripts.reads2profileDeletions">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">reads2profileDeletions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand=5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMtranscripts.reads2profileDeletions" title="Link to this definition"></a></dt>
<dd><p>Takes a list of aligned reads and transforms them into profiles for each transcript, including deletions. Tested only for PLUS strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – The name of the file (without extension) containing the aligned reads.</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – The directory path where the file is located.</p></li>
<li><p><strong>df_details</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing details about the transcripts, including their lengths.</p></li>
<li><p><strong>expand</strong> (<em>int</em>) – The number of bases to expand around deletions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the output DataFrames with profiles (reads, deletions, deletions with expansion) and a log list with status messages.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple (pandas.DataFrame, pandas.DataFrame, pandas.DataFrame, list)</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">output_df</span><span class="p">,</span> <span class="n">outputDel_df</span><span class="p">,</span> <span class="n">outputDelExt_df</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="n">reads2profileDeletions</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;aligned_reads&#39;</span><span class="p">,</span> <span class="n">dirPath</span><span class="o">=</span><span class="s1">&#39;/path/to/files&#39;</span><span class="p">,</span> <span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMtranscripts.sam2profiles">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">sam2profiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneList=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deletions=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMtranscripts.sam2profiles" title="Link to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles. Executed using wrapping script SAM2profiles.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The name of the SAM file to be processed.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – The directory path where the SAM file is located.</p></li>
<li><p><strong>geneList</strong> (<em>list</em>) – List of transcripts to be selected.</p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – Element of filename to be removed, defaults to ‘’.</p></li>
<li><p><strong>df_details</strong> (<em>pandas.DataFrame</em>) – DataFrame containing details of transcripts.</p></li>
<li><p><strong>deletions</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to generate profile of deletions, defaults to False.</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bases to expand around deletions, defaults to 5.</p></li>
<li><p><strong>pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save output in pickle format, defaults to False.</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of chunks to read SAM file in, defaults to 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sam2profiles</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example.sam&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;/path/to/files&quot;</span><span class="p">,</span> <span class="n">geneList</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene1&quot;</span><span class="p">,</span> <span class="s2">&quot;gene2&quot;</span><span class="p">],</span> <span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMtranscripts.transcript2profile">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">transcript2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMtranscripts.transcript2profile" title="Link to this definition"></a></dt>
<dd><p>Takes a list of tuples (position, CIGARstring) and generates a profile. 
Works with entire reads, for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>list</em>) – List of tuples where each tuple contains a position and a CIGAR string.</p></li>
<li><p><strong>length</strong> (<em>int</em>) – The length of the profile to be generated.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas Series representing the profile, with positions as the index and counts as the values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">l</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;10M&#39;</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s1">&#39;5M&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">length</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">profile</span> <span class="o">=</span> <span class="n">transcript2profile</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">profile</span><span class="p">)</span>
<span class="go">1    1.0</span>
<span class="go">2    2.0</span>
<span class="go">3    2.0</span>
<span class="go">4    2.0</span>
<span class="go">5    2.0</span>
<span class="go">6    2.0</span>
<span class="go">7    1.0</span>
<span class="go">8    1.0</span>
<span class="go">9    1.0</span>
<span class="go">10   1.0</span>
<span class="go">dtype: float64</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.sam.SAMtranscripts.transcript2profileDeletions">
<span class="sig-prename descclassname"><span class="pre">trxtools.sam.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">transcript2profileDeletions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.sam.SAMtranscripts.transcript2profileDeletions" title="Link to this definition"></a></dt>
<dd><p>Takes a list of tuples (position, CIGARstring) and generates a profile for deletions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>list</em>) – List of tuples where each tuple contains a position and a CIGAR string.</p></li>
<li><p><strong>expand</strong> (<em>int</em>) – The number of bases to expand around deletions.</p></li>
<li><p><strong>length</strong> (<em>int</em>) – The length of the profile to be generated.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas Series representing the profile, with positions as the index and counts as the values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">profile</span> <span class="o">=</span> <span class="n">transcript2profileDeletions</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">expand</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="trxtools-profiles-module">
<h2>trxtools.profiles module<a class="headerlink" href="#trxtools-profiles-module" title="Link to this heading"></a></h2>
<section id="module-trxtools.profiles.BigWig">
<span id="trxtools-profiles-bigwig-module"></span><h3>trxtools.profiles.BigWig module<a class="headerlink" href="#module-trxtools.profiles.BigWig" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.BigWig.foldingFromBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.BigWig.</span></span><span class="sig-name descname"><span class="pre">foldingFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">range5end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.BigWig.foldingFromBigWig" title="Link to this definition"></a></dt>
<dd><p>Extracts folding energy for a given gene from BigWig files, 
and applies a folding offset to the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – Name of the gene to extract data for.</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the directory containing BigWig files.</p></li>
<li><p><strong>data_files</strong> (<em>list</em>) – List of BigWig file names to extract data from.</p></li>
<li><p><strong>gtf</strong> (<em>GTF</em>) – GTF file object containing gene annotations.</p></li>
<li><p><strong>ranges</strong> (<em>int</em><em>, </em><em>optional</em>) – Range of nucleotides to extract, defaults to 0.</p></li>
<li><p><strong>range5end</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bases to exclude from the 5’ end, defaults to 0.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bases to shift the data by, defaults to 15.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing nucleotide sequences and associated data with applied folding offset.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gtf</span> <span class="o">=</span> <span class="n">GTF</span><span class="p">(</span><span class="s2">&quot;/path/to/annotations.gtf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">foldingFromBigWig</span><span class="p">(</span><span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/bigwig/&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;file1.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;file2.bw&quot;</span><span class="p">],</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">range5end</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.BigWig.geneFromBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.BigWig.</span></span><span class="sig-name descname"><span class="pre">geneFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.BigWig.geneFromBigWig" title="Link to this definition"></a></dt>
<dd><p>Extracts nucleotide sequences and associated data for a given gene from BigWig files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – Name of the gene to extract data for.</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the directory containing BigWig files.</p></li>
<li><p><strong>data_files</strong> (<em>list</em>) – List of BigWig file names to extract data from.</p></li>
<li><p><strong>gtf</strong> (<em>GTF</em>) – GTF file object containing gene annotations.</p></li>
<li><p><strong>ranges</strong> – Range of nucleotides to extract, defaults to 0.</p></li>
<li><p><strong>verbose</strong> – If True, prints the name of each BigWig file being processed, defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing nucleotide sequences and associated data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gtf</span> <span class="o">=</span> <span class="n">GTF</span><span class="p">(</span><span class="s2">&quot;/path/to/annotations.gtf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">geneFromBigWig</span><span class="p">(</span><span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/bigwig/&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;file1.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;file2.bw&quot;</span><span class="p">],</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.BigWig.geneFromBigWigTree">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.BigWig.</span></span><span class="sig-name descname"><span class="pre">geneFromBigWigTree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.bw'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.BigWig.geneFromBigWigTree" title="Link to this definition"></a></dt>
<dd><p>Extracts nucleotide sequences and associated data for a given gene from BigWig files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – Name of the gene to extract data for.</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the directory containing subdirectories containing BigWig files.</p></li>
<li><p><strong>gtf</strong> (<em>GTF</em>) – GTF file object containing gene annotations.</p></li>
<li><p><strong>ranges</strong> – Range of nucleotides to extract, defaults to 0.</p></li>
<li><p><strong>verbose</strong> – If True, prints the name of each BigWig file being processed, defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing nucleotide sequences and associated data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gtf</span> <span class="o">=</span> <span class="n">GTF</span><span class="p">(</span><span class="s2">&quot;/path/to/annotations.gtf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">geneFromBigWig</span><span class="p">(</span><span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/bigwig/&quot;</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.BigWig.getSeqData">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.BigWig.</span></span><span class="sig-name descname"><span class="pre">getSeqData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.bw'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.BigWig.getSeqData" title="Link to this definition"></a></dt>
<dd><p>Load BigWig data for a gene/chromosome region and return as pandas Series.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – Name of the gene/chromosome to retrieve data for.</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the directory containing BigWig files.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Base name of the BigWig files (without strand suffix and extension). Can be obtained using strip_BigWig_names().</p></li>
<li><p><strong>gtf</strong> (<em>object</em>) – GTF object that provides gene information such as strand, chromosome, and coordinates.</p></li>
<li><p><strong>ranges</strong> – Number of bases to extend the region on both sides, defaults to 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>BigWig data for the specified gene or region as a pandas Series.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gtf</span> <span class="o">=</span> <span class="n">GTF</span><span class="p">(</span><span class="s1">&#39;/path/to/gtf_file.gtf&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">getSeqData</span><span class="p">(</span><span class="s1">&#39;BRCA1&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/bigwig/&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.BigWig.strip_BigWig_names">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.BigWig.</span></span><span class="sig-name descname"><span class="pre">strip_BigWig_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.bw'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.BigWig.strip_BigWig_names" title="Link to this definition"></a></dt>
<dd><p>Remove _fwd.bw and _rev.bw from the list of BigWig files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>files</strong> (<em>list</em><em>, </em><em>required</em>) – list of filenames, defaults to list()</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of unique names</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">strip_BigWig_names</span><span class="p">([</span><span class="s2">&quot;file1_fwd.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;file1_rev.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;file2_fwd.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;file2_rev.bw&quot;</span><span class="p">])</span>
<span class="go">[&#39;file2&#39;, &#39;file1&#39;]</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-trxtools.profiles.metaprofiles">
<span id="trxtools-profiles-metaprofiles-module"></span><h3>trxtools.profiles.metaprofiles module<a class="headerlink" href="#module-trxtools.profiles.metaprofiles" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.bed_split_strands">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">bed_split_strands</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.bed_split_strands" title="Link to this definition"></a></dt>
<dd><p>Splits a BED dataframe into two separate dataframes based on strand information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – A dataframe containing BED format data. The strand information is expected to be in the 6th column (index 5).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing two dataframes:
- bed_plus (pandas.DataFrame): Dataframe containing entries with the ‘+’ strand.
- bed_minus (pandas.DataFrame): Dataframe containing entries with the ‘-’ strand.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_plus</span><span class="p">,</span> <span class="n">bed_minus</span> <span class="o">=</span> <span class="n">bed_split_strands</span><span class="p">(</span><span class="n">bed_regions</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.binMultipleMatrices">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">binMultipleMatrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">mm={},</span> <span class="pre">bins=[50,</span> <span class="pre">10,</span> <span class="pre">50],</span> <span class="pre">bed_df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[],</span> <span class="pre">flank_5=None,</span> <span class="pre">flank_3=None,</span> <span class="pre">region_col=None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.binMultipleMatrices" title="Link to this definition"></a></dt>
<dd><p>Bin multiple matrices of gene profiles into a single dataframe
:param mm: dictionary of matrices of gene profiles 
:type mm: dict
:param bins: How many bins to use for each region part (5’ flank, region body, 3’ flank), defaults to [50, 10, 50]
:type bins: list, optional
:param bed_df: dataframe in BED format containing genomic coordinates of target regions, defaults to pd.DataFrame()
:type bed_df: pandas.DataFrame, optional
:param flank_5: length of 5’flank to extend BED regions by, defaults to None
:type flank_5: int, optional
:param flank_3: length of 3’flank to extend BED regions by, defaults to None
:type flank_3: int, optional
:raises ValueError: if bins is not a list of three integers
:raises ValueError: if input_value is not an integer or a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary of binned matrices of gene profiles</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatrices</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">binned_mm</span> <span class="o">=</span> <span class="n">binMultipleMatrices</span><span class="p">(</span><span class="n">mm</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="mi">50</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">50</span><span class="p">],</span> <span class="n">bed_df</span><span class="o">=</span><span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.getMultipleMatrices">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">getMultipleMatrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_paths_plus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw_paths_minus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_libsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_zeroes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_baselimit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.getMultipleMatrices" title="Link to this definition"></a></dt>
<dd><p>Get score matrices for positions in given regions (with optional flanks) from multiple BigWig files.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_paths_plus</strong> (<em>list</em>) – list of paths to BigWig files (+ strand)</p></li>
<li><p><strong>bw_paths_minus</strong> (<em>list</em>) – list of paths to BigWig files (- strand)</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – dataframe in BED format containing genomic coordinates of target regions</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 5’ side, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 3’ side, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – _description_, defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to datapoints, defaults to None</p></li>
<li><p><strong>normalize_libsize</strong> (<em>bool</em><em>, </em><em>optional</em>) – normalization to library size (sum of scores in a bigwig file), defaults to True</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – if True, align profiles at the 3’end of features. Otherwise align at 5’end, defaults to False</p></li>
<li><p><strong>len_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – If set, regions longer than len_cutoff will be trimmed to len_cutoff. Incompatible with flank_3 when align_3end=False and with flank_5 when align_3end=True, defaults to None</p></li>
<li><p><strong>skip_zeroes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, regions with no coverage in the BigWig will be skipped to reduce memory usage, defaults to False</p></li>
<li><p><strong>chunk_baselimit</strong> (<em>int</em><em>, </em><em>optional</em>) – If set, data will be processed in chunks to limit memory usage. Value is approximate total number of bases in a chunk, defaults to None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, print progress messages for troubleshooting purposes, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing score matrices for individual BigWig files. Dictionary keys are BigWig file names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatrices</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.getMultipleMatricesFromPeak">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">getMultipleMatricesFromPeak</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peak_paths=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g='references/GRCh38.primary_assembly.genome.cleaned.len'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.getMultipleMatricesFromPeak" title="Link to this definition"></a></dt>
<dd><p>Get score matrices for positions in given regions (with optional flanks) from multiple peak files.
Matrix rows correspond to regions in BED. Columns correpond to nucleotide positions in regions + flanks.
:param peak_paths: list of paths to peak files (broadPeak or narrowPeak format)
:type peak_paths: list
:param bed_df: dataframe in BED format containing genomic coordinates of target regions
:type bed_df: pandas.DataFrame
:param g: Path to genome file (chromosome lengths), defaults to ‘references/GRCh38.primary_assembly.genome.cleaned.len’
:type g: str
:param flank_5: length of 5’flank to extend BED regions by, defaults to 0
:type flank_5: int, optional
:param flank_3: length of 3’flank to extend BED regions by, defaults to 0
:type flank_3: int, optional</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary of score matrices for each peak file</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">peak_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_peaks.narrowPeak&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_peaks.narrowPeak&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatricesFromPeak</span><span class="p">(</span><span class="n">peak_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span> 
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.getMultipleMatricesSparse">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">getMultipleMatricesSparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_paths_plus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw_paths_minus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_libsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_zeroes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_baselimit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.getMultipleMatricesSparse" title="Link to this definition"></a></dt>
<dd><p>getMultipleMatrices optimized for sparse data (features with mostly 0 coverage).
Use when your bait protein binds in well defined sites (e.g TFs) as opposed to e.g. processive helicases or polymerases.
Get score matrices for positions in given regions (with optional flanks) from multiple BigWig files.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_paths_plus</strong> (<em>list</em>) – list of paths to BigWig files (+ strand)</p></li>
<li><p><strong>bw_paths_minus</strong> (<em>list</em>) – list of paths to BigWig files (- strand)</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – dataframe in BED format containing genomic coordinates of target regions</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 5’ side, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 3’ side, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – _description_, defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to datapoints, defaults to None</p></li>
<li><p><strong>normalize_libsize</strong> (<em>bool</em><em>, </em><em>optional</em>) – normalization to library size (sum of scores in a bigwig file), defaults to True</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – if True, align profiles at the 3’end of features. Otherwise align at 5’end, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing score matrices for individual BigWig files. Dictionary keys are BigWig file names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatricesSparse</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.get_bw_data">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">get_bw_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.get_bw_data" title="Link to this definition"></a></dt>
<dd><p>!!! This is an internal function, use getMultipleMatrices() instead.
Retrieve BigWig scores for positions in a given region, optionally including flanks of given length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bed_row</strong> (<em>list</em><em> or </em><em>pandas.Series</em>) – A row from a BED file, containing chromosome, start, end, and strand information.</p></li>
<li><p><strong>bw</strong> (<em>pyBigWig</em>) – A pyBigWig object to retrieve values from.</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of the 5’ flank to include, defaults to 0.</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of the 3’ flank to include, defaults to 0.</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to align the series to the 3’ end, defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas Series containing the BigWig scores for the specified region.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;example.bw&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">get_bw_data</span><span class="p">(</span><span class="n">bed_regions</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">bw</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">align_3end</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.get_chrom_lens">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">get_chrom_lens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.get_chrom_lens" title="Link to this definition"></a></dt>
<dd><p>Get chromosome lengths from a BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bw</strong> (<em>pyBigWig</em>) – A pyBigWig object to retrieve chromosome lengths from.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with chromosome names as keys and their lengths as values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw</span> <span class="o">=</span> <span class="n">pbw</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;example_fwd.bw&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">meta</span><span class="o">.</span><span class="n">get_chrom_lens</span><span class="p">(</span><span class="n">bw</span><span class="p">)</span>
<span class="go">{&#39;chr1&#39;: 248956422,</span>
<span class="go">&#39;chr2&#39;: 242193529,</span>
<span class="go">&#39;chr3&#39;: 198295559,</span>
<span class="go">&#39;chr4&#39;: 190214555,</span>
<span class="go">&#39;chr5&#39;: 181538259,</span>
<span class="go">&#39;chr6&#39;: 170805979,</span>
<span class="go">&#39;chr7&#39;: 159345973,</span>
<span class="go">&#39;chr8&#39;: 145138636,</span>
<span class="go">&#39;chr9&#39;: 138394717,</span>
<span class="go">&#39;chr10&#39;: 133797422,</span>
<span class="go">&#39;chr11&#39;: 135086622,</span>
<span class="go">&#39;chr12&#39;: 133275309,</span>
<span class="go">&#39;chr13&#39;: 114364328,</span>
<span class="go">&#39;chr14&#39;: 107043718,</span>
<span class="go">&#39;chr15&#39;: 101991189,</span>
<span class="go">&#39;chr16&#39;: 90338345,</span>
<span class="go">&#39;chr17&#39;: 83257441,</span>
<span class="go">&#39;chr18&#39;: 80373285,</span>
<span class="go">&#39;chr19&#39;: 58617616,</span>
<span class="go">&#39;chr20&#39;: 64444167,</span>
<span class="go">&#39;chr21&#39;: 46709983,</span>
<span class="go">&#39;chr22&#39;: 50818468,</span>
<span class="go">&#39;chrX&#39;: 156040895,</span>
<span class="go">&#39;chrY&#39;: 57227415,</span>
<span class="go">&#39;chrM&#39;: 16569}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.get_multiple_matrices">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">get_multiple_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_paths_plus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw_paths_minus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_libsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.get_multiple_matrices" title="Link to this definition"></a></dt>
<dd><p>WARNINIG: This function is replaced by getMultipleMatrices()</p>
<p>Get score matrices for positions in given regions (with optional flanks) from multiple BigWig files.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_paths_plus</strong> (<em>list</em>) – list of paths to BigWig files (+ strand)</p></li>
<li><p><strong>bw_paths_minus</strong> (<em>list</em>) – list of paths to BigWig files (- strand)</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – dataframe in BED format containing genomic coordinates of target regions</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 5’ side, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 3’ side, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – _description_, defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to datapoints, defaults to None</p></li>
<li><p><strong>normalize_libsize</strong> (<em>bool</em><em>, </em><em>optional</em>) – normalization to library size (sum of scores in a bigwig file), defaults to True</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – if True, align profiles at the 3’end of features. Otherwise align at 5’end, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing score matrices for individual BigWig files. Dictionary keys are BigWig file names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function is deprecated and will be removed in future versions.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.join_strand_matrices">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">join_strand_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">plus_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minus_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.join_strand_matrices" title="Link to this definition"></a></dt>
<dd><p>Combine score matrices for regions on + and - strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>plus_dict</strong> (<em>dict</em>) – Dictionary containing score matrices for regions on the + strand.</p></li>
<li><p><strong>minus_dict</strong> (<em>dict</em>) – Dictionary containing score matrices for regions on the - strand.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Exception</strong> – If keys in dictionaries do not match.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary containing combined score matrices for regions on both strands.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.matrixFromBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">matrixFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_zeroes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_baselimit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.matrixFromBigWig" title="Link to this definition"></a></dt>
<dd><p>!!! This is an internal function, use getMultipleMatrices() instead.
Get matrix with BigWig scores for all regions in a bed df from a single BigWig file.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.
Flanks are strand-aware.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_path</strong> (<em>str</em>) – Path to target BigWig file</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – DataFrame with BED data (use read_bed())</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 5’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 3’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true replace NaN values with 0 (pybigwig returns positions with 0 coverage as NaN), defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to retrieved scores, defaults to None</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, position 0 in the resulting matrix will be set at the target region’s 3’end instead of 5’end, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with ‘regions’ (numpy array), ‘matrix’ (numpy array), and ‘lengths’ (numpy array) keys</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrixFromBigWig</span><span class="p">(</span><span class="s1">&#39;example.bw&#39;</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">align_3end</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.matrixFromBigWigSparse">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">matrixFromBigWigSparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_zeroes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_baselimit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.matrixFromBigWigSparse" title="Link to this definition"></a></dt>
<dd><p>!!! This is an internal function, use getMultipleMatrices() instead.
matrixFromBigWig optimized for sparse data (features with mostly 0 coverage).
Use when your bait protein binds in well defined sites (e.g TFs) as opposed to e.g. processive helicases or polymerases.
Get matrix with BigWig scores for all regions in a bed df from a single BigWig file.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.
Flanks are strand-aware.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_path</strong> (<em>str</em>) – Path to target BigWig file</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – DataFrame with BED data (use read_bed())</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 5’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 3’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true replace NaN values with 0 (pybigwig returns positions with 0 coverage as NaN), defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to retrieved scores, defaults to None</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, position 0 in the resulting matrix will be set at the target region’s 3’end instead of 5’end, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with ‘regions’ (numpy array), ‘matrix’ (numpy array), and ‘lengths’ (numpy array) keys</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.matrix_from_bigwig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">matrix_from_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.matrix_from_bigwig" title="Link to this definition"></a></dt>
<dd><p>WARNINIG: This function is replaced by matrixFromBigWig()</p>
<p>Get matrix with BigWig scores for all regions in a bed df from a single BigWig file.
Matrix rows correspond to regions in BED.
Columns correpond to nucleotide positions in regions + flanks.
Flanks are strand-aware.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_path</strong> (<em>str</em>) – Path to target BigWig file</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – DataFrame with BED data (use read_bed())</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 5’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 3’flank to extend BED regions by, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true replace NaN values with 0 (pybigwig returns positions with 0 coverage as NaN), defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to retrieved scores, defaults to None</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, position 0 in the resulting matrix will be set at the target region’s 3’end instead of 5’end, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with the result score matrix</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function is deprecated and will be removed in future versions.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.metaprofile">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">metaprofile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_internal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.metaprofile" title="Link to this definition"></a></dt>
<dd><p>Calculate metaprofiles for each sample from score matrices by aggregating each position across all regions.
:param matrix_dict: Dict containing score matrices returned by getMultipleMatrices()
:type matrix_dict: dict
:param agg_type: Type of aggregation to use. Available: ‘mean’, ‘median’, ‘sum’, defaults to ‘mean’
:type agg_type: str, optional
:param normalize_internal: if true, normalize each profile internally (i.e. gene-wise) (x/sum(x)) before aggregating, defaults to False
:type normalize_internal: bool, optional
:raises Exception: “Wrong agg_type; available values: ‘mean’, ‘median’, ‘sum’”</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>dataframe containing metaprofile values for each position in each of the input matrices (i.e. bigwig files)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatrices</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mp</span> <span class="o">=</span> <span class="n">metaprofile</span><span class="p">(</span><span class="n">mm</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mp</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.normalizeToLibrary">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">normalizeToLibrary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metaprofile=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigwig_plus=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigwig_minus=[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.normalizeToLibrary" title="Link to this definition"></a></dt>
<dd><p>Normalize metaprofile to library size (sum of scores in a bigwig file).
:param metaprofile: DataFrame with metaprofile data
:type metaprofile: pandas.DataFrame
:param bigwig_plus: list of paths to BigWig files for + strand
:type bigwig_plus: list
:param bigwig_minus: list of paths to BigWig files for - strand
:type bigwig_minus: list
:raises Exception: if number of + and - strand BigWig files do not match</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>normalized metaprofile DataFrame</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metaprofile_df</span> <span class="o">=</span> <span class="n">trxtools</span><span class="o">.</span><span class="n">metaprofile</span><span class="p">(</span><span class="n">matrix_dict</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normalized_metaprofile</span> <span class="o">=</span> <span class="n">trxtools</span><span class="o">.</span><span class="n">normalizeToLibrary</span><span class="p">(</span><span class="n">metaprofile_df</span><span class="p">,</span> <span class="n">bigwig_plus</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">],</span> <span class="n">bigwig_minus</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.peak2matrice">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">peak2matrice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peak_file_path=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g='references/GRCh38.primary_assembly.genome.cleaned.len'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fillna=True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.peak2matrice" title="Link to this definition"></a></dt>
<dd><p>Get score matrices for positions in given regions (with optional flanks) from a single nroadPeak or narrowPeak file. 
Matrix rows correspond to regions in BED. Columns correpond to nucleotide positions in regions + flanks. Flanks are strand-aware.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – DataFrame with BED data (can use read_bed())</p></li>
<li><p><strong>peak_file_path</strong> (<em>str</em>) – Path to peak file (broadPeak or narrowPeak format)</p></li>
<li><p><strong>g</strong> (<em>str</em>) – Path to genome file (chromosome lengths), defaults to ‘references/GRCh38.primary_assembly.genome.cleaned.len’</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 5’flank to extend BED regions by, strand-aware, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – length of 3’flank to extend BED regions by, strand-aware, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true replace NaN values with 0 (pybigwig returns positions with 0 coverage as NaN), defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with the result score matrix</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.readBED">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">readBED</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.readBED" title="Link to this definition"></a></dt>
<dd><p>Simple BED file parser</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bed_path</strong> (<em>str</em>) – Path to BED file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Contents of BED file in DataFrame form</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_df</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">metaprofiles</span><span class="o">.</span><span class="n">readBED</span><span class="p">(</span><span class="s1">&#39;regions.bed&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.read_bed">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">read_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.read_bed" title="Link to this definition"></a></dt>
<dd><p>Simple BED file parser</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bed_path</strong> (<em>str</em>) – Path to BED file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Contents of BED file in DataFrame form</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function is deprecated and will be removed in future versions.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.regionScore">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">regionScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bw_paths_plus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bw_paths_minus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_libsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.regionScore" title="Link to this definition"></a></dt>
<dd><p>Calculate coverage (when agg_type=’sum’) or statistic for multiple regions in multiple BigWig files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bw_paths_plus</strong> (<em>list</em>) – list of paths to BigWig files (+ strand)</p></li>
<li><p><strong>bw_paths_minus</strong> (<em>list</em>) – list of paths to BigWig files (- strand)</p></li>
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – dataframe in BED format containing genomic coordinates of target regions</p></li>
<li><p><strong>agg_type</strong> (<em>str</em><em>, </em><em>optional</em>) – operation to perform on region scores. Available options: ‘sum’, ‘mean’, ‘median’, defaults to ‘sum’</p></li>
<li><p><strong>flank_5</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 5’ side, defaults to 0</p></li>
<li><p><strong>flank_3</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nt that input regions will be extended by on 3’ side, defaults to 0</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, NaNs will be replaced with zeroes (recommended, as pyBigWig reports positions with 0 score as NaN), defaults to True</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – pseudocounts to add to datapoints, defaults to None</p></li>
<li><p><strong>normalize_libsize</strong> (<em>bool</em><em>, </em><em>optional</em>) – normalization to library size (sum of scores in a bigwig file), defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with calculated scores. Rows are regions/genes, columns are BigWig files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">region_scores</span> <span class="o">=</span> <span class="n">regionScore</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">agg_type</span><span class="o">=</span><span class="s1">&#39;sum&#39;</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.selectSubsetMM">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">selectSubsetMM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.selectSubsetMM" title="Link to this definition"></a></dt>
<dd><p>Select a subset of regions from the matrix dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix_dict</strong> (<em>dict</em>) – Dictionary containing score matrices.</p></li>
<li><p><strong>subset</strong> (<em>Union</em><em>[</em><em>pd.DataFrame</em><em>, </em><em>list</em><em>]</em><em>, </em><em>optional</em>) – DataFrame or list containing the subset of regions to select, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing the selected subset of score matrices.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bw_plus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_plus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_plus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bw_minus_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample1_minus.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample2_minus.bw&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">2500</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mm</span> <span class="o">=</span> <span class="n">getMultipleMatrices</span><span class="p">(</span><span class="n">bw_plus_files</span><span class="p">,</span> <span class="n">bw_minus_files</span><span class="p">,</span> <span class="n">bed_regions</span><span class="p">,</span> <span class="n">flank_5</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flank_3</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">normalize_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subset_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subset_mm</span> <span class="o">=</span> <span class="n">selectSubsetMM</span><span class="p">(</span><span class="n">mm</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="n">subset_regions</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.metaprofiles.trim_bed">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.metaprofiles.</span></span><span class="sig-name descname"><span class="pre">trim_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_3end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_cutoff</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.metaprofiles.trim_bed" title="Link to this definition"></a></dt>
<dd><p>Trim BED region to final length of len_cutoff bases starting from 3’ or 5’ end (strand-aware).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<em>list</em><em> or </em><em>pandas.Series</em>) – A row from a BED file, containing chromosome, start, end, and strand information.</p></li>
<li><p><strong>align_3end</strong> (<em>bool</em>) – If true, the 3’ end will be kept, i.e. the upstream region will be trimmed from the 5’ end of the feature.</p></li>
<li><p><strong>len_cutoff</strong> (<em>int</em>) – The final length to which the region should be trimmed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The trimmed BED row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list or pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_df_trimmed</span> <span class="o">=</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">trim_bed</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">align_3end</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">len_cutoff</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_df</span>
<span class="go">   chrom  start  end  name   score  strand</span>
<span class="go">0  chr1   100    500  gene1  .      -</span>
<span class="go">1  chr2   439    1021 gene2  .      +</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">trim_bed</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">align_3end</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">len_cutoff</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="go">   chrom  start  end  name   score  strand</span>
<span class="go">0  chr1   100    200  gene1  .      -</span>
<span class="go">1  chr2   921    1021 gene2  .      +</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-trxtools.profiles.profileTools">
<span id="trxtools-profiles-profiletools-module"></span><h3>trxtools.profiles.profileTools module<a class="headerlink" href="#module-trxtools.profiles.profileTools" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.FoldingFromBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">FoldingFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwFWD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwREV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dG65nt&#64;30C'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.FoldingFromBigWig" title="Link to this definition"></a></dt>
<dd><p>Pulls folding information from BigWig folding data for a given gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – str, name of the gene</p></li>
<li><p><strong>gtf</strong> (<em>pyCRAC.GTF2</em>) – pyCRAC.GTF2 object with GTF and TAB files loaded</p></li>
<li><p><strong>bwFWD</strong> (<em>dict</em>) – dict of pyBigWig objects for forward strand</p></li>
<li><p><strong>bwREV</strong> (<em>dict</em>) – dict of pyBigWig objects for reverse strand</p></li>
<li><p><strong>ranges</strong> (<em>int</em>) – int, flanks to be added for the gene, default 0</p></li>
<li><p><strong>offset</strong> (<em>int</em>) – int, offset for folding data, default 15</p></li>
<li><p><strong>fold</strong> (<em>str</em>) – str, name of output column, default “<a class="reference external" href="mailto:dG65nt&#37;&#52;&#48;30C">dG65nt<span>&#64;</span>30C</a>”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with folding information</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">FoldingFromBigWig</span><span class="p">(</span><span class="s1">&#39;gene1&#39;</span><span class="p">,</span> <span class="n">gtf_object</span><span class="p">,</span> <span class="n">bwFWD_dict</span><span class="p">,</span> <span class="n">bwREV_dict</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.binCollect3">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">binCollect3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">s1=Series([],</span> <span class="pre">dtype:</span> <span class="pre">float64),</span> <span class="pre">lengths=[500,</span> <span class="pre">72,</span> <span class="pre">500],</span> <span class="pre">bins=[50,</span> <span class="pre">10,</span> <span class="pre">50]</span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.binCollect3" title="Link to this definition"></a></dt>
<dd><p>Collects and sums data from a series into bins of specified lengths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<em>pd.Series</em>) – Series containing the input data</p></li>
<li><p><strong>lengths</strong> (<em>list</em>) – List of lengths for each bin</p></li>
<li><p><strong>bins</strong> (<em>list</em>) – List of number of bins for each length</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with the summed data from the bins</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">binCollect3</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">lengths</span><span class="o">=</span><span class="p">[</span><span class="mi">500</span><span class="p">,</span><span class="mi">72</span><span class="p">,</span><span class="mi">500</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">50</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.calculateFDR">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">calculateFDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations=100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_FDR=0.05</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.calculateFDR" title="Link to this definition"></a></dt>
<dd><p>Calculates False Discovery Rate (FDR) for a given dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.Series</em>) – Series containing the input data</p></li>
<li><p><strong>iterations</strong> (<em>int</em>) – Number of iterations for random dataset generation, default 100</p></li>
<li><p><strong>target_FDR</strong> (<em>float</em>) – Target False Discovery Rate, default 0.05</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with FDR applied</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">calculateFDR</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">iterations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">target_FDR</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.compare1toRef">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">compare1toRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heatmap=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relative=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.compare1toRef" title="Link to this definition"></a></dt>
<dd><p>Takes Series and compare this with reference DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to csv file or DataFrame</p></li>
<li><p><strong>dataset</strong> (<em>pd.Series</em>) – Series containing the dataset to compare</p></li>
<li><p><strong>ranges</strong> (<em>str</em>) – ‘mm’ for min-max or ‘qq’ for q1-q3, default ‘mm’</p></li>
<li><p><strong>heatmap</strong> (<em>bool</em>) – If True, return Series of differences to plot heatmap, default False</p></li>
<li><p><strong>relative</strong> (<em>bool</em>) – If True, recalculate differences according to the peak size, default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame (heatmap=False) or Series (heatmap=True)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame or pd.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">compare1toRef</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="s1">&#39;mm&#39;</span><span class="p">,</span> <span class="n">heatmap</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.compareMoretoRef">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">compareMoretoRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.compareMoretoRef" title="Link to this definition"></a></dt>
<dd><p>Compares a DataFrame created by filter_df with a reference DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to csv file or DataFrame containing reference data</p></li>
<li><p><strong>dataset</strong> (<em>pd.DataFrame</em>) – DataFrame containing the dataset to compare</p></li>
<li><p><strong>ranges</strong> (<em>str</em>) – ‘mm’ for min-max or ‘qq’ for q1-q3, default ‘mm’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with comparison results</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">compareMoretoRef</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="s1">&#39;mm&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.dictBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">dictBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strands</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.dictBigWig" title="Link to this definition"></a></dt>
<dd><p>Preload BigWig files to memory using pyBigWig tools.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>files</strong> (<em>list</em>) – list of str, filenames</p></li>
<li><p><strong>path</strong> (<em>str</em>) – str, path to the files</p></li>
<li><p><strong>strands</strong> (<em>bool</em>) – bool, if True handle strand-specific data, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dict or tuple of dicts, BigWig objects</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or tuple</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dictBigWig</span><span class="p">([</span><span class="s1">&#39;sample1_fwd.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample1_rev.bw&#39;</span><span class="p">],</span> <span class="s1">&#39;/path/to/files&#39;</span><span class="p">)</span>
<span class="go">({&#39;sample1&#39;: &lt;pyBigWig object&gt;}, {&#39;sample1&#39;: &lt;pyBigWig object&gt;})</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.findPeaks">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">findPeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=20</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.findPeaks" title="Link to this definition"></a></dt>
<dd><p>Find local extrema using SciPy argrelextrema function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<em>pd.Series</em>) – Series data to localize peaks</p></li>
<li><p><strong>window</strong> (<em>int</em>) – To smooth data before peak-calling, default 1 (no smoothing)</p></li>
<li><p><strong>win_type</strong> (<em>str</em>) – Type of smoothing window, default “blackman”</p></li>
<li><p><strong>order</strong> (<em>int</em>) – Minimal spacing between peaks, argrelextrema order parameter, default 20</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of peaks</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">findPeaks</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.findTroughs">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">findTroughs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=20</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.findTroughs" title="Link to this definition"></a></dt>
<dd><p>Find local minima using SciPy argrelextrema function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<em>pd.Series</em>) – Series data to localize troughs</p></li>
<li><p><strong>window</strong> (<em>int</em>) – To smooth data before trough-calling, default 1 (no smoothing)</p></li>
<li><p><strong>win_type</strong> (<em>str</em>) – Type of smoothing window, default “blackman”</p></li>
<li><p><strong>order</strong> (<em>int</em>) – Minimal spacing between minima, argrelextrema order parameter, default 20</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of troughs</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">findTroughs</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.geneFromBigWig">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">geneFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwFWD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwREV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toStrip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.geneFromBigWig" title="Link to this definition"></a></dt>
<dd><p>Pull genome coverage from BigWig data for a given gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> (<em>str</em>) – str, name of the gene</p></li>
<li><p><strong>gtf</strong> (<em>pyCRAC.GTF2</em>) – pyCRAC.GTF2 object, GTF and TAB files loaded</p></li>
<li><p><strong>bwFWD</strong> (<em>dict</em>) – dict, pyBigWig objects for forward strand</p></li>
<li><p><strong>bwREV</strong> (<em>dict</em>) – dict, pyBigWig objects for reverse strand</p></li>
<li><p><strong>toStrip</strong> (<em>str</em>) – str, name to be stripped</p></li>
<li><p><strong>ranges</strong> (<em>int</em>) – int, flanks to be added for the gene, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame, genome coverage data</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">geneFromBigWig</span><span class="p">(</span><span class="s1">&#39;gene1&#39;</span><span class="p">,</span> <span class="n">gtf_object</span><span class="p">,</span> <span class="n">bwFWD_dict</span><span class="p">,</span> <span class="n">bwREV_dict</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.ntotal">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">ntotal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop=True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.ntotal" title="Link to this definition"></a></dt>
<dd><p>Normalize data in DataFrame to fraction of total column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame to be normalized</p></li>
<li><p><strong>drop</strong> (<em>bool</em>) – If True, drop ‘position’ and ‘nucleotide’ columns, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Normalized DataFrame</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ntotal</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.parseConcatFile">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">parseConcatFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reads'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RPM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.parseConcatFile" title="Link to this definition"></a></dt>
<dd><p>Parse concat file to extract gene data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – str, path of the concat file</p></li>
<li><p><strong>gtf</strong> (<em>pyCRAC.GTF2</em>) – pyCRAC.GTF2 object, GTF and TAB files loaded</p></li>
<li><p><strong>use</strong> (<em>str</em>) – str, column name to use [‘reads’, ‘substitutions’, ‘deletions’], default “reads”</p></li>
<li><p><strong>RPM</strong> (<em>bool</em>) – bool, if True use reads per million, default False</p></li>
<li><p><strong>ranges</strong> (<em>int</em>) – int, flanks to be added for the gene, default 1000</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dict, DataFrames using gene name as a key</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parseConcatFile</span><span class="p">(</span><span class="s1">&#39;/path/to/concat/file&#39;</span><span class="p">,</span> <span class="n">gtf_object</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function is deprecated and will be removed in future versions.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.preprocess">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">preprocess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">input_df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[],</span> <span class="pre">let_in=[''],</span> <span class="pre">let_out=['wont_find_this_string'],</span> <span class="pre">stats=False,</span> <span class="pre">smooth=True,</span> <span class="pre">window=10,</span> <span class="pre">win_type='blackman',</span> <span class="pre">pseudocounts_param=True,</span> <span class="pre">pseudo_value=0.01,</span> <span class="pre">ntotal_param=True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.preprocess" title="Link to this definition"></a></dt>
<dd><p>Combines methods.filterExp and expStats. Returns DataFrame with chosen experiments, optionally apply smoothing and stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the input data</p></li>
<li><p><strong>let_in</strong> (<em>list</em>) – List of words that characterize experiment, default [‘’]</p></li>
<li><p><strong>let_out</strong> (<em>list</em>) – List of words that disqualify experiments, default [‘wont_find_this_string’]</p></li>
<li><p><strong>stats</strong> (<em>bool</em>) – If True, return stats for all experiments, default False</p></li>
<li><p><strong>smooth</strong> (<em>bool</em>) – If True, apply smoothing window, default True</p></li>
<li><p><strong>window</strong> (<em>int</em>) – Smoothing window size, default 10</p></li>
<li><p><strong>win_type</strong> (<em>str</em>) – Type of smoothing window, default “blackman”</p></li>
<li><p><strong>pseudocounts_param</strong> (<em>bool</em>) – If True, add 0.01 pseudocounts, default True</p></li>
<li><p><strong>ntotal_param</strong> (<em>bool</em>) – If True, apply ntotal normalization, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with ‘mean’, ‘median’, ‘min’, ‘max’ and quartiles if more than 2 experiments</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">preprocess</span><span class="p">(</span><span class="n">input_df</span><span class="p">,</span> <span class="n">let_in</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;exp1&#39;</span><span class="p">],</span> <span class="n">let_out</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;exp2&#39;</span><span class="p">],</span> <span class="n">stats</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">,</span> <span class="n">pseudocounts_param</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ntotal_param</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.pseudocounts">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">pseudocounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value=0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop=True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.pseudocounts" title="Link to this definition"></a></dt>
<dd><p>Add pseudocounts to data to avoid zero values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame to which pseudocounts will be added</p></li>
<li><p><strong>value</strong> (<em>float</em>) – Value of the pseudocount to be added, default 0.01</p></li>
<li><p><strong>drop</strong> (<em>bool</em>) – If True, drop ‘position’ and ‘nucleotide’ columns, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with pseudocounts added</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pseudocounts</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.save_csv">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">save_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_ref=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasets=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.save_csv" title="Link to this definition"></a></dt>
<dd><p>Saves DataFrame to a CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_ref</strong> (<em>pd.DataFrame</em>) – DataFrame containing reference data with ‘position’ and ‘nucleotide’ columns</p></li>
<li><p><strong>datasets</strong> (<em>pd.DataFrame</em>) – DataFrame containing experimental data</p></li>
<li><p><strong>path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save the CSV file, default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with combined reference and experimental data</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">save_csv</span><span class="p">(</span><span class="n">data_ref</span><span class="p">,</span> <span class="n">datasets</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s1">&#39;/path/to/save.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.profiles.profileTools.stripBigWignames">
<span class="sig-prename descclassname"><span class="pre">trxtools.profiles.profileTools.</span></span><span class="sig-name descname"><span class="pre">stripBigWignames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.profiles.profileTools.stripBigWignames" title="Link to this definition"></a></dt>
<dd><p>Strip “_rev.bw” and “_fwd.bw” from file names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>files</strong> (<em>list</em>) – list of str, filenames</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str, unique names</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stripBigWignames</span><span class="p">([</span><span class="s1">&#39;sample1_fwd.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;sample1_rev.bw&#39;</span><span class="p">])</span>
<span class="go">[&#39;sample1&#39;]</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="trxtools-folding-module">
<h2>trxtools.folding module<a class="headerlink" href="#trxtools-folding-module" title="Link to this heading"></a></h2>
<section id="module-trxtools.folding.assays">
<span id="trxtools-folding-assays-submodule"></span><h3>trxtools.folding.assays submodule<a class="headerlink" href="#module-trxtools.folding.assays" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.bulkInputIDT">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">bulkInputIDT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.bulkInputIDT" title="Link to this definition"></a></dt>
<dd><p>Transforms dataframe with columns “template” and “non-template” and prepares table to be used as bulk input.
Discards oligos longer than 200 nt.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>DataFrame</em>) – Table with columns “template” and “non-template”</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table with sequences to order</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.extruded">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">extruded</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buried</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.extruded" title="Link to this definition"></a></dt>
<dd><p>Returns the extruded sequence by removing the buried nucleotides from the end</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – sequence</p></li>
<li><p><strong>buried</strong> (<em>int</em>) – length of sequence buried within RNAP</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>extruded sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.findPrimer">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">findPrimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.findPrimer" title="Link to this definition"></a></dt>
<dd><p>Find query sequence within given sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – sequence</p></li>
<li><p><strong>query</strong> (<em>str</em>) – query within searched sequences</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(start,stop) in python indexing</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">findPrimer</span><span class="p">(</span><span class="s2">&quot;ATCGATCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ATCG&quot;</span><span class="p">)</span>
<span class="go">(0, 4)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.nonTemplateDNA">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">nonTemplateDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.nonTemplateDNA" title="Link to this definition"></a></dt>
<dd><p>Returns the DNA sequence of the non-template strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – sequence</p></li>
<li><p><strong>primer</strong> (<em>str</em>) – primer sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DNA sequence of non-template strand</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.sequenceConstrain">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">sequenceConstrain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.sequenceConstrain" title="Link to this definition"></a></dt>
<dd><p>Returns sequence constrained with the 5prime RNA primer</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<em>str</em>) – sequence structure</p></li>
<li><p><strong>stall</strong> (<em>str</em>) – single letter representing the stall sequence</p></li>
<li><p><strong>RNAprimer</strong> (<em>str</em>) – RNA primer sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>sequence constraints for folding algorithm</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.stalled">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">stalled</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.stalled" title="Link to this definition"></a></dt>
<dd><p>Finds and returns stalled sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – sequence</p></li>
<li><p><strong>stall</strong> (<em>str</em>) – stall sequence, default “AAA”</p></li>
<li><p><strong>primer</strong> (<em>str</em>) – primer sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>stalled sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.structureFile">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">structureFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.structureFile" title="Link to this definition"></a></dt>
<dd><p>Saves structure file in current directory</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<em>str</em>) – sequence structure</p></li>
<li><p><strong>stall</strong> (<em>str</em>) – single letter representing the stall sequence</p></li>
<li><p><strong>RNAprimer</strong> (<em>str</em>) – RNA primer sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.templateDNA">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">templateDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.templateDNA" title="Link to this definition"></a></dt>
<dd><p>Returns the DNA sequence of the template strand. Takes reverse complement of RNA sequence and the 5’end DNA overhang as input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – sequence of RNA</p></li>
<li><p><strong>overhang5end</strong> (<em>str</em>) – sequence of the 5’end DNA overhang</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DNA sequence of template strand</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.testScaffold">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">testScaffold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNA_primer=''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.testScaffold" title="Link to this definition"></a></dt>
<dd><p>Prints scaffold to test before ordering</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – DataFrame with columns “template” and “non-template”</p></li>
<li><p><strong>overhang5end</strong> (<em>str</em>) – sequence of the 5’end DNA overhang</p></li>
<li><p><strong>RNA_primer</strong> (<em>str</em>) – RNA primer sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.assays.toOrder">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.assays.</span></span><span class="sig-name descname"><span class="pre">toOrder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buried='GUCUGUUUUGUGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stallFull='AAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">afterStall='TGATCGGTAC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end='TGA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNA_primer='AGGCCGAAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkInput=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lengthMax=200</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.assays.toOrder" title="Link to this definition"></a></dt>
<dd><p>Prepares sequences for ordering by applying folding functions to DNA sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DataFrame</em>) – DataFrame with columns “sequence” and “name”</p></li>
<li><p><strong>buried</strong> (<em>str</em>) – sequence buried within RNAP, default “GUCUGUUUUGUGG”</p></li>
<li><p><strong>stallFull</strong> (<em>str</em>) – stall sequence, default “AAA”</p></li>
<li><p><strong>afterStall</strong> (<em>str</em>) – sequence after stall, default “TGATCGGTAC”</p></li>
<li><p><strong>overhang5end</strong> (<em>str</em>) – sequence of the 5’end DNA overhang, default “TGA”</p></li>
<li><p><strong>RNA_primer</strong> (<em>str</em>) – RNA primer sequence, default “AGGCCGAAA”</p></li>
<li><p><strong>bulkInput</strong> (<em>bool</em>) – whether to prepare bulk input for ordering, default True</p></li>
<li><p><strong>test</strong> (<em>bool</em>) – whether to print scaffold for testing before ordering, default True</p></li>
<li><p><strong>lengthMax</strong> (<em>int</em>) – maximum length of sequences to order, default 200</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame of sequences to order</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-trxtools.folding.nascent">
<span id="trxtools-folding-nascent-submodule"></span><h3>trxtools.folding.nascent submodule<a class="headerlink" href="#module-trxtools.folding.nascent" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">Fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tempDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold.RNAfold">
<span class="sig-name descname"><span class="pre">RNAfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold.RNAfold" title="Link to this definition"></a></dt>
<dd><p>Calculates dG using RNAfold (ViennaRNA).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>{list</em><em>, </em><em>pd.Series</em><em>, </em><em>pd.DataFrame}</em>) – Input data, can be a list, Series, or DataFrame with “name” column.</p></li>
<li><p><strong>saveData</strong> (<em>bool</em>) – If True, saves the data, default is False.</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – Temperature for folding, default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with calculated dG values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold.RNAinvert">
<span class="sig-name descname"><span class="pre">RNAinvert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quick</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold.RNAinvert" title="Link to this definition"></a></dt>
<dd><p>Returns n sequences with a given structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<em>str</em>) – Secondary RNA structure.</p></li>
<li><p><strong>saveData</strong> (<em>bool</em>) – If True, saves the data, default is False.</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – Temperature for folding, default is None.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – Number of output sequences, default is 5.</p></li>
<li><p><strong>RNAprimer</strong> (<em>str</em>) – RNA primer sequence.</p></li>
<li><p><strong>stall</strong> (<em>str</em>) – Nucleotide sequence to avoid.</p></li>
<li><p><strong>quick</strong> (<em>bool</em>) – If False, uses -Fmp -f 0.01 params, default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with generated sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold.RNAinvertStall">
<span class="sig-name descname"><span class="pre">RNAinvertStall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'A'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold.RNAinvertStall" title="Link to this definition"></a></dt>
<dd><p>Returns sequences without the nucleotide that is present in stall.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<em>str</em>) – Secondary RNA structure.</p></li>
<li><p><strong>RNAprimer</strong> (<em>str</em>) – RNA primer sequence.</p></li>
<li><p><strong>stall</strong> (<em>str</em>) – Nucleotide sequence to avoid, default is “A”.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – Number of sequences to generate, default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with filtered sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold.UNAfold">
<span class="sig-name descname"><span class="pre">UNAfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold.UNAfold" title="Link to this definition"></a></dt>
<dd><p>Calculates dG using UNAfold.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>{list</em><em>, </em><em>pd.Series</em><em>, </em><em>pd.DataFrame}</em>) – Input data, can be a list, Series, or DataFrame with “name” column.</p></li>
<li><p><strong>saveData</strong> (<em>bool</em>) – If True, saves the data, default is False.</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – Temperature for folding, default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with calculated dG values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Fold.bashFolding">
<span class="sig-name descname"><span class="pre">bashFolding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RNA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Fold.bashFolding" title="Link to this definition"></a></dt>
<dd><p>Runs RNA folding using bash.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>method</strong> (<em>str</em>) – Folding method, “RNA” for ViennaRNA or “UNA” for UNAfold.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Hybrid">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">Hybrid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tempDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Hybrid" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Hybrid.RNAhybrid">
<span class="sig-name descname"><span class="pre">RNAhybrid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Hybrid.RNAhybrid" title="Link to this definition"></a></dt>
<dd><p>Calculates dG using hybrid-min.
This function calculates the hybridization energy (dG) between RNA sequences using the hybrid-min tool.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>{list</em><em>, </em><em>pd.Series</em><em>, </em><em>pd.DataFrame}</em>) – Input data, can be a list, Series, or DataFrame with “name” column.</p></li>
<li><p><strong>saveData</strong> (<em>bool</em>) – If True, saves the data, default is False.</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – Temperature for hybridization, default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with calculated dG values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hybrid</span> <span class="o">=</span> <span class="n">Hybrid</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AGCUAGUCA&quot;</span><span class="p">,</span> <span class="s2">&quot;CGAUCGUAG&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hybrid</span><span class="o">.</span><span class="n">RNAhybrid</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="trxtools.folding.nascent.Hybrid.bashHybrid">
<span class="sig-name descname"><span class="pre">bashHybrid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.Hybrid.bashHybrid" title="Link to this definition"></a></dt>
<dd><p>Runs hybrid-min using bash.
This function executes the hybrid-min command using bash to calculate the hybridization energy between two sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.extendingWindow">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">extendingWindow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.extendingWindow" title="Link to this definition"></a></dt>
<dd><p>Returns DataFrame of sequences of all possible lengths between minimum (m) and length of input sequence -1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – str, input sequence</p></li>
<li><p><strong>name</strong> (<em>str</em>) – name of the sequence, default “name”</p></li>
<li><p><strong>strand</strong> (<em>str</em>) – strand type {“plus”, “minus”}, default “plus”</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – temperature, default 30</p></li>
<li><p><strong>m</strong> (<em>int</em>) – minimum length of sequence, default 7</p></li>
<li><p><strong>toAdd</strong> (<em>int</em>) – additional length to add, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame of sequences with names according to nascent.slidingWindow convention</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.foldNascentElem">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">foldNascentElem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.foldNascentElem" title="Link to this definition"></a></dt>
<dd><p>Fold the very 3’ of nascent elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pd.DataFrame</em>) – DataFrame containing folded sequences.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with additional columns for nascent element folding energies.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.handleInput">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">handleInput</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keepNames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.handleInput" title="Link to this definition"></a></dt>
<dd><p>Input data with columns: “seq” or “sequence” and “name” (optional).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>{str</em><em>, </em><em>list</em><em>, </em><em>pd.Series</em><em>, </em><em>pd.DataFrame}</em>) – input data {str, list, Series, DataFrame}</p></li>
<li><p><strong>keepNames</strong> (<em>bool</em>) – if True use given names, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame where index become name of sequence to fold</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.join2d">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">join2d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='format'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.join2d" title="Link to this definition"></a></dt>
<dd><p>Joins 2D data from RNA folding results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing RNA folding results.</p></li>
<li><p><strong>use</strong> (<em>str</em>) – Column to use for joining, default is ‘format’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of DataFrames, one for each gene/chromosome.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.merge2d">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">merge2d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.merge2d" title="Link to this definition"></a></dt>
<dd><p>Merges 2D data from RNA folding results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing RNA folding results.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with statistical description of the merged data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.name2index">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">name2index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">object)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.name2index" title="Link to this definition"></a></dt>
<dd><p>Extracts position from sequence name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>s1</strong> (<em>pd.Series</em>) – Series with names from prepareNascent function</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.nascentElems">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentElems</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.nascentElems" title="Link to this definition"></a></dt>
<dd><p>Describe elements of secondary structure: stems, multistems.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna RNA secondary structure notation.</p></li>
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing the sequence of the last nascent element and its distance from the end.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.nascentElemsDataFrame">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentElemsDataFrame</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.nascentElemsDataFrame" title="Link to this definition"></a></dt>
<dd><p>Apply nascentElems to DataFrame of folded sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pd.DataFrame</em>) – DataFrame containing folded sequences with ‘vienna’ and ‘sequence’ columns.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with additional columns for nascent element sequences and distances.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.nascentFolding">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentFolding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.nascentFolding" title="Link to this definition"></a></dt>
<dd><p>Combines folding function: fold RNA, locate last nascent element and calculate dG of it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence.</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – Folding temperature, default is 30.</p></li>
<li><p><strong>window</strong> (<em>int</em>) – Window size for folding, default is 100.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with folding results and nascent element information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.parseFoldingName">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">parseFoldingName</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.parseFoldingName" title="Link to this definition"></a></dt>
<dd><p>Parse folding names into separate columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing folding names.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with parsed columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.prepareNascent">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">prepareNascent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.prepareNascent" title="Link to this definition"></a></dt>
<dd><p>Divide long transcript into short sequences. Combines output of extendingWindow and slidingWindow.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – input sequence</p></li>
<li><p><strong>name</strong> (<em>str</em>) – name of the sequence, default “name”</p></li>
<li><p><strong>strand</strong> (<em>str</em>) – strand type {“plus”, “minus”}, default “plus”</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – temperature, default 30</p></li>
<li><p><strong>window</strong> (<em>int</em>) – window size, default 100</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with sequences</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.nascent.slidingWindow">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.nascent.</span></span><span class="sig-name descname"><span class="pre">slidingWindow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.nascent.slidingWindow" title="Link to this definition"></a></dt>
<dd><p>Slices sequence using sliding window.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – input sequence</p></li>
<li><p><strong>name</strong> (<em>str</em>) – name of the sequence, default “name”</p></li>
<li><p><strong>strand</strong> (<em>str</em>) – strand type {“plus”, “both”, “minus”}, default “plus”</p></li>
<li><p><strong>temp</strong> (<em>int</em>) – temperature, default 30</p></li>
<li><p><strong>window</strong> (<em>int</em>) – window size, default 100</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with sliding windows</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-trxtools.folding.secondary">
<span id="trxtools-folding-secondary-submodule"></span><h3>trxtools.folding.secondary submodule<a class="headerlink" href="#module-trxtools.folding.secondary" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.Lstem">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">Lstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.Lstem" title="Link to this definition"></a></dt>
<dd><p>Returns a list of positions where “(” is found using coordinates {1,inf}.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of positions with “(”</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p>Example:
&gt;&gt;&gt; Lstem(“((..))”)
[1, 2]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.Rstem">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">Rstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.Rstem" title="Link to this definition"></a></dt>
<dd><p>Returns a list of positions where “)” is found using coordinates {1,inf}.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of positions with “)”</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p>Example:
&gt;&gt;&gt; Rstem(“((..))”)
[5, 6]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.checkVienna">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">checkVienna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.checkVienna" title="Link to this definition"></a></dt>
<dd><p>Validates the integrity of a Vienna file by checking the sequence and structure lengths and the balance of parentheses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence</p></li>
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the Vienna file is valid, False otherwise</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkVienna</span><span class="p">(</span><span class="s2">&quot;GCAU&quot;</span><span class="p">,</span> <span class="s2">&quot;((..))&quot;</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.loopStems">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">loopStems</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loopsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testPrint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.loopStems" title="Link to this definition"></a></dt>
<dd><p>Returns positions of stems of single hairpins and multiloop stems using coordinates {1:inf}.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p></li>
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence (optional)</p></li>
<li><p><strong>loopsList</strong> (<em>list</em>) – List of loop positions (optional)</p></li>
<li><p><strong>testPrint</strong> (<em>bool</em>) – Boolean to print test output, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of stem positions and list of multistem positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loopStems</span><span class="p">(</span><span class="s2">&quot;((..))&quot;</span><span class="p">,</span> <span class="s2">&quot;GCAU&quot;</span><span class="p">)</span>
<span class="go">([(1, 6)], [])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.loops">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">loops</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.loops" title="Link to this definition"></a></dt>
<dd><p>Returns the first positions outside the loop, e.g., “.((….)).” returns [(3,8)].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of tuples with loop positions</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p>Example:
&gt;&gt;&gt; loops(“.((….)).”)
[(3, 8)]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.substructures">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">substructures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.substructures" title="Link to this definition"></a></dt>
<dd><p>Lists sub-structures of the given structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p></li>
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series of sub-structures</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">substructures</span><span class="p">(</span><span class="s2">&quot;((..))&quot;</span><span class="p">,</span> <span class="s2">&quot;GCAU&quot;</span><span class="p">)</span>
<span class="go">stem_1    GCAU</span>
<span class="go">dtype: object</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.test">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stems</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multistems</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linkers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.test" title="Link to this definition"></a></dt>
<dd><p>Prints the Vienna structure with given features marked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p></li>
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence (optional)</p></li>
<li><p><strong>loops</strong> (<em>list</em>) – List of tuples with loop positions (optional)</p></li>
<li><p><strong>stems</strong> (<em>list</em>) – List of tuples with stem positions (optional)</p></li>
<li><p><strong>multistems</strong> (<em>list</em>) – List of multistem positions (optional)</p></li>
<li><p><strong>linkers</strong> (<em>list</em>) – List of linker positions (optional)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p>Example:
&gt;&gt;&gt; test(“((..))”, “GCAU”, loops=[(3, 4)])
1         
((..))
GCAU
__O_O</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.folding.secondary.vienna2format">
<span class="sig-prename descclassname"><span class="pre">trxtools.folding.secondary.</span></span><span class="sig-name descname"><span class="pre">vienna2format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loopsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stemsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multistemsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testPrint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.folding.secondary.vienna2format" title="Link to this definition"></a></dt>
<dd><p>Converts Vienna format to a custom format with letters: O - loop, S - stem, M - multiloop stem, and L - linker.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> (<em>str</em>) – Vienna format secondary structure</p></li>
<li><p><strong>sequence</strong> (<em>str</em>) – RNA sequence (optional)</p></li>
<li><p><strong>loopsList</strong> (<em>list</em>) – List of loop positions (optional)</p></li>
<li><p><strong>stemsList</strong> (<em>list</em>) – List of stem positions (optional)</p></li>
<li><p><strong>multistemsList</strong> (<em>list</em>) – List of multistem positions (optional)</p></li>
<li><p><strong>testPrint</strong> (<em>bool</em>) – Boolean to print test output, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Custom format string</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vienna2format</span><span class="p">(</span><span class="s2">&quot;((..))&quot;</span><span class="p">,</span> <span class="s2">&quot;GCAU&quot;</span><span class="p">)</span>
<span class="go">&#39;SSLLSS&#39;</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="trxtools-tables-module">
<h2>trxtools.tables module<a class="headerlink" href="#trxtools-tables-module" title="Link to this heading"></a></h2>
<section id="module-trxtools.tables.goEnrichment">
<span id="trxtools-tables-goenrichment-submodule"></span><h3>trxtools.tables.goEnrichment submodule<a class="headerlink" href="#module-trxtools.tables.goEnrichment" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.tables.goEnrichment.get_enrichment">
<span class="sig-prename descclassname"><span class="pre">trxtools.tables.goEnrichment.</span></span><span class="sig-name descname"><span class="pre">get_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biological_process'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_reference_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FISHER'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FDR'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.tables.goEnrichment.get_enrichment" title="Link to this definition"></a></dt>
<dd><p>Run a GO term enrichment test using PANTHER API</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_genes</strong> (<em>list</em>) – List of sequence identifiers of queried genes (e.g. transcript ids, gene ids)</p></li>
<li><p><strong>organism</strong> (<em>str</em>) – Taxid of query species (e.g. “9606” for H. sapiens)</p></li>
<li><p><strong>go_dataset</strong> (<em>str</em>) – Which annotation dataset to query, “biological_process” or “molecular_function”</p></li>
<li><p><strong>use_reference_set</strong> (<em>bool</em>) – Use a custom set of rerence (background) genes? Default False. If True, ref_genes and ref_species need to be specifed.</p></li>
<li><p><strong>ref_genes</strong> (<em>list</em>) – <em>optional</em> list of reference genes. Specifying None (default) will use the whole genome of species specified in organism. When passing a list, ref_organism taxid must also be provided.</p></li>
<li><p><strong>ref_species</strong> (<em>str</em>) – Taxid of reference species, required when ref_genes is not None</p></li>
<li><p><strong>test_type</strong> (<em>str</em>) – Which tatistical test to use. Available: “FISHER” (default), “BINOMIAL”</p></li>
<li><p><strong>correction</strong> (<em>str</em>) – Which multiple testing correction method to use. Available: “FDR” (default), “BONFERRONI”, “NONE”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Unfiltered DataFrame of results.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame:</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="module-trxtools.methods">
<span id="trxtools-methods-module"></span><h2>trxtools.methods module<a class="headerlink" href="#module-trxtools.methods" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.DNA_stretch_positions">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">DNA_stretch_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_len</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_len</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sequence'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.DNA_stretch_positions" title="Link to this definition"></a></dt>
<dd><p>Finds all character stretches of given length in a string (i.e. DNA sequence) and reports their positions. Wrapper for DNA_string_positions()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Input dataframe</p></li>
<li><p><strong>char</strong> (<em>str</em>) – Query character (base)</p></li>
<li><p><strong>min_len</strong> (<em>int</em>) – minimum length of stretch to look for</p></li>
<li><p><strong>min_len</strong> – maximum length of stretch to look for</p></li>
<li><p><strong>name_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequence names, defaults to ‘name’</p></li>
<li><p><strong>seq_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequences, defaults to ‘sequence’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe in long format containing positions of found string</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.DNA_string_positions">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">DNA_string_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sequence'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.DNA_string_positions" title="Link to this definition"></a></dt>
<dd><p>Finds all occurences of a given substring in a string (i.e. DNA sequence) and reports their positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Input dataframe</p></li>
<li><p><strong>string</strong> (<em>str</em>) – Query string</p></li>
<li><p><strong>name_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequence names, defaults to ‘name’</p></li>
<li><p><strong>seq_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequences, defaults to ‘sequence’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe in long format containing positionsof found string</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.addCluster">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">addCluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n=10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.addCluster" title="Link to this definition"></a></dt>
<dd><p>Assigns n clusters to the data using KMeans algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame containing the data to be clustered</p></li>
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of clusters to form, defaults to 10</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with an additional ‘cluster’ column indicating cluster assignment</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">addCluster</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">   x  y  cluster</span>
<span class="go">0  1  5        1</span>
<span class="go">1  2  4        1</span>
<span class="go">2  3  3        0</span>
<span class="go">3  4  2        0</span>
<span class="go">4  5  1        0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.bashCommand">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">bashCommand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bashCommand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.bashCommand" title="Link to this definition"></a></dt>
<dd><p>Run command in bash using subprocess.call()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bashCommand</strong> – str with command</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.bed2len">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">bed2len</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.bed2len" title="Link to this definition"></a></dt>
<dd><p>Convert a bed DataFrame to a length DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bed</strong> (<em>pandas.DataFrame</em>) – pandas DataFrame containing bed data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas DataFrame with length data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chr&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span> <span class="s1">&#39;stop&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">150</span><span class="p">,</span> <span class="mi">250</span><span class="p">],</span> <span class="s1">&#39;region&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;region1&#39;</span><span class="p">,</span> <span class="s1">&#39;region2&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed2len</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
<span class="go">region</span>
<span class="go">region1    50</span>
<span class="go">region2    50</span>
<span class="go">dtype: int64</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.calGC">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">calGC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[],</span> <span class="pre">calFor=['G',</span> <span class="pre">'C']</span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.calGC" title="Link to this definition"></a></dt>
<dd><p>Returns GC content in a given string - uses [‘nucleotide’] column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataset</strong> (<em>pandas.DataFrame</em>) – DataFrame() with “nucleotide” column</p></li>
<li><p><strong>calFor</strong> (<em>list</em>) – list of nucleotides to calculate GC content</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fraction of GC content, float</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.cleanNames">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">cleanNames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.cleanNames" title="Link to this definition"></a></dt>
<dd><p>Cleans the names in the given data by removing specified strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>dict</em><em> or </em><em>pandas.DataFrame</em>) – The data to be cleaned. It can be either a dictionary or a pandas DataFrame.</p></li>
<li><p><strong>strings</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of strings to be removed from the names. Defaults to an empty list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The cleaned data with names modified according to the specified strings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.define_experiments">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">define_experiments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">paths_in</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">whole_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_hittable_reads.txt'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.define_experiments" title="Link to this definition"></a></dt>
<dd><p>Parse file names and extract experiment name from them</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>paths_in</strong> (<em>list</em>) – List of file paths</p></li>
<li><p><strong>whole_name</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use the whole file name as the experiment name, defaults to False</p></li>
<li><p><strong>strip</strong> (<em>str</em><em>, </em><em>optional</em>) – String to strip from the file name, defaults to ‘_hittable_reads.txt’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of experiment names and list of paths</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.enriched">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">enriched</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.enriched" title="Link to this definition"></a></dt>
<dd><p>Filter DataFrame for enriched genes based on adjusted p-value and log2 fold change.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame containing gene expression data</p></li>
<li><p><strong>padj</strong> (<em>float</em><em>, </em><em>optional</em>) – Adjusted p-value threshold, defaults to 0.05</p></li>
<li><p><strong>fc</strong> (<em>float</em><em>, </em><em>optional</em>) – Log2 fold change threshold, defaults to 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Filtered DataFrame with enriched genes</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.expNameParser">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">expNameParser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'b_d_e_p'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.expNameParser" title="Link to this definition"></a></dt>
<dd><p>Function handles experiment name; recognizes AB123456 as experiment date; BY4741 or HTP or given string as bait protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Experiment name</p></li>
<li><p><strong>additional_tags</strong> (<em>list</em>) – List of additional tags</p></li>
<li><p><strong>order</strong> (<em>str</em><em>, </em><em>optional</em>) – Order of elements in the output, defaults to ‘b_d_e_p’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Reordered experiment name</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.expStats">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">expStats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.expStats" title="Link to this definition"></a></dt>
<dd><p>Returns DataFrame with ‘mean’, ‘median’, ‘min’, ‘max’ and quartiles if more than 2 experiments</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_df</strong> (<em>pandas.DataFrame</em>) – DataFrame containing the input data</p></li>
<li><p><strong>smooth</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to apply smoothing window, defaults to True</p></li>
<li><p><strong>window</strong> (<em>int</em><em>, </em><em>optional</em>) – Smoothing window size, defaults to 10</p></li>
<li><p><strong>win_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Type of smoothing window, defaults to ‘blackman’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with calculated statistics</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.filterExp">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">filterExp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datasets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">let_in</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">let_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['wont_find_this_string']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.filterExp" title="Link to this definition"></a></dt>
<dd><p>Returns object with filtered columns/keys.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>datasets</strong> – DataFrame or dict with exp name as a key</p></li>
<li><p><strong>let_in</strong> (<em>list</em>) – List of elements of name to filter in</p></li>
<li><p><strong>let_out</strong> (<em>list</em>) – List of elements of name to filter out</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, prints the filtered columns/keys, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame or dict with filtered columns/keys</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame or dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.find_pol3_terminators">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">find_pol3_terminators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sequence'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.find_pol3_terminators" title="Link to this definition"></a></dt>
<dd><p>Finds all Pol3 terminators (T stretches) of given length in a string (i.e. DNA sequence) and reports their positions. Wrapper for DNA_string_positions()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Input dataframe</p></li>
<li><p><strong>min_T</strong> (<em>int</em>) – minimum length of T stretch to look for</p></li>
<li><p><strong>min_T</strong> – maximum length of T stretch to look for</p></li>
<li><p><strong>region_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequence names, defaults to ‘region’</p></li>
<li><p><strong>seq_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing sequences, defaults to ‘sequence’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe in long format containing positionsof found string</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.groupCRACsamples">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">groupCRACsamples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='protein'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toDrop=[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.groupCRACsamples" title="Link to this definition"></a></dt>
<dd><p>Parse CRAC names and annotates them using one of the following features:
[‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’, ‘sample’, ‘sampleRep’]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame with CRAC names as index</p></li>
<li><p><strong>use</strong> (<em>str</em><em>, </em><em>optional</em>) – Feature to annotate, choose from [‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’, ‘sample’, ‘sampleRep’], defaults to ‘protein’</p></li>
<li><p><strong>toDrop</strong> (<em>list</em><em>, </em><em>optional</em>) – List of words in CRAC name that will qualify the sample for rejection, defaults to []</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with added column [‘group’]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.indexOrder">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">indexOrder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_tags=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output='root'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order='b_d_e_p'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.indexOrder" title="Link to this definition"></a></dt>
<dd><p>Apply expNameParser to whole DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame where names of columns are names of experiments</p></li>
<li><p><strong>additional_tags</strong> (<em>list</em>) – List of additional tags to consider in expNameParser</p></li>
<li><p><strong>output</strong> (<em>str</em>) – Not used in the function, kept for compatibility</p></li>
<li><p><strong>order</strong> (<em>str</em><em>, </em><em>optional</em>) – Order of elements in the output, defaults to ‘b_d_e_p’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with new names</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.is_inside">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">is_inside</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inner_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inner_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outer_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outer_end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.is_inside" title="Link to this definition"></a></dt>
<dd><p>Check if one region is inside another string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>inner_start</strong> (<em>int</em>) – Start position of the inner region</p></li>
<li><p><strong>inner_end</strong> (<em>int</em>) – End position of the inner region</p></li>
<li><p><strong>outer_start</strong> (<em>int</em>) – Start position of the outer region</p></li>
<li><p><strong>outer_end</strong> (<em>int</em>) – End position of the outer region</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the inner region is inside the outer region, False otherwise</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_inside</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_inside</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.letterContent">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">letterContent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'A'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.letterContent" title="Link to this definition"></a></dt>
<dd><p>Calculates the content of a given letter in a sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – Sequence to calculate the content of the given letter in</p></li>
<li><p><strong>letter</strong> (<em>str</em>) – letter to calculate the content of in the sequence, defaults to “A”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Content of the given letter in the sequence, round to two decimal places</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">letterContent</span><span class="p">(</span><span class="s2">&quot;ATCG&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">0.25</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.listPaths">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">listPaths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.listPaths" title="Link to this definition"></a></dt>
<dd><p>List all file paths in a folder with a specific suffix and optional name element.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>folder</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder to list files from, defaults to “.”</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix of the files to list, defaults to an empty string</p></li>
<li><p><strong>nameElem</strong> (<em>str</em><em>, </em><em>optional</em>) – Optional element that must be in the file name, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of file paths that match the criteria</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">listPaths</span><span class="p">(</span><span class="n">folder</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.txt&quot;</span><span class="p">,</span> <span class="n">nameElem</span><span class="o">=</span><span class="s2">&quot;file&quot;</span><span class="p">)</span>
<span class="go">[&#39;file1.txt&#39;, &#39;file2.txt&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.loadGTF">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">loadGTF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gtf_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.loadGTF" title="Link to this definition"></a></dt>
<dd><p>Load GTF file into a DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gtf_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the GTF file, defaults to “”</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing GTF data</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.nested_region_cleanup">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">nested_region_cleanup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'end'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.nested_region_cleanup" title="Link to this definition"></a></dt>
<dd><p>Helper function to remove shorter regions nested in longer ones</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – input dataframe</p></li>
<li><p><strong>start_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing region start positions, defaults to ‘start’</p></li>
<li><p><strong>end_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing region end positions, defaults to ‘end’</p></li>
<li><p><strong>length_col</strong> (<em>str</em><em>, </em><em>optional</em>) – name of column containing region lengths, defaults to ‘length’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dataframe with nested regions removed</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.normalize">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts=0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CPM=True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.normalize" title="Link to this definition"></a></dt>
<dd><p>Normalize the DataFrame by adding pseudocounts, dividing by the sum, and optionally applying log2 transformation and CPM scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame to be normalized</p></li>
<li><p><strong>log2</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to apply log2 transformation, defaults to False</p></li>
<li><p><strong>pseudocounts</strong> (<em>float</em><em>, </em><em>optional</em>) – Value to add to each element to avoid log of zero, defaults to 0.1</p></li>
<li><p><strong>CPM</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the data to counts per million, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Normalized DataFrame</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.parseCRACname">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">parseCRACname</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=&lt;class</span> <span class="pre">'pandas.core.series.Series'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.parseCRACname" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>Parse CRAC name into [‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’] using this order.</dt><dd><p>“_” is used to split the name</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>s1</strong> (<em>pandas.Series</em>) – Series containing CRAC names</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with parsed CRAC names</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.quantileCategory">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">quantileCategory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q=4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.quantileCategory" title="Link to this definition"></a></dt>
<dd><p>Quantile-based discretization function based on pandas.qcut function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<em>pandas.Series</em>) – Series to be discretized</p></li>
<li><p><strong>q</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of quantiles, defaults to 4</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with quantile categories</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">quantileCategory</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">]),</span> <span class="n">q</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">0       0</span>
<span class="go">1       0</span>
<span class="go">2       1</span>
<span class="go">3       1</span>
<span class="go">4       1</span>
<span class="go">dtype: int64</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.randomDNAall">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">randomDNAall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CGTA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.randomDNAall" title="Link to this definition"></a></dt>
<dd><p>Generates all possible random sequences of a given length</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<em>int</em>) – Length of the sequence</p></li>
<li><p><strong>letters</strong> (<em>str</em>) – str with letters that will be used</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of all possible random sequences of the given length</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">randomDNAall</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="s2">&quot;CGTA&quot;</span><span class="p">)</span>
<span class="go">[&#39;CC&#39;, &#39;CG&#39;, &#39;CT&#39;, &#39;CA&#39;, &#39;GC&#39;, &#39;GG&#39;, &#39;GT&#39;, &#39;GA&#39;, &#39;TC&#39;, &#39;TG&#39;, &#39;TT&#39;, &#39;TA&#39;, &#39;AC&#39;, &#39;AG&#39;, &#39;AT&#39;, &#39;AA&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.randomDNAsingle">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">randomDNAsingle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CGTA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.randomDNAsingle" title="Link to this definition"></a></dt>
<dd><p>Random generator of nucleotide sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<em>int</em>) – Length of the sequence</p></li>
<li><p><strong>letters</strong> (<em>str</em>) – str with letters that will be used</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Random sequence of the given length</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">randomDNAsingle</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="s2">&quot;CGTA&quot;</span><span class="p">)</span>
<span class="go">&#39;CGTACGTAAC&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.readSalmon">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">readSalmon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NumReads'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.readSalmon" title="Link to this definition"></a></dt>
<dd><p>Reads multiple Salmon quant.sf files to one DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nameElem</strong> (<em>str</em>) – Element to load, present in all files</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to directory with files</p></li>
<li><p><strong>toLoad</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional parameter for filtering, defaults to nameElem</p></li>
<li><p><strong>toClear</strong> (<em>list</em>) – List of strings to be removed from file names</p></li>
<li><p><strong>toAdd</strong> (<em>str</em>) – String to be added to file names</p></li>
<li><p><strong>column</strong> (<em>str</em><em>, </em><em>optional</em>) – Column to extract from quant.sf file, defaults to ‘NumReads’</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame</em><em>, </em><em>optional</em>) – DataFrame to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – Boolean, allows for overwriting during appending, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with all files, where columns are values from ‘Name’ column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_DEseq">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_DEseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_DEseq" title="Link to this definition"></a></dt>
<dd><p>WARNING: Not tested properly. May not work as expected.
Read DESeq2 output file and add gene names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>p</strong> (<em>str</em>) – Path to the DESeq2 output file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with DESeq2 results and gene names</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_HTSeq_output">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_HTSeq_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'classes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_HTSeq_output" title="Link to this definition"></a></dt>
<dd><p>Reads multiple HTSeq tab files to one DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Path to directory with files</p></li>
<li><p><strong>toClear</strong> (<em>list</em>) – List of strings to be removed from file names</p></li>
<li><p><strong>toAdd</strong> (<em>str</em>) – String to be added to file names</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame</em><em>, </em><em>optional</em>) – DataFrame to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – Boolean, allows for overwriting during appending, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with all files, where columns are values from ‘name’ column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_STARstats">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_STARstats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_STARstats" title="Link to this definition"></a></dt>
<dd><p>Reads multiple STAR Log final output files to one DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Path to directory with files</p></li>
<li><p><strong>toClear</strong> (<em>list</em>) – List of strings to be removed from file names</p></li>
<li><p><strong>toAdd</strong> (<em>str</em>) – String to be added to file names</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame</em><em>, </em><em>optional</em>) – DataFrame to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – Boolean, allows for overwriting during appending, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with all files, where columns are values from ‘name’ column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_featureCount">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_featureCount</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_featureCount" title="Link to this definition"></a></dt>
<dd><p>Read featureCount files with common first column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nameElem</strong> (<em>str</em>) – Element to load, present in all files</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to directory with files</p></li>
<li><p><strong>toLoad</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional parameter for filtering, defaults to nameElem</p></li>
<li><p><strong>toClear</strong> (<em>list</em>) – List of strings to be removed from file names</p></li>
<li><p><strong>toAdd</strong> (<em>str</em>) – String to be added to file names</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame</em><em>, </em><em>optional</em>) – DataFrame to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – Boolean, allows for overwriting during appending, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with all files, where columns are values from ‘name’ column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_list">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_list" title="Link to this definition"></a></dt>
<dd><p>Read list from file. Each row becomes item in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filepath</strong> (<em>str</em>) – Path to the file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of items from the file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">read_list</span><span class="p">(</span><span class="s2">&quot;file.txt&quot;</span><span class="p">)</span>
<span class="go">[&#39;item1&#39;, &#39;item2&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.read_tabFile">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">read_tabFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.read_tabFile" title="Link to this definition"></a></dt>
<dd><p>Read tab files with common first column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nameElem</strong> (<em>str</em>) – part of a filename which is present in all files</p></li>
<li><p><strong>path</strong> (<em>str</em>) – path to directory with files</p></li>
<li><p><strong>toLoad</strong> (<em>str</em>) – to be present in file name</p></li>
<li><p><strong>toClear</strong> (<em>str</em>) – part of filename to be removed from file name</p></li>
<li><p><strong>toAdd</strong> (<em>str</em>) – string to be added to file name</p></li>
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – DataFrame, to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – boolean, allows for overwriting during appending, default = False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dataframe with all files, where columns are values from ‘name’ column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.reverse_complement">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Reverse complement of DNA or RNA sequence. Identifies RNA by presence of ‘U’ in the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – DNA or RNA sequence to reverse complement</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Reverse complement of the input sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement</span><span class="p">(</span><span class="s2">&quot;AUCG&quot;</span><span class="p">)</span>
<span class="go">&#39;CGAU&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.reverse_complement_DNA">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement_DNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.reverse_complement_DNA" title="Link to this definition"></a></dt>
<dd><p>Reverse complement of DNA sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – DNA sequence to reverse complement</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Reverse complement of the input sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement_DNA</span><span class="p">(</span><span class="s2">&quot;ATCG&quot;</span><span class="p">)</span>
<span class="go">&#39;CGAT&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.reverse_complement_RNA">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement_RNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.reverse_complement_RNA" title="Link to this definition"></a></dt>
<dd><p>Reverse complement of RNA sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – RNA sequence to reverse complement</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Reverse complement of the input sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement_RNA</span><span class="p">(</span><span class="s2">&quot;AUCG&quot;</span><span class="p">)</span>
<span class="go">&#39;CGAU&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.rollingGC">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">rollingGC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s=&lt;class</span> <span class="pre">'pandas.core.series.Series'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.rollingGC" title="Link to this definition"></a></dt>
<dd><p>Calculates GC from sequence, uses ‘boxcar’ window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s</strong> (<em>pandas.Series</em>) – Series containing sequence</p></li>
<li><p><strong>window</strong> (<em>int</em>) – window size for GC calculation</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Series with GC calculated, center=False</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.runPCA">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">runPCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components=2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.runPCA" title="Link to this definition"></a></dt>
<dd><p>Run PCA analysis and re-assigns column names and index names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pandas.DataFrame</em>) – DataFrame containing the input data</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of principal components to compute, defaults to 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple consisting of DataFrame with PCA results and a list of explained variance ratios</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Example<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">runPCA</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">]]),</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">(          PC1       PC2</span>
<span class="go">0 -4.242641  0.000000</span>
<span class="go">1  0.000000  0.000000</span>
<span class="go">2  4.242641  0.000000, [100.0, 0.0])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.timestamp">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">timestamp</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.timestamp" title="Link to this definition"></a></dt>
<dd><p>Returns current timestamp as a string</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>timestamp in a format ‘YYYYMMDD_HHMMSS’</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.methods.timestampRandomInt">
<span class="sig-prename descclassname"><span class="pre">trxtools.methods.</span></span><span class="sig-name descname"><span class="pre">timestampRandomInt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.methods.timestampRandomInt" title="Link to this definition"></a></dt>
<dd><p>Returns current timestamp with a random integer as a string</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>timestamp in a format ‘YYYYMMDD_HHMMSS_RANDOMINT’</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-trxtools.plotting">
<span id="trxtools-plotting-module"></span><h2>trxtools.plotting module<a class="headerlink" href="#module-trxtools.plotting" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.GOterm">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">GOterm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'term.label'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalizer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pValue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.GOterm" title="Link to this definition"></a></dt>
<dd><p>Generate a bar plot for GO terms based on significance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input DataFrame containing GO term data.</p></li>
<li><p><strong>x</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name for the x-axis values (e.g., FDR). Defaults to ‘fdr’.</p></li>
<li><p><strong>y</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name for the y-axis values (e.g., term label). Defaults to ‘term.label’.</p></li>
<li><p><strong>normalizer</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name used for normalization (e.g., p-value). Defaults to ‘pValue’.</p></li>
<li><p><strong>cutoff</strong> (<em>float</em><em>, </em><em>optional</em>) – The cutoff value for significance. Defaults to 0.05.</p></li>
<li><p><strong>count</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of top terms to display. Defaults to 20.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The figure size. Defaults to (4, 4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the saved figure. Defaults to 300.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the plot. Defaults to None.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the figure. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">GOterm</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fdr&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;term.label&#39;</span><span class="p">,</span> <span class="n">normalizer</span><span class="o">=</span><span class="s1">&#39;pValue&#39;</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;GO Term Plot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.PCA">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">PCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PClimit=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">7)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PCval=[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.PCA" title="Link to this definition"></a></dt>
<dd><p>Plot PCA plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – DataFrame containing PCA results.</p></li>
<li><p><strong>names</strong> (<em>list</em><em>, </em><em>optional</em>) – List of names to annotate.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot.</p></li>
<li><p><strong>PClimit</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of principal components to plot. Defaults to 1.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 7).</p></li>
<li><p><strong>PCval</strong> (<em>list</em><em>, </em><em>optional</em>) – List of explained variance for each principal component. Defaults to an empty list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">PCA</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Sample1&#39;</span><span class="p">,</span> <span class="s1">&#39;Sample2&#39;</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;PCA Plot&#39;</span><span class="p">,</span> <span class="n">PClimit</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.boxplot1">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">boxplot1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(7,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Spectral'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.boxplot1" title="Link to this definition"></a></dt>
<dd><p>Generate a box plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>pd.DataFrame</em><em> or </em><em>pd.Series</em>) – Data to plot.</p></li>
<li><p><strong>labels</strong> (<em>list</em><em>, </em><em>optional</em>) – Labels for the data. Defaults to None.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot. Defaults to an empty string.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 6).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure. Defaults to 150.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use a logarithmic scale. Defaults to False.</p></li>
<li><p><strong>lim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to None.</p></li>
<li><p><strong>name_cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the colormap to use. Defaults to ‘Spectral’.</p></li>
<li><p><strong>vert</strong> (<em>int</em><em>, </em><em>optional</em>) – Whether to plot the boxes vertically. Defaults to 1.</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color of the boxes. Defaults to None.</p></li>
<li><p><strong>grid</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display a grid. Defaults to False.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">boxplot1</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Box Plot&#39;</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.clusterClusterMap">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">clusterClusterMap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.clusterClusterMap" title="Link to this definition"></a></dt>
<dd><p>Generate a clustermap for clusters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing cluster data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">clusterClusterMap</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.cumulativeDifference">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">cumulativeDifference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df2=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_pos=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_between=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equal_weight=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color1='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.cumulativeDifference" title="Link to this definition"></a></dt>
<dd><p>Plot cumulative differences for peaks metaplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to CSV file or DataFrame containing reference data.</p></li>
<li><p><strong>df2</strong> (<em>pd.DataFrame</em>) – DataFrame containing the second dataset.</p></li>
<li><p><strong>local_pos</strong> (<em>list</em>) – List of positions of features (peaks/troughs).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
<li><p><strong>fill_between</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to fill the area between the plot and the x-axis. Defaults to True.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot. Defaults to an empty string.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for filtering local positions. Defaults to None.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for filtering local positions. Defaults to None.</p></li>
<li><p><strong>window</strong> (<em>int</em><em>, </em><em>optional</em>) – Window size around each feature. Defaults to 50.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (4, 3).</p></li>
<li><p><strong>equal_weight</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to normalize the data to equal weight. Defaults to False.</p></li>
<li><p><strong>color1</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the difference plot. Defaults to ‘green’.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the vertical line at position 0. Defaults to ‘red’.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. Defaults to None.</p></li>
<li><p><strong>use</strong> (<em>str</em><em>, </em><em>optional</em>) – Method to aggregate data (‘mean’, ‘median’, ‘sum’). Defaults to ‘mean’.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cumulativeDifference</span><span class="p">(</span><span class="s1">&#39;reference.csv&#39;</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">local_pos</span><span class="o">=</span><span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">300</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Cumulative Differences&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.cumulativePeaks">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">cumulativePeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df2=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_pos=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">equal_weight=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color1='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color2='magenta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.cumulativePeaks" title="Link to this definition"></a></dt>
<dd><p>Plot cumulative peaks metaplot for a single gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to CSV file or DataFrame containing reference data.</p></li>
<li><p><strong>df2</strong> (<em>pd.DataFrame</em>) – DataFrame containing the second dataset.</p></li>
<li><p><strong>local_pos</strong> (<em>list</em>) – List of positions of features (peaks/troughs).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot. Defaults to an empty string.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for filtering local positions. Defaults to None.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for filtering local positions. Defaults to None.</p></li>
<li><p><strong>window</strong> (<em>int</em><em>, </em><em>optional</em>) – Window size around each feature. Defaults to 50.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (4, 3).</p></li>
<li><p><strong>equal_weight</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to normalize the data to equal weight. Defaults to False.</p></li>
<li><p><strong>color1</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the reference dataset plot. Defaults to ‘green’.</p></li>
<li><p><strong>color2</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the second dataset plot. Defaults to ‘magenta’.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the vertical line at position 0. Defaults to ‘red’.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. Defaults to None.</p></li>
<li><p><strong>use</strong> (<em>str</em><em>, </em><em>optional</em>) – Method to aggregate data (‘mean’, ‘median’, ‘sum’). Defaults to ‘mean’.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cumulativePeaks</span><span class="p">(</span><span class="s1">&#39;reference.csv&#39;</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">local_pos</span><span class="o">=</span><span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">300</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Cumulative Peaks&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.enhancedVolcano">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">enhancedVolcano</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'log2FoldChange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'padj'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pval_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.enhancedVolcano" title="Link to this definition"></a></dt>
<dd><p>Generate an enhanced volcano plot based on DataFrame values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input DataFrame containing the data.</p></li>
<li><p><strong>x_fc</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name for the x-axis values (fold change). Defaults to ‘log2FoldChange’.</p></li>
<li><p><strong>y_pval</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name for the y-axis values (adjusted p-value). Defaults to ‘padj’.</p></li>
<li><p><strong>pval_cutoff</strong> (<em>float</em><em>, </em><em>optional</em>) – The p-value cutoff for significance. Defaults to 0.01.</p></li>
<li><p><strong>fc_cutoff</strong> (<em>float</em><em>, </em><em>optional</em>) – The fold change cutoff for significance. Defaults to 2.</p></li>
<li><p><strong>plot_n</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot non-significant points. Defaults to True.</p></li>
<li><p><strong>labels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to add labels to significant points. Defaults to True.</p></li>
<li><p><strong>xlim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The x-axis limits. Defaults to None.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The y-axis limits. Defaults to None.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The figure size. Defaults to (4, 4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the saved figure. Defaults to 300.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the plot. Defaults to None.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the figure. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">enhancedVolcano</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x_fc</span><span class="o">=</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span> <span class="n">y_pval</span><span class="o">=</span><span class="s1">&#39;padj&#39;</span><span class="p">,</span> <span class="n">pval_cutoff</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">fc_cutoff</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Volcano Plot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.generateSubplots">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">generateSubplots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.generateSubplots" title="Link to this definition"></a></dt>
<dd><p>Generate subplots for multiple dataframes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframes</strong> (<em>list</em>) – A list of dataframes to plot. Each element is a list of lists containing title, dataframe, and optional colormap.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The size of the figure. Defaults to (5, 3).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the figure in dots per inch. Defaults to 300.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">12</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">generateSubplots</span><span class="p">([</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">[[</span><span class="s1">&#39;Title 1&#39;</span><span class="p">,</span> <span class="n">df1</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span> 
<span class="gp">&gt;&gt;&gt; </span>     <span class="p">[</span><span class="s1">&#39;Title 2&#39;</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="kc">None</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="s1">&#39;plot.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.hplotSTARstats_readLen">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_readLen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.hplotSTARstats_readLen" title="Link to this definition"></a></dt>
<dd><p>Plot the average input read length as a horizontal bar plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hplotSTARstats_readLen</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.hplotSTARstats_reads">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.hplotSTARstats_reads" title="Link to this definition"></a></dt>
<dd><p>Plot the number of reads after pre-processing as a horizontal bar plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hplotSTARstats_reads</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.metaprofileAndHeatmap">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">metaprofileAndHeatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_metaplot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_heatmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'title':</span> <span class="pre">None}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_internal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">differential_heatmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[50,</span> <span class="pre">10,</span> <span class="pre">50]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.metaprofileAndHeatmap" title="Link to this definition"></a></dt>
<dd><p>Generate metaprofile and heatmap plots for multiple subsets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_metaplot</strong> (<em>dict</em>) – Data for metaprofile plot.</p></li>
<li><p><strong>data_heatmap</strong> (<em>dict</em>) – Data for heatmap plot.</p></li>
<li><p><strong>subsets</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of subsets with titles as keys and subset data as values. Defaults to {“title”: None}.</p></li>
<li><p><strong>agg_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Aggregation type for metaprofile (‘mean’, ‘median’, ‘sum’). Defaults to ‘mean’.</p></li>
<li><p><strong>normalize_internal</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to normalize data internally. Defaults to True.</p></li>
<li><p><strong>differential_heatmap</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot differential heatmap. Defaults to True.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The size of the figure. Defaults to (5, 3).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the figure in dots per inch. Defaults to 300.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the figure. If not provided, the figure will be displayed.</p></li>
<li><p><strong>bins</strong> (<em>list</em><em>, </em><em>optional</em>) – List of bin sizes for gene scheme. Defaults to [50, 10, 50].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metaprofileAndHeatmap</span><span class="p">(</span><span class="n">data_metaplot</span><span class="p">,</span> <span class="n">data_heatmap</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Subset 1&quot;</span><span class="p">:</span> <span class="n">subset1</span><span class="p">,</span> <span class="s2">&quot;Subset 2&quot;</span><span class="p">:</span> <span class="n">subset2</span><span class="p">},</span> 
<span class="gp">&gt;&gt;&gt; </span>                      <span class="n">agg_type</span><span class="o">=</span><span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="n">normalize_internal</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">differential_heatmap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>                      <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="s1">&#39;plot.png&#39;</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">50</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotAndFolding">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotAndFolding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dG=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name='median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotAndFolding" title="Link to this definition"></a></dt>
<dd><p>Plot a figure similar to a box plot with additional delta G values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the data with columns: [‘position’, ‘mean’, ‘median’, ‘std’] and optionally [‘nucleotide’, ‘q1’, ‘q3’, ‘max’, ‘min’].</p></li>
<li><p><strong>dG</strong> (<em>pd.Series</em>) – Series containing delta G values.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for the plot.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for the plot.</p></li>
<li><p><strong>legend</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display the legend. Defaults to True.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 3).</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to (None, 0.01).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure in dots per inch. Defaults to 150.</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the median line. Defaults to ‘green’.</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the median line in the legend. Defaults to ‘median’.</p></li>
<li><p><strong>h_lines</strong> (<em>list</em><em>, </em><em>optional</em>) – List of horizontal lines to add to the plot.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the horizontal lines. Defaults to ‘red’.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – Number to offset position if 5’ flank was used. Defaults to 0.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotAndFolding</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">dG</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Plot with Delta G&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="n">h_lines</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">60</span><span class="p">],</span> <span class="n">lc</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">fname</span><span class="o">=</span><span class="s1">&#39;plot.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats" title="Link to this definition"></a></dt>
<dd><p>Plot all STAR statistics including reads, read length, mapping, mismatches, and chimeric reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats_chimeric">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_chimeric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats_chimeric" title="Link to this definition"></a></dt>
<dd><p>Plot chimeric reads and splices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats_chimeric</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats_mapping">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats_mapping" title="Link to this definition"></a></dt>
<dd><p>Plot STAR mapping statistics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats_mapping</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats_mistmatches">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_mistmatches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats_mistmatches" title="Link to this definition"></a></dt>
<dd><p>Plot mapping errors including mismatches, deletions, and insertions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats_mistmatches</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats_readLen">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_readLen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats_readLen" title="Link to this definition"></a></dt>
<dd><p>Plot the average input read length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats_readLen</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plotSTARstats_reads">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plotSTARstats_reads" title="Link to this definition"></a></dt>
<dd><p>Plot the number of reads after pre-processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing STAR statistics.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the plot in dots per inch. Defaults to 150.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotSTARstats_reads</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plot_as_box_plot">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_as_box_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name='median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plot_as_box_plot" title="Link to this definition"></a></dt>
<dd><p>Plot a figure similar to a box plot: median, 2nd and 3rd quartiles, and min-max range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the data with columns: [‘position’, ‘mean’, ‘median’, ‘std’] and optionally [‘nucleotide’, ‘q1’, ‘q3’, ‘max’, ‘min’].</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for the plot.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for the plot.</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the median line in the legend. Defaults to ‘median’.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 3).</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to (None, 0.01).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure in dots per inch. Defaults to 150.</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the median line. Defaults to ‘green’.</p></li>
<li><p><strong>h_lines</strong> (<em>list</em><em>, </em><em>optional</em>) – List of horizontal lines to add to the plot.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the horizontal lines. Defaults to ‘red’.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – Number to offset position if 5’ flank was used. Defaults to 0.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;position&#39;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;median&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;q1&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;q3&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;min&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">)})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_as_box_plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Box Plot&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">h_lines</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">60</span><span class="p">],</span> <span class="n">lc</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">fname</span><span class="o">=</span><span class="s1">&#39;box_plot.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plot_diff">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label1='reference'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label2=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_medians=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_ranges=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plot_diff" title="Link to this definition"></a></dt>
<dd><p>Plot given dataset and reference, highlighting differences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to CSV file or DataFrame containing reference data.</p></li>
<li><p><strong>dataset</strong> (<em>pd.DataFrame</em>) – DataFrame containing the data to compare with columns: [‘position’, ‘mean’, ‘median’, ‘std’] and optionally [‘nucleotide’, ‘q1’, ‘q3’, ‘max’, ‘min’].</p></li>
<li><p><strong>ranges</strong> (<em>str</em><em>, </em><em>optional</em>) – Type of range to highlight (‘mm’ for min-max or ‘qq’ for q1-q3). Defaults to ‘mm’.</p></li>
<li><p><strong>label1</strong> (<em>str</em><em>, </em><em>optional</em>) – Label for the reference dataset in the legend. Defaults to “reference”.</p></li>
<li><p><strong>label2</strong> (<em>str</em><em>, </em><em>optional</em>) – Label for the comparison dataset in the legend.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for the plot.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for the plot.</p></li>
<li><p><strong>plot_medians</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot medians. Defaults to True.</p></li>
<li><p><strong>plot_ranges</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot ranges. Defaults to True.</p></li>
<li><p><strong>legend</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display the legend. Defaults to True.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure in dots per inch. Defaults to 150.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 3).</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to (None, 0.01).</p></li>
<li><p><strong>h_lines</strong> (<em>list</em><em>, </em><em>optional</em>) – List of horizontal lines to add to the plot.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the horizontal lines. Defaults to ‘red’.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – Number to offset position if 5’ flank was used. Defaults to 0.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plot_diff</span><span class="p">(</span><span class="s1">&#39;reference.csv&#39;</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="s1">&#39;qq&#39;</span><span class="p">,</span> <span class="n">label1</span><span class="o">=</span><span class="s1">&#39;Reference&#39;</span><span class="p">,</span> <span class="n">label2</span><span class="o">=</span><span class="s1">&#39;Comparison&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Difference Plot&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">plot_medians</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_ranges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">h_lines</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">60</span><span class="p">],</span> <span class="n">lc</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">fname</span><span class="o">=</span><span class="s1">&#39;difference_plot.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plot_heatmap">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title='Heatmap</span> <span class="pre">of</span> <span class="pre">differences</span> <span class="pre">between</span> <span class="pre">dataset</span> <span class="pre">and</span> <span class="pre">reference</span> <span class="pre">plot</span> <span class="pre">for</span> <span class="pre">RDN37-1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plot_heatmap" title="Link to this definition"></a></dt>
<dd><p>Plot heatmap of differences between dataset and reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the differences.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot. Defaults to ‘Heatmap of differences between dataset and reference plot for RDN37-1’.</p></li>
<li><p><strong>vmin</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum value for the colormap. Defaults to None.</p></li>
<li><p><strong>vmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for the colormap. Defaults to None.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (20, 10).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure in dots per inch. Defaults to 300.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Difference Heatmap&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">fname</span><span class="o">=</span><span class="s1">&#39;heatmap.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.plot_to_compare">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_to_compare</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color1='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color2='black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_label=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.plot_to_compare" title="Link to this definition"></a></dt>
<dd><p>Plot a figure to compare two datasets similar to a box plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to CSV file or DataFrame containing reference data.</p></li>
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – DataFrame containing the data to compare.</p></li>
<li><p><strong>color1</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the reference dataset plot. Defaults to ‘green’.</p></li>
<li><p><strong>color2</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the comparison dataset plot. Defaults to ‘black’.</p></li>
<li><p><strong>ref_label</strong> (<em>str</em><em>, </em><em>optional</em>) – Label for the reference dataset in the legend.</p></li>
<li><p><strong>label</strong> (<em>str</em><em>, </em><em>optional</em>) – Label for the comparison dataset in the legend.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the plot.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start position for the plot.</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – Stop position for the plot.</p></li>
<li><p><strong>legend</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display the legend. Defaults to True.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. Defaults to (7, 3).</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Limits for the y-axis. Defaults to (None, 0.01).</p></li>
<li><p><strong>h_lines</strong> (<em>list</em><em>, </em><em>optional</em>) – List of horizontal lines to add to the plot.</p></li>
<li><p><strong>lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the horizontal lines. Defaults to ‘red’.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the figure in dots per inch. Defaults to 150.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – Number to offset position if 5’ flank was used. Defaults to 300.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the figure. If not provided, the figure will be displayed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plot_to_compare</span><span class="p">(</span><span class="s1">&#39;reference.csv&#39;</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">color1</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">color2</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">ref_label</span><span class="o">=</span><span class="s1">&#39;Reference&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Comparison&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Comparison Plot&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">h_lines</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">60</span><span class="p">],</span> <span class="n">lc</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">fname</span><span class="o">=</span><span class="s1">&#39;comparison_plot.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.select_colors">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">select_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Spectral'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.select_colors" title="Link to this definition"></a></dt>
<dd><p>Select a list of colors from a colormap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of colors to select, defaults to int()</p></li>
<li><p><strong>name_cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the colormap to use, defaults to ‘Spectral’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of colors</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="trxtools.plotting.vennDiagram">
<span class="sig-prename descclassname"><span class="pre">trxtools.plotting.</span></span><span class="sig-name descname"><span class="pre">vennDiagram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">dataframes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['df1',</span> <span class="pre">'df2',</span> <span class="pre">'df3']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('skyblue',</span> <span class="pre">'lightgreen',</span> <span class="pre">'lightpink')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#trxtools.plotting.vennDiagram" title="Link to this definition"></a></dt>
<dd><p>Generate a Venn diagram for 2 or 3 DataFrames.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframes</strong> (<em>list</em><em> of </em><em>pd.DataFrame</em>) – Input DataFrames containing the data.</p></li>
<li><p><strong>labels</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Labels for the DataFrames. Defaults to [‘df1’,’df2’,’df3’].</p></li>
<li><p><strong>colors</strong> (<em>tuple</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Colors for the Venn diagram. Defaults to (‘skyblue’, ‘lightgreen’, ‘lightpink’).</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the plot. Defaults to None.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the figure. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Example<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vennDiagram</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Group 1&#39;</span><span class="p">,</span> <span class="s1">&#39;Group 2&#39;</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Venn Diagram&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="scripts.html" class="btn btn-neutral float-left" title="Scripts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="release.html" class="btn btn-neutral float-right" title="Release information" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Turowski Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>