<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>TTools.SAMgenome &mdash; TTools 0.1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> TTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../TTools.html">TTools package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">TTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>TTools.SAMgenome</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for TTools.SAMgenome</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">TTools.SAMgeneral</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">TTools.SAMtranscripts</span> <span class="kn">import</span> <span class="n">transcript2profile</span>

<span class="c1">####################################################</span>
<span class="c1">#   intermediate functions level -2 (from final)   #</span>
<span class="c1">####################################################</span>

<div class="viewcode-block" id="chromosome2profile3end"><a class="viewcode-back" href="../../TTools.html#TTools.SAMgenome.chromosome2profile3end">[docs]</a><span class="k">def</span> <span class="nf">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="p">[],</span> <span class="n">length</span><span class="o">=</span><span class="nb">int</span><span class="p">(),</span> <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;FWD&#39;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Generates profile for the 3&#39; ends of reads and saves position of non-coded end</span>

<span class="sd">    :param l: list of triple tuples (position, cigar_string, sequence), defaults to []</span>
<span class="sd">    :type l: list</span>
<span class="sd">    :param length: length of chromosome</span>
<span class="sd">    :type length: int</span>
<span class="sd">    :param strand: ``&#39;FWD&#39;`` or ``&#39;REV&#39;``, defaults to &#39;FWD&#39;</span>
<span class="sd">    :type strand: str, optional</span>
<span class="sd">    :return: profile, noncoded</span>
<span class="sd">    :rtype: np.array, list of tuples</span>

<span class="sd">    &gt;&gt;&gt; chromosome2profile3end(l=[(10,&quot;3S15M1D9M2S&quot;,&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;)],length=50,strand=&quot;FWD&quot;)</span>
<span class="sd">    (0     0.0</span>
<span class="sd">    1     0.0</span>
<span class="sd">    ...</span>
<span class="sd">    34    0.0</span>
<span class="sd">    35    1.0</span>
<span class="sd">    36    0.0</span>
<span class="sd">    ...</span>
<span class="sd">    50    0.0</span>
<span class="sd">    dtype: float64,</span>
<span class="sd">    [(35, &#39;CC&#39;)])</span>
<span class="sd">    &gt;&gt;&gt; chromosome2profile3end(l=[(40,&quot;3S15M1D9M2S&quot;,&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;)],length=50,strand=&quot;REV&quot;)</span>
<span class="sd">    (0     0.0</span>
<span class="sd">    1     0.0</span>
<span class="sd">    ...</span>
<span class="sd">    39    0.0</span>
<span class="sd">    40    1.0</span>
<span class="sd">    41    0.0</span>
<span class="sd">    ...</span>
<span class="sd">    50    0.0</span>
<span class="sd">    dtype: float64,</span>
<span class="sd">    [(40, &#39;AAA&#39;)])</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">noncoded</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">cigar_string</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">in</span> <span class="n">l</span><span class="p">:</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">groupCIGAR</span><span class="p">(</span><span class="n">cigar_string</span><span class="p">)</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">stripCIGAR</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="c1">#removes H</span>

        <span class="c1"># outputs</span>
        <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="s2">&quot;FWD&quot;</span><span class="p">:</span> <span class="c1">#strand &quot;+&quot;</span>
            <span class="n">nonMatchThree</span> <span class="o">=</span> <span class="n">tostripCIGARthree</span><span class="p">(</span><span class="n">match</span><span class="p">)</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">stripSubstitutions</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="c1">#removes S</span>
            <span class="n">readLength</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">match</span><span class="p">])</span>
            <span class="n">readEnd</span> <span class="o">=</span> <span class="n">position</span><span class="o">+</span><span class="n">readLength</span>
            <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">readEnd</span><span class="p">)</span> <span class="c1">#appending</span>
            <span class="k">if</span> <span class="n">nonMatchThree</span><span class="p">:</span> 
                <span class="n">noncoded</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">readEnd</span><span class="p">,</span><span class="n">sequence</span><span class="p">[</span><span class="o">-</span><span class="n">nonMatchThree</span><span class="p">:]))</span>
                
        <span class="k">elif</span> <span class="n">strand</span><span class="o">==</span><span class="s2">&quot;REV&quot;</span><span class="p">:</span> <span class="c1">#strand &quot;-&quot;</span>
            <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="c1">#appending</span>
            <span class="n">nonMatchFive</span> <span class="o">=</span> <span class="n">tostripCIGARfive</span><span class="p">(</span><span class="n">match</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">nonMatchFive</span><span class="p">:</span>
                <span class="n">revcomp</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">reverse_complement_DNA</span><span class="p">(</span><span class="n">sequence</span><span class="p">[:</span><span class="n">nonMatchFive</span><span class="p">])</span>
                <span class="n">noncoded</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">position</span><span class="p">,</span><span class="n">revcomp</span><span class="p">))</span>
      
    <span class="c1"># returning the profile</span>
    <span class="n">profile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">hits</span><span class="p">))</span>  <span class="c1"># faster that using zip and numpy</span>
    <span class="n">profile</span> <span class="o">=</span> <span class="n">profile</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">RangeIndex</span><span class="p">(</span><span class="n">length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># fills spaces with 0 counts</span>
    <span class="k">return</span> <span class="n">profile</span><span class="p">,</span> <span class="n">noncoded</span></div>

<span class="c1">####################################################</span>
<span class="c1">#   intermediate functions level -1 (from final)   #</span>
<span class="c1">####################################################</span>

<div class="viewcode-block" id="reads2genome"><a class="viewcode-back" href="../../TTools.html#TTools.SAMgenome.reads2genome">[docs]</a><span class="k">def</span> <span class="nf">reads2genome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="nb">str</span><span class="p">(),</span> <span class="n">dirPath</span><span class="o">=</span><span class="nb">str</span><span class="p">(),</span> <span class="n">df_details</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
    <span class="sd">&#39;&#39;&#39;Function used by sam2genome. Works for both strands.</span>

<span class="sd">    :param name: name of experiment</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param dirPath:</span>
<span class="sd">    :type dirPath: str</span>
<span class="sd">    :param df_details: lengths of chromosomes</span>
<span class="sd">    :type df_details: DataFrame</span>
<span class="sd">    :return: output_df_fwd, output_df_rev, log</span>
<span class="sd">    :rtype: DataFrame, DataFrame, list</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="s1">&#39;CIGAR&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;details&#39;</span><span class="p">]</span>
    <span class="n">df_input_fwd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_fwd.tab&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span>
    <span class="n">df_input_rev</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_rev.tab&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span>

    <span class="n">output_df_fwd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">output_df_rev</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">log</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_input_fwd</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="n">df_details</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
            <span class="n">l</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CIGAR&#39;</span><span class="p">]))</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">transcript2profile</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">)</span>
            <span class="n">output_df_fwd</span> <span class="o">=</span> <span class="n">output_df_fwd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s1</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - FWD profile generated successfully&quot;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - FWD profile FAILED&quot;</span><span class="p">)</span>
            
    <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_input_rev</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="n">df_details</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
            <span class="n">l</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CIGAR&#39;</span><span class="p">]))</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">transcript2profile</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">)</span>
            <span class="n">output_df_rev</span> <span class="o">=</span> <span class="n">output_df_rev</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s1</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - REV profile generated successfully&quot;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - REV profile FAILED&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">output_df_fwd</span><span class="p">,</span> <span class="n">output_df_rev</span><span class="p">,</span> <span class="n">log</span></div>


<div class="viewcode-block" id="reads2genome3end"><a class="viewcode-back" href="../../TTools.html#TTools.SAMgenome.reads2genome3end">[docs]</a><span class="k">def</span> <span class="nf">reads2genome3end</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="nb">str</span><span class="p">(),</span> <span class="n">dirPath</span><span class="o">=</span><span class="nb">str</span><span class="p">(),</span> <span class="n">df_details</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span><span class="n">noncoded</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Function used by sam2genome3end. Works for both strands.</span>

<span class="sd">    :param name: name of experiment</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param dirPath:</span>
<span class="sd">    :type dirPath: str</span>
<span class="sd">    :param df_details: lengths of chromosomes</span>
<span class="sd">    :type df_details: DataFrame</span>
<span class="sd">    :param noncoded: If True then will parse and save non-coded ends, defaults to True</span>
<span class="sd">    :type noncoded: bool, optional</span>
<span class="sd">    :return: output_df_fwd, output_df_rev, log, noncoded_fwd, noncoded_rev</span>
<span class="sd">    :rtype: DataFrame, DataFrame, list, DataFrame, DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="s1">&#39;CIGAR&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;details&#39;</span><span class="p">]</span>
    <span class="n">df_input_fwd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_fwd.tab&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span>
    <span class="n">df_input_rev</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_rev.tab&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span>

    <span class="n">output_df_fwd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">output_df_rev</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">noncoded_fwd</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">noncoded_rev</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">log</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># strand &quot;+&quot;</span>
    <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_input_fwd</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="n">df_details</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
            <span class="n">l</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CIGAR&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]))</span>         
            <span class="c1">#processing reads</span>
            <span class="n">s1_profile</span><span class="p">,</span> <span class="n">l1_noncoded</span> <span class="o">=</span> <span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;FWD&#39;</span><span class="p">)</span>
            <span class="n">output_df_fwd</span> <span class="o">=</span> <span class="n">output_df_fwd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s1_profile</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">noncoded</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
                <span class="n">noncoded_fwd</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">l1_noncoded</span>

            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - FWD profile and/or list of noncoded ends generated successfully&quot;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - FWD profile and/or list of noncoded ends FAILED&quot;</span><span class="p">)</span>
    
    <span class="c1"># strand &quot;-&quot;</span>
    <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_input_rev</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="n">df_details</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span>
            <span class="n">l</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;CIGAR&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]))</span>         
            <span class="c1">#processing reads</span>
            <span class="n">s1_profile</span><span class="p">,</span> <span class="n">l1_noncoded</span> <span class="o">=</span> <span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;REV&#39;</span><span class="p">)</span>
            <span class="n">output_df_rev</span> <span class="o">=</span> <span class="n">output_df_rev</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s1_profile</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">noncoded</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
                <span class="n">noncoded_rev</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">l1_noncoded</span>

            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - REV profile and/or list of noncoded ends generated successfully&quot;</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="s2">&quot; - REV profile and/or list of noncoded ends FAILED&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">output_df_fwd</span><span class="p">,</span> <span class="n">output_df_rev</span><span class="p">,</span> <span class="n">log</span><span class="p">,</span> <span class="n">noncoded_fwd</span><span class="p">,</span> <span class="n">noncoded_rev</span></div>


<span class="c1">####################################################</span>
<span class="c1">#              final functions (level 0)           #</span>
<span class="c1">####################################################</span>

<div class="viewcode-block" id="sam2genome"><a class="viewcode-back" href="../../TTools.html#TTools.SAMgenome.sam2genome">[docs]</a><span class="k">def</span> <span class="nf">sam2genome</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">toClear</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">pickle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">chunks</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Function handling SAM files and generating profiles. Executed using wrapping script SAM2profilesGenomic.py.</span>

<span class="sd">    :param filename: </span>
<span class="sd">    :type filename: str</span>
<span class="sd">    :param path: </span>
<span class="sd">    :type path: str</span>
<span class="sd">    :param toClear: element of filename to be removed, defaults to &#39;&#39;</span>
<span class="sd">    :type toClear: str, optional</span>
<span class="sd">    :param pickle: save output in pickle format, defaults to False</span>
<span class="sd">    :type pickle: bool, optional</span>
<span class="sd">    :param chunks: Read SAM file in chunks, defaults to 0</span>
<span class="sd">    :type chunks: int, optional</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># making working directory</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">filename</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">toClear</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">toClear</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">timestamp</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">timestampRandomInt</span><span class="p">()</span>
    <span class="n">dirPath</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="n">timestamp</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dirPath</span><span class="p">)</span>

    <span class="c1">#geneList = chromosome list from the @SQ header of SAM file&#39;</span>
    <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep @SQ &quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | cut -f2 | sed &#39;s/SN\://&#39; &gt; &quot;</span> <span class="o">+</span> <span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.list&quot;</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    <span class="n">geneList</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">read_list</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.list&quot;</span><span class="p">)</span>
    <span class="c1">#chromosome lengths from SAM header</span>
    <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep @SQ &quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | cut -f4 | sed &#39;s/LN\://&#39; &gt; &quot;</span> <span class="o">+</span> <span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.len&quot;</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    <span class="n">geneLen</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">read_list</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.len&quot;</span><span class="p">)</span>
    
    <span class="n">df_details</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrName&quot;</span> <span class="p">:</span> <span class="n">geneList</span><span class="p">,</span> <span class="s2">&quot;length&quot;</span> <span class="p">:</span> <span class="n">geneLen</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;chrName&#39;</span><span class="p">)</span>
    <span class="n">df_details</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_details</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    
    <span class="c1"># tempfiles</span>
    <span class="c1">##list of files if red as chunks</span>
    <span class="k">if</span> <span class="n">chunks</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">chunkList</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">chunks</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">chunkList</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">chunks</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>

    <span class="c1">##genes means chromosomes for this function</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">geneList</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chunkList</span><span class="p">:</span>
        <span class="n">geneListFileName</span> <span class="o">=</span> <span class="s2">&quot;/geneList_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.tab&quot;</span>
        <span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">chunks</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="n">geneListFileName</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
    <span class="c1">##Select FWD and REV reads from SAM file</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">dirPath</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chunkList</span><span class="p">:</span>
        <span class="n">geneListFileName</span> <span class="o">=</span> <span class="s2">&quot;geneList_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.tab&quot;</span>
        
        <span class="nb">print</span>
        <span class="c1">#FLAG 0 and 256 for single end reads, aligned, forward match</span>
        <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep -v ^@ ../&quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | grep -f &quot;</span> <span class="o">+</span> <span class="n">geneListFileName</span> <span class="o">+</span>\
                  <span class="s2">&quot; | awk -F&#39;</span><span class="se">\t</span><span class="s2">&#39; &#39;BEGIN{OFS = FS} $2==0||$2==256{print $2,$3, $4, $6, $10, $12}&#39; &gt; &quot;</span><span class="o">+</span>\
                  <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_fwd.tab&quot;</span>
        <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
        
        <span class="c1">#FLAG 16 and 272 for single end reads, aligned, forward match</span>
        <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep -v ^@ ../&quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | grep -f &quot;</span> <span class="o">+</span> <span class="n">geneListFileName</span> <span class="o">+</span>\
                  <span class="s2">&quot; | awk -F&#39;</span><span class="se">\t</span><span class="s2">&#39; &#39;BEGIN{OFS = FS} $2==16||$2==272{print $2,$3, $4, $6, $10, $12}&#39; &gt; &quot;</span><span class="o">+</span>\
                  <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_rev.tab&quot;</span>
        <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="s2">&quot;cat &quot;</span><span class="o">+</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;*fwd.tab &gt; &quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span><span class="s2">&quot;_fwd.tab&quot;</span><span class="p">)</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="s2">&quot;cat &quot;</span><span class="o">+</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;*rev.tab &gt; &quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span><span class="s2">&quot;_rev.tab&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reads selected.&quot;</span><span class="p">)</span>

    <span class="c1">#Reads to profiles</span>
    <span class="n">df_profiles_fwd</span><span class="p">,</span> <span class="n">df_profiles_rev</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="n">reads2genome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">dirPath</span><span class="o">=</span><span class="n">dirPath</span><span class="p">,</span><span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">)</span>
    
    <span class="c1">#save output</span>
    <span class="k">if</span> <span class="n">pickle</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
        <span class="n">df_profiles_fwd</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_reads_fwd.pcl&quot;</span><span class="p">)</span>
        <span class="n">df_profiles_rev</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_reads_rev.pcl&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">pickle</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
        <span class="n">df_profiles_fwd</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_reads_fwd.csv&quot;</span><span class="p">)</span>
        <span class="n">df_profiles_rev</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_reads_rev.csv&quot;</span><span class="p">)</span>
    <span class="c1">#save log</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_reads.log&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">log</span><span class="p">:</span>
            <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># clean</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="n">timestamp</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done.&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="sam2genome3end"><a class="viewcode-back" href="../../TTools.html#TTools.SAMgenome.sam2genome3end">[docs]</a><span class="k">def</span> <span class="nf">sam2genome3end</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">toClear</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">pickle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">chunks</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">noncoded</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Function handling SAM files and generating profiles for the 3&#39; end of reads. Executed using wrapping script SAM2profilesGenomic.py.</span>

<span class="sd">    :param filename:</span>
<span class="sd">    :type filename: str</span>
<span class="sd">    :param path: </span>
<span class="sd">    :type path: str</span>
<span class="sd">    :param toClear: element of filename to be removed, defaults to &#39;&#39;</span>
<span class="sd">    :type toClear: str, optional</span>
<span class="sd">    :param pickle: save output in pickle format, defaults to False</span>
<span class="sd">    :type pickle: bool, optional</span>
<span class="sd">    :param chunks: Read SAM file in chunks, defaults to 0</span>
<span class="sd">    :type chunks: int, optional</span>
<span class="sd">    :param noncoded: Process and save non-coded ends, defaults to True</span>
<span class="sd">    :type noncoded: bool, optional</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="c1"># making working directory</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">filename</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">toClear</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">toClear</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">timestamp</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">timestampRandomInt</span><span class="p">()</span>
    <span class="n">dirPath</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="n">timestamp</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dirPath</span><span class="p">)</span>

    <span class="c1">#geneList = chromosome list from the @SQ header of SAM file&#39;</span>
    <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep @SQ &quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | cut -f2 | sed &#39;s/SN\://&#39; &gt; &quot;</span> <span class="o">+</span> <span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.list&quot;</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    <span class="n">geneList</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">read_list</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.list&quot;</span><span class="p">)</span>
    <span class="c1">#chromosome lengths from SAM header</span>
    <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep @SQ &quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | cut -f4 | sed &#39;s/LN\://&#39; &gt; &quot;</span> <span class="o">+</span> <span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.len&quot;</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    <span class="n">geneLen</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">read_list</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_chr.len&quot;</span><span class="p">)</span>
    
    <span class="n">df_details</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrName&quot;</span> <span class="p">:</span> <span class="n">geneList</span><span class="p">,</span> <span class="s2">&quot;length&quot;</span> <span class="p">:</span> <span class="n">geneLen</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;chrName&#39;</span><span class="p">)</span>
    <span class="n">df_details</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_details</span><span class="p">[</span><span class="s1">&#39;length&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    
    <span class="c1"># tempfiles</span>
    <span class="c1">##list of files if red as chunks</span>
    <span class="k">if</span> <span class="n">chunks</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">chunkList</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">chunks</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">chunkList</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">chunks</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>

    <span class="c1">##genes means chromosomes for this function</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">geneList</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chunkList</span><span class="p">:</span>
        <span class="n">geneListFileName</span> <span class="o">=</span> <span class="s2">&quot;/geneList_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.tab&quot;</span>
        <span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">chunks</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">dirPath</span> <span class="o">+</span> <span class="n">geneListFileName</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
    <span class="c1">##Select FWD and REV reads from SAM file</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">dirPath</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chunkList</span><span class="p">:</span>
        <span class="n">geneListFileName</span> <span class="o">=</span> <span class="s2">&quot;geneList_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.tab&quot;</span>
        
        <span class="nb">print</span>
        <span class="c1">#FLAG 0 and 256 for single end reads, aligned, forward match</span>
        <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep -v ^@ ../&quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | grep -f &quot;</span> <span class="o">+</span> <span class="n">geneListFileName</span> <span class="o">+</span>\
                  <span class="s2">&quot; | awk -F&#39;</span><span class="se">\t</span><span class="s2">&#39; &#39;BEGIN{OFS = FS} $2==0||$2==256{print $2,$3, $4, $6, $10, $12}&#39; &gt; &quot;</span><span class="o">+</span>\
                  <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_fwd.tab&quot;</span>
        <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
        
        <span class="c1">#FLAG 16 and 272 for single end reads, aligned, forward match</span>
        <span class="n">command</span> <span class="o">=</span> <span class="s2">&quot;grep -v ^@ ../&quot;</span> <span class="o">+</span> <span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot; | grep -f &quot;</span> <span class="o">+</span> <span class="n">geneListFileName</span> <span class="o">+</span>\
                  <span class="s2">&quot; | awk -F&#39;</span><span class="se">\t</span><span class="s2">&#39; &#39;BEGIN{OFS = FS} $2==16||$2==272{print $2,$3, $4, $6, $10, $12}&#39; &gt; &quot;</span><span class="o">+</span>\
                  <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_rev.tab&quot;</span>
        <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
    
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="s2">&quot;cat &quot;</span><span class="o">+</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;*fwd.tab &gt; &quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span><span class="s2">&quot;_fwd.tab&quot;</span><span class="p">)</span>
    <span class="n">tt</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">bashCommand</span><span class="p">(</span><span class="s2">&quot;cat &quot;</span><span class="o">+</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;*rev.tab &gt; &quot;</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span><span class="s2">&quot;_rev.tab&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reads selected.&quot;</span><span class="p">)</span>

    <span class="c1">#Reads to profiles</span>
    <span class="n">df_profiles_fwd</span><span class="p">,</span> <span class="n">df_profiles_rev</span><span class="p">,</span> <span class="n">log</span><span class="p">,</span> <span class="n">noncoded_fwd</span><span class="p">,</span> <span class="n">noncoded_rev</span> <span class="o">=</span> <span class="n">reads2genome3end</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">dirPath</span><span class="o">=</span><span class="n">dirPath</span><span class="p">,</span><span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">,</span><span class="n">noncoded</span><span class="o">=</span><span class="n">noncoded</span><span class="p">)</span>
    
    <span class="c1">#parse raw non-coded ends, keep only &gt;=3 nt long</span>
    <span class="n">noncoded_fwd</span> <span class="o">=</span> <span class="n">parseNoncoded</span><span class="p">(</span><span class="n">noncoded_fwd</span><span class="p">,</span> <span class="n">minLen</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    <span class="n">noncoded_rev</span> <span class="o">=</span> <span class="n">parseNoncoded</span><span class="p">(</span><span class="n">noncoded_rev</span><span class="p">,</span> <span class="n">minLen</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    
    <span class="c1">#select only polyA reads (&gt;=3 and 75% of A content) and prepare profile</span>
    <span class="n">noncoded_profile_fwd</span> <span class="o">=</span> <span class="n">noncoded2profile</span><span class="p">(</span><span class="n">selectPolyA</span><span class="p">(</span><span class="n">noncoded_fwd</span><span class="p">),</span><span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">)</span>
    <span class="n">noncoded_profile_rev</span> <span class="o">=</span> <span class="n">noncoded2profile</span><span class="p">(</span><span class="n">selectPolyA</span><span class="p">(</span><span class="n">noncoded_rev</span><span class="p">),</span><span class="n">df_details</span><span class="o">=</span><span class="n">df_details</span><span class="p">)</span>

    <span class="c1">#save output</span>
    <span class="k">if</span> <span class="n">pickle</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
        <span class="n">df_profiles_fwd</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_3end_fwd.pcl&quot;</span><span class="p">)</span>
        <span class="n">df_profiles_rev</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_3end_rev.pcl&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">pickle</span><span class="o">==</span><span class="kc">False</span><span class="p">:</span>
        <span class="n">df_profiles_fwd</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_3end_fwd.csv&quot;</span><span class="p">)</span>
        <span class="n">df_profiles_rev</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_3end_rev.csv&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">noncoded</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
        <span class="c1"># noncoded_fwd.to_csv(path + name + &quot;_noncoded_3end_fwd.csv&quot;)</span>
        <span class="c1"># noncoded_rev.to_csv(path + name + &quot;_noncoded_3end_rev.csv&quot;)</span>
        <span class="n">noncoded_profile_fwd</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_noncoded_PROFILES_3end_fwd.csv&quot;</span><span class="p">)</span>
        <span class="n">noncoded_profile_rev</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_noncoded_PROFILES_3end_rev.csv&quot;</span><span class="p">)</span>
    
    <span class="c1">#save log</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;_PROFILES_3end.log&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">log</span><span class="p">:</span>
            <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># clean</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="n">timestamp</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done.&quot;</span><span class="p">)</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Tomasz W. Turowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>