<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Scripts &mdash; trxtools 0.2.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="trxtools" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> trxtools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#main-scripts">Main scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sam2profilesgenomic-py">SAM2profilesGenomic.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sam2profiles-py">SAM2profiles.py</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#supporting-scripts">Supporting scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#csv2pickle-py">csv2pickle.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mergesalmon-py">mergeSalmon.py</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">trxtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Scripts</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/scripts.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="scripts">
<h1>Scripts<a class="headerlink" href="#scripts" title="Permalink to this heading"></a></h1>
<section id="main-scripts">
<h2>Main scripts<a class="headerlink" href="#main-scripts" title="Permalink to this heading"></a></h2>
<section id="sam2profilesgenomic-py">
<h3>SAM2profilesGenomic.py<a class="headerlink" href="#sam2profilesgenomic-py" title="Permalink to this heading"></a></h3>
<p>usage: Generates genomic profiles from SAM files.</p>
<dl>
<dt>optional arguments:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
</dd>
<dt>Options for input files:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span></kbd></dt>
<dd><p>SAM file (default: None)</p>
</dd>
</dl>
<p>-u {read,3end,5end}  Generate profiles using reads (read) or the 3’ ends (end) (default: 3end)
-n                   Save non-coded polyA ends. Can be used ONLY with: -u 3end (default: False)
-r                   Save all non-coded ends. Can be used ONLY with: -u 3end (default: False)
–csv                Save output as csv. Default is pickle (default: False)
-c TOCLEAR           String of signs to be cleared from the name of SAM file (default: _comp_flexbar_STARAligned.out)
–chunks CHUNKS      Divide list of genes into a chunks of given size (default: 0)</p>
</dd>
</dl>
</section>
<section id="sam2profiles-py">
<h3>SAM2profiles.py<a class="headerlink" href="#sam2profiles-py" title="Permalink to this heading"></a></h3>
<p>usage: Generates profiles from SAM files for a list of given transcripts.</p>
<dl>
<dt>optional arguments:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
</dd>
<dt>Options for input files:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span></kbd></dt>
<dd><p>SAM file (default: None)</p>
</dd>
<dt><kbd><span class="option">-l <var>FILE</var></span></kbd></dt>
<dd><p>list of genes file (default: None)</p>
</dd>
<dt><kbd><span class="option">-d <var>DETAILS_FILE</var></span></kbd></dt>
<dd><p>file with transcriptome details. (default:
/homes/tollervey/COVID19/databases/transcriptome_details_biomart_tRNA_rRNA_UPDATEJan2021.tab)</p>
</dd>
<dt><kbd><span class="option">--del</span></kbd></dt>
<dd><p>Generate additional profiles for deletions (default: False)</p>
</dd>
<dt><kbd><span class="option">-p</span></kbd></dt>
<dd><p>SAve output as pickle DataFrame (default: False)</p>
</dd>
<dt><kbd><span class="option">-e <var>EXPAND</var></span></kbd></dt>
<dd><p>For deletions position can be expanded by the value on each side (e=5 gives 10 nt long) (default: 5)</p>
</dd>
<dt><kbd><span class="option">-c <var>TOCLEAR</var></span></kbd></dt>
<dd><p>String of signs to be cleared from the name of SAM file (default: _comp_flexbar_STARAligned.out)</p>
</dd>
<dt><kbd><span class="option">--chunks <var>CHUNKS</var></span></kbd></dt>
<dd><p>Divide list of genes into a chunks of given size (default: 0)</p>
</dd>
</dl>
</dd>
</dl>
</section>
</section>
<section id="supporting-scripts">
<h2>Supporting scripts<a class="headerlink" href="#supporting-scripts" title="Permalink to this heading"></a></h2>
<section id="csv2pickle-py">
<h3>csv2pickle.py<a class="headerlink" href="#csv2pickle-py" title="Permalink to this heading"></a></h3>
<p>usage: Read CSV file and save as a pickle</p>
<dl>
<dt>optional arguments:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
</dd>
<dt>Options for input files:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-f <var>FILE</var></span></kbd></dt>
<dd><p>SAM file (default: None)</p>
</dd>
<dt><kbd><span class="option">--gzip</span></kbd></dt>
<dd><p>gzip compression (default: False)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="mergesalmon-py">
<h3>mergeSalmon.py<a class="headerlink" href="#mergesalmon-py" title="Permalink to this heading"></a></h3>
<p>usage: merges Salmon quantification files to one tab file</p>
<dl>
<dt>optional arguments:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
</dd>
<dt>Options for input files:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-i <var>INPUT</var></span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory (default: None)</p>
</dd>
<dt><kbd><span class="option">-o <var>OUTPUT</var></span></kbd></dt>
<dd><p>Name of output file (default: merged)</p>
</dd>
<dt><kbd><span class="option">-u <var>USE</var></span></kbd></dt>
<dd><p>column to be used {TPM,NumReads} (default: NumReads)</p>
</dd>
<dt><kbd><span class="option">-f <var>FILTER</var></span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory. Optional as additional filter (default: None)</p>
</dd>
<dt><kbd><span class="option">-c <var>CLEAR</var></span></kbd></dt>
<dd><p>String of signs to be found in Salmon output directory, will be cleared (default: None)</p>
</dd>
<dt><kbd><span class="option">-a <var>ADD</var></span></kbd></dt>
<dd><p>String of signs to be added to experiment name (default: None)</p>
</dd>
</dl>
</dd>
</dl>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="trxtools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Tomasz W. Turowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>