Search.setIndex({"docnames": ["TTools", "index", "scripts"], "filenames": ["TTools.rst", "index.rst", "scripts.rst"], "titles": ["Modules", "TTools", "Scripts"], "terms": {"wa": [0, 1], "initi": 1, "storag": 1, "function": [0, 1], "script": [0, 1], "us": [0, 1, 2], "turowski": 1, "lab": 1, "conduct": 1, "variou": 1, "bioinformat": 1, "There": 1, "i": [0, 1, 2], "intent": 1, "creat": [0, 1], "comprehens": 1, "packag": 1, "can": [0, 1, 2], "oper": 1, "non": [0, 1, 2], "experienc": 1, "user": 1, "some": [0, 1], "ar": [0, 1], "easi": 1, "run": [0, 1], "everyon": 1, "importantli": 1, "provid": [0, 1], "skill": 1, "individu": 1, "major": 1, "focu": 1, "put": 1, "postprocess": 1, "high": 1, "troughput": 1, "sequenc": [0, 1], "data": [0, 1], "origin": 1, "from": [0, 1, 2], "followin": 1, "method": 1, "crac": [0, 1], "clip": [0, 1], "rna": [0, 1], "seq": [0, 1], "trna": 1, "net": 1, "chip": 1, "specif": [0, 1], "file": [0, 1, 2], "name": [0, 1, 2], "system": 1, "expid": [0, 1], "expdat": [0, 1], "protein": [0, 1], "condition1": [0, 1], "condition2": [0, 1], "condition3": [0, 1], "e": [0, 1, 2], "c123_jk050420_polr2a": 1, "flag_wt_arsenite30min": 1, "an": [0, 1], "experi": [0, 1, 2], "c123": 1, "person": 1, "jk": 1, "020420": 1, "thi": [0, 1, 2], "puldown": 1, "polr2a": 1, "flag": 1, "wt": 1, "cell": 1, "after": 1, "arsenite30min": 1, "main": 1, "sam2profilesgenom": [0, 1], "py": [0, 1], "sam2profil": [0, 1], "support": 1, "csv2pickl": 1, "mergesalmon": 1, "modul": 1, "samgener": 1, "samtranscript": 1, "samgenom": 1, "assai": 1, "go_enrich": 1, "nascent": 1, "plot": 1, "profil": [1, 2], "secondari": 1, "index": [0, 1], "search": [0, 1], "page": 1, "usag": 2, "gener": [0, 2], "genom": [0, 2], "sam": [0, 2], "option": [0, 2], "argument": 2, "h": [0, 2], "help": 2, "show": 2, "messag": 2, "exit": 2, "input": [0, 2], "f": [0, 2], "default": [0, 2], "none": [0, 2], "u": 2, "read": [0, 2], "3end": [0, 2], "5end": [0, 2], "3": [0, 2], "end": [0, 2], "n": [0, 2], "save": [0, 2], "code": [0, 2], "polya": [0, 2], "onli": [0, 2], "fals": [0, 2], "r": 2, "all": [0, 2], "csv": [0, 2], "output": [0, 2], "pickl": [0, 2], "c": [0, 2], "toclear": [0, 2], "string": [0, 2], "sign": 2, "clear": 2, "_comp_flexbar_staralign": 2, "out": [0, 2], "chunk": [0, 2], "divid": [0, 2], "list": [0, 2], "gene": [0, 2], "given": [0, 2], "size": [0, 2], "0": [0, 2], "transcript": [0, 2], "l": [0, 2], "d": [0, 2], "details_fil": 2, "transcriptom": 2, "detail": [0, 2], "home": 2, "tollervei": 2, "covid19": 2, "databas": 2, "transcriptome_details_biomart_trna_rrna_updatejan2021": 2, "tab": [0, 2], "del": 2, "addit": [0, 2], "delet": [0, 2], "p": [0, 2], "datafram": [0, 2], "expand": [0, 2], "For": 2, "posit": [0, 2], "valu": [0, 2], "each": [0, 2], "side": [0, 2], "5": [0, 2], "give": 2, "10": [0, 2], "nt": [0, 2], "long": [0, 2], "gzip": [0, 2], "compress": 2, "merg": [0, 2], "salmon": 2, "quantif": 2, "one": [0, 2], "found": [0, 2], "directori": [0, 2], "o": [0, 2], "column": [0, 2], "tpm": 2, "numread": [0, 2], "filter": [0, 2], "add": [0, 2], "ad": [0, 2], "foldingfrombigwig": 0, "gene_nam": 0, "gtf": 0, "bwfwd": 0, "bwrev": 0, "rang": 0, "offset": 0, "15": 0, "fold": 0, "dg65nt": 0, "30c": 0, "sourc": 0, "pull": 0, "inform": 0, "bigwig": 0, "paramet": 0, "str": 0, "pycrac": 0, "gtf2": 0, "object": 0, "load": 0, "dict": 0, "pybigwig": 0, "int": 0, "flank": 0, "return": 0, "calculatefdr": 0, "seri": 0, "dtype": 0, "float64": 0, "iter": 0, "100": 0, "target_fdr": 0, "05": 0, "calcul": 0, "discoveri": 0, "rate": 0, "fdr": 0, "dataset": 0, "float": 0, "detault": 0, "compare1toref": 0, "ref": 0, "mm": 0, "heatmap": 0, "rel": 0, "take": 0, "compar": 0, "refer": 0, "path": 0, "min": 0, "max": 0, "qq": 0, "q1": 0, "q3": 0, "boolean": 0, "reference_above_experi": 0, "minimum": 0, "etc": 0, "rae_min": 0, "rae_max": 0, "ear_min": 0, "ear_max": 0, "true": 0, "differ": 0, "recalcul": 0, "accord": 0, "peak": 0, "warn": 0, "neg": 0, "1": 0, "higher": 0, "than": 0, "comparemoretoref": 0, "empti": 0, "filter_df": 0, "dictbigwig": 0, "strand": 0, "preload": 0, "memori": 0, "tool": 0, "findpeak": 0, "s1": 0, "window": 0, "win_typ": 0, "blackman": 0, "order": 0, "20": 0, "find": 0, "local": 0, "extrema": 0, "scipi": 0, "argrelextrema": 0, "To": 0, "smooth": 0, "befor": 0, "call": 0, "type": 0, "minim": 0, "space": 0, "between": 0, "findtrough": 0, "minima": 0, "trough": 0, "genefrombigwig": 0, "tostrip": 0, "coverag": 0, "One": 0, "strip": 0, "ntotal": 0, "df": 0, "class": 0, "panda": 0, "core": 0, "frame": 0, "drop": 0, "normal": 0, "fraction": 0, "total": 0, "nucleotid": 0, "parseconcatfil": 0, "rpm": 0, "1000": 0, "pars": 0, "concat": 0, "tu": 0, "substitut": 0, "kei": 0, "preprocess": 0, "input_df": 0, "let_in": 0, "let_out": 0, "wont_find_this_str": 0, "stat": 0, "combin": 0, "filterexp": 0, "expstat": 0, "choosen": 0, "appli": 0, "word": 0, "character": 0, "disqualifi": 0, "mean": 0, "median": 0, "quartil": 0, "more": 0, "2": 0, "pseudocount": 0, "01": 0, "save_csv": 0, "data_ref": 0, "containinig": 0, "experiment": 0, "stripbigwignam": 0, "_rev": 0, "bw": 0, "_fwd": 0, "form": 0, "filenam": 0, "uniqu": 0, "countdelet": 0, "tupl": 0, "cigarstr": 0, "map": 0, "count": 0, "onc": 0, "longer": 0, "common": 0, "expans": 0, "first": 0, "cigar": 0, "number": 0, "np": 0, "arrai": 0, "400": 0, "3s15m1d9m2": 0, "415": 0, "412": 0, "413": 0, "414": 0, "416": 0, "417": 0, "418": 0, "countmiddl": 0, "middl": 0, "409": 0, "410": 0, "411": 0, "countread": 0, "401": 0, "402": 0, "403": 0, "404": 0, "405": 0, "406": 0, "407": 0, "408": 0, "419": 0, "420": 0, "421": 0, "422": 0, "423": 0, "424": 0, "groupcigar": 0, "cigar_str": 0, "split": 0, "3s44m1s1h": 0, "": 0, "44": 0, "m": 0, "noncoded2profil": 0, "df_input": 0, "df_detail": 0, "turn": 0, "parsenoncod": 0, "chromosom": 0, "length": 0, "minlen": 0, "structur": 0, "dictionari": 0, "chromosm": 0, "chri": 0, "chrii": 0, "keep": 0, "40": 0, "aaa": 0, "35": 0, "aacaa": 0, "chr": 0, "nan": 0, "selectpolya": 0, "select": 0, "A": 0, "content": 0, "abov": 0, "75": 0, "modifi": 0, "stripcigar": 0, "match": 0, "to_strip": 0, "remov": 0, "mark": 0, "stripsubstitut": 0, "substutiotn": 0, "both": 0, "tostripcigarf": 0, "soft": 0, "tostripcigarthre": 0, "without": 0, "align": 0, "reads2profil": 0, "dirpath": 0, "transform": 0, "transctipt": 0, "test": 0, "plu": 0, "reads2profiledelet": 0, "genelist": 0, "handl": 0, "execut": 0, "wrap": 0, "element": 0, "bool": 0, "format": 0, "transcript2profil": 0, "work": 0, "entir": 0, "transcript2profiledelet": 0, "Not": 0, "minu": 0, "chromosome2profile3end": 0, "fwd": 0, "tripl": 0, "rev": 0, "noncod": 0, "tttgcgcagtcgtgcggggcgcagcgccc": 0, "50": 0, "34": 0, "36": 0, "cc": 0, "39": 0, "41": 0, "parsehead": 0, "reads2genom": 0, "sam2genom": 0, "output_df_fwd": 0, "output_df_rev": 0, "log": 0, "reads2genome3end": 0, "sam2genome3end": 0, "If": 0, "noncoded_fwd": 0, "noncoded_rev": 0, "reads2genome5end": 0, "sam2genome5end": 0, "noncoded_pa": 0, "noncoded_raw": 0, "bulkinputidt": 0, "tranform": 0, "colum": 0, "templat": 0, "prepar": 0, "tabl": 0, "bulk": 0, "discard": 0, "oligo": 0, "200": 0, "extrud": 0, "buri": 0, "13": 0, "within": 0, "rnap": 0, "findprim": 0, "queri": 0, "contain": 0, "start": 0, "stop": 0, "nontemplatedna": 0, "primer": 0, "dna": 0, "sequenceconstrain": 0, "stall": 0, "rnaprim": 0, "constrain": 0, "5prime": 0, "singl": 0, "letter": 0, "algorithm": 0, "structurefil": 0, "current": 0, "templatedna": 0, "overhang5end": 0, "ovethang": 0, "testscaffold": 0, "rna_prim": 0, "print": 0, "scaffold": 0, "overhang": 0, "shift": 0, "toorder": 0, "gucuguuuugugg": 0, "stallful": 0, "afterstal": 0, "tgatcggtac": 0, "tga": 0, "aggccgaaa": 0, "bulkinput": 0, "lengthmax": 0, "transfer": 0, "get_enrich": 0, "query_gen": 0, "organ": 0, "use_reference_set": 0, "ref_gen": 0, "ref_organ": 0, "go_dataset": 0, "biological_process": 0, "test_typ": 0, "fisher": 0, "correct": 0, "go": 0, "term": 0, "enrich": 0, "panther": 0, "api": 0, "identifi": 0, "g": 0, "id": 0, "taxid": 0, "speci": 0, "9606": 0, "sapien": 0, "custom": 0, "set": 0, "rerenc": 0, "background": 0, "ref_speci": 0, "need": 0, "specifi": 0, "whole": 0, "when": 0, "pass": 0, "must": 0, "also": 0, "requir": 0, "which": 0, "annot": 0, "molecular_funct": 0, "tatist": 0, "avail": 0, "binomi": 0, "multipl": 0, "bonferroni": 0, "unfilt": 0, "result": 0, "addclust": 0, "assign": 0, "cluster": 0, "kmean": 0, "param": 0, "bashcommand": 0, "command": 0, "bash": 0, "subprocess": 0, "calgc": 0, "calfor": 0, "gc": 0, "cleannam": 0, "additional_tag": 0, "clean": 0, "problem": 0, "exist": 0, "where": 0, "new": 0, "define_experi": 0, "paths_in": 0, "whole_nam": 0, "_hittable_read": 0, "txt": 0, "extract": 0, "them": 0, "As": 0, "a_b_c": 0, "expnamepars": 0, "b_d_e_p": 0, "recogn": 0, "ab123456": 0, "date": 0, "by4741": 0, "htp": 0, "bait": 0, "tag": 0, "root": 0, "other": 0, "defoult": 0, "b": 0, "prefix": 0, "reorder": 0, "verbos": 0, "exp": 0, "groupcracsampl": 0, "todrop": 0, "follow": 0, "featur": 0, "sampl": 0, "samplerep": 0, "choos": 0, "qualifi": 0, "reject": 0, "group": 0, "indexord": 0, "lettercont": 0, "list_paths_in_current_dir": 0, "suffix": 0, "stdin": 0, "indic": 0, "standard": 0, "instead": 0, "log2": 0, "parsecracnam": 0, "_": 0, "quantilecategori": 0, "q": 0, "4": 0, "quantil": 0, "base": 0, "discret": 0, "qcut": 0, "decil": 0, "randomdnaal": 0, "cgta": 0, "possibl": 0, "random": 0, "randomdnasingl": 0, "readsalmon": 0, "nameelem": 0, "toload": 0, "toadd": 0, "overwrit": 0, "elem": 0, "equal": 0, "pd": 0, "read_htseq_output": 0, "htseq": 0, "append": 0, "allow": 0, "dure": 0, "read_starstat": 0, "read_featurecount": 0, "present": 0, "read_list": 0, "filepath": 0, "row": 0, "becom": 0, "item": 0, "read_tabfil": 0, "reverse_compl": 0, "revers": 0, "complement": 0, "reverse_complement_dna": 0, "reverse_complement_rna": 0, "rollinggc": 0, "boxcar": 0, "center": 0, "runpca": 0, "n_compon": 0, "pca": 0, "analysi": 0, "re": 0, "consist": 0, "pc": 0, "timestamp": 0, "timestamprandomint": 0, "tempdir": 0, "rnafold": 0, "savedata": 0, "temp": 0, "dg": 0, "viennarna": 0, "rnainvert": 0, "quick": 0, "fmp": 0, "rnainvertstal": 0, "unafold": 0, "bashfold": 0, "una": 0, "hybrid": 0, "rnahybrid": 0, "bashhybrid": 0, "hybri": 0, "analyseviennamarkgc": 0, "vienna": 0, "leav": 0, "stem": 0, "extendingwindow": 0, "30": 0, "7": 0, "doe": 0, "ani": 0, "shorter": 0, "even": 0, "deg": 0, "slidingwindow": 0, "convent": 0, "foldnascentelem": 0, "veri": 0, "handleinput": 0, "keepnam": 0, "join2d": 0, "markvienna": 0, "merge2d": 0, "name2index": 0, "preparenasc": 0, "nascentelem": 0, "describ": 0, "multistem": 0, "nascentelemsdatafram": 0, "nascentfold": 0, "locat": 0, "last": 0, "parsefoldingnam": 0, "short": 0, "selectfoldedn": 0, "pattern": 0, "most": 0, "energi": 0, "slice": 0, "slide": 0, "80": 0, "clusterclustermap": 0, "clustermap": 0, "hplotstarstat": 0, "dpi": 0, "150": 0, "hplotstarstats_chimer": 0, "hplotstarstats_map": 0, "hplotstarstats_mistmatch": 0, "hplotstarstats_readlen": 0, "hplotstarstats_read": 0, "plotcumulativepeak": 0, "df2": 0, "local_po": 0, "titl": 0, "figsiz": 0, "color1": 0, "green": 0, "color2": 0, "magenta": 0, "lc": 0, "red": 0, "metaplot": 0, "plotpca": 0, "pclimit": 0, "pcval": 0, "import": [], "example_df": 0, "pca_result": [], "plotstarstat": 0, "plotstarstats_chimer": 0, "plotstarstats_map": 0, "plotstarstats_mistmatch": 0, "plotstarstats_readlen": 0, "plotstarstats_read": 0, "plot_as_box_plot": 0, "ylim": 0, "color": 0, "h_line": 0, "figur": 0, "similar": 0, "box": 0, "std": 0, "oi": 0, "ax": 0, "lim": 0, "horizont": 0, "line": 0, "plot_diff": 0, "label": 0, "plot_median": 0, "plot_rang": 0, "plot_heatmap": 0, "rdn37": 0, "vmin": 0, "vmax": 0, "heat": 0, "plot_to_compar": 0, "black": 0, "ref_label": 0, "300": 0, "lstem": 0, "coordin": 0, "inf": 0, "rstem": 0, "checkvienna": 0, "valid": 0, "integr": 0, "loopstem": 0, "loopslist": 0, "testprint": 0, "postion": 0, "hairpin": 0, "multiloop": 0, "warninig": 0, "loop": 0, "outsid": 0, "8": 0, "substructur": 0, "sub": 0, "linker": 0, "vienna2format": 0, "stemslist": 0, "multistemslist": 0, "convert": 0, "defaul": 0, "noncoded2profile1": 0, "parsenoncodedlist": 0, "lit": 0, "savebigwig": 0, "bw_name": 0, "chrom": 0, "_type_": 0, "_description_": 0, "selectend": 0, "wrapper": 0, "selectsortpath": 0, "sort": 0, "chromsom": 0, "lognam": 0}, "objects": {"TTools": [[0, 0, 0, "-", "SAMgeneral"], [0, 0, 0, "-", "SAMgenome"], [0, 0, 0, "-", "SAMtranscripts"], [0, 0, 0, "-", "assays"], [0, 0, 0, "-", "go_enrichment"], [0, 0, 0, "-", "methods"], [0, 0, 0, "-", "nascent"], [0, 0, 0, "-", "plotting"], [0, 0, 0, "-", "profiles"], [0, 0, 0, "-", "secondary"]], "TTools.SAMgeneral": [[0, 1, 1, "", "countDeletion"], [0, 1, 1, "", "countMiddle"], [0, 1, 1, "", "countRead"], [0, 1, 1, "", "groupCIGAR"], [0, 1, 1, "", "noncoded2profile"], [0, 1, 1, "", "noncoded2profile1"], [0, 1, 1, "", "parseNoncoded"], [0, 1, 1, "", "parseNoncodedList"], [0, 1, 1, "", "saveBigWig"], [0, 1, 1, "", "selectEnds"], [0, 1, 1, "", "selectPolyA"], [0, 1, 1, "", "selectSortPaths"], [0, 1, 1, "", "stripCIGAR"], [0, 1, 1, "", "stripSubstitutions"], [0, 1, 1, "", "tostripCIGARfive"], [0, 1, 1, "", "tostripCIGARthree"]], "TTools.SAMgenome": [[0, 1, 1, "", "chromosome2profile3end"], [0, 1, 1, "", "parseHeader"], [0, 1, 1, "", "reads2genome"], [0, 1, 1, "", "reads2genome3end"], [0, 1, 1, "", "reads2genome5end"], [0, 1, 1, "", "sam2genome"]], "TTools.SAMtranscripts": [[0, 1, 1, "", "reads2profile"], [0, 1, 1, "", "reads2profileDeletions"], [0, 1, 1, "", "sam2profiles"], [0, 1, 1, "", "transcript2profile"], [0, 1, 1, "", "transcript2profileDeletions"]], "TTools.assays": [[0, 1, 1, "", "bulkInputIDT"], [0, 1, 1, "", "extruded"], [0, 1, 1, "", "findPrimer"], [0, 1, 1, "", "nonTemplateDNA"], [0, 1, 1, "", "sequenceConstrain"], [0, 1, 1, "", "stalled"], [0, 1, 1, "", "structureFile"], [0, 1, 1, "", "templateDNA"], [0, 1, 1, "", "testScaffold"], [0, 1, 1, "", "toOrder"]], "TTools.go_enrichment": [[0, 1, 1, "", "get_enrichment"]], "TTools.methods": [[0, 1, 1, "", "addCluster"], [0, 1, 1, "", "bashCommand"], [0, 1, 1, "", "calGC"], [0, 1, 1, "", "cleanNames"], [0, 1, 1, "", "define_experiments"], [0, 1, 1, "", "expNameParser"], [0, 1, 1, "", "expStats"], [0, 1, 1, "", "filterExp"], [0, 1, 1, "", "groupCRACsamples"], [0, 1, 1, "", "indexOrder"], [0, 1, 1, "", "letterContent"], [0, 1, 1, "", "list_paths_in_current_dir"], [0, 1, 1, "", "normalize"], [0, 1, 1, "", "parseCRACname"], [0, 1, 1, "", "quantileCategory"], [0, 1, 1, "", "randomDNAall"], [0, 1, 1, "", "randomDNAsingle"], [0, 1, 1, "", "readSalmon"], [0, 1, 1, "", "read_HTSeq_output"], [0, 1, 1, "", "read_STARstats"], [0, 1, 1, "", "read_featureCount"], [0, 1, 1, "", "read_list"], [0, 1, 1, "", "read_tabFile"], [0, 1, 1, "", "reverse_complement"], [0, 1, 1, "", "reverse_complement_DNA"], [0, 1, 1, "", "reverse_complement_RNA"], [0, 1, 1, "", "rollingGC"], [0, 1, 1, "", "runPCA"], [0, 1, 1, "", "timestamp"], [0, 1, 1, "", "timestampRandomInt"]], "TTools.nascent": [[0, 2, 1, "", "Fold"], [0, 2, 1, "", "Hybrid"], [0, 1, 1, "", "analyseViennamarkGC"], [0, 1, 1, "", "extendingWindow"], [0, 1, 1, "", "foldNascentElem"], [0, 1, 1, "", "handleInput"], [0, 1, 1, "", "join2d"], [0, 1, 1, "", "markVienna"], [0, 1, 1, "", "merge2d"], [0, 1, 1, "", "name2index"], [0, 1, 1, "", "nascentElems"], [0, 1, 1, "", "nascentElemsDataFrame"], [0, 1, 1, "", "nascentFolding"], [0, 1, 1, "", "parseFoldingName"], [0, 1, 1, "", "prepareNascent"], [0, 1, 1, "", "selectFoldedN"], [0, 1, 1, "", "slidingWindow"]], "TTools.nascent.Fold": [[0, 3, 1, "", "RNAfold"], [0, 3, 1, "", "RNAinvert"], [0, 3, 1, "", "RNAinvertStall"], [0, 3, 1, "", "UNAfold"], [0, 3, 1, "", "bashFolding"]], "TTools.nascent.Hybrid": [[0, 3, 1, "", "RNAhybrid"], [0, 3, 1, "", "bashHybrid"]], "TTools.plotting": [[0, 1, 1, "", "clusterClusterMap"], [0, 1, 1, "", "hplotSTARstats"], [0, 1, 1, "", "hplotSTARstats_chimeric"], [0, 1, 1, "", "hplotSTARstats_mapping"], [0, 1, 1, "", "hplotSTARstats_mistmatches"], [0, 1, 1, "", "hplotSTARstats_readLen"], [0, 1, 1, "", "hplotSTARstats_reads"], [0, 1, 1, "", "plotCumulativePeaks"], [0, 1, 1, "", "plotPCA"], [0, 1, 1, "", "plotSTARstats"], [0, 1, 1, "", "plotSTARstats_chimeric"], [0, 1, 1, "", "plotSTARstats_mapping"], [0, 1, 1, "", "plotSTARstats_mistmatches"], [0, 1, 1, "", "plotSTARstats_readLen"], [0, 1, 1, "", "plotSTARstats_reads"], [0, 1, 1, "", "plot_as_box_plot"], [0, 1, 1, "", "plot_diff"], [0, 1, 1, "", "plot_heatmap"], [0, 1, 1, "", "plot_to_compare"]], "TTools.profiles": [[0, 1, 1, "", "FoldingFromBigWig"], [0, 1, 1, "", "calculateFDR"], [0, 1, 1, "", "compare1toRef"], [0, 1, 1, "", "compareMoretoRef"], [0, 1, 1, "", "dictBigWig"], [0, 1, 1, "", "findPeaks"], [0, 1, 1, "", "findTroughs"], [0, 1, 1, "", "geneFromBigWig"], [0, 1, 1, "", "ntotal"], [0, 1, 1, "", "parseConcatFile"], [0, 1, 1, "", "preprocess"], [0, 1, 1, "", "pseudocounts"], [0, 1, 1, "", "save_csv"], [0, 1, 1, "", "stripBigWignames"]], "TTools.secondary": [[0, 1, 1, "", "Lstem"], [0, 1, 1, "", "Rstem"], [0, 1, 1, "", "checkVienna"], [0, 1, 1, "", "loopStems"], [0, 1, 1, "", "loops"], [0, 1, 1, "", "substructures"], [0, 1, 1, "", "test"], [0, 1, 1, "", "vienna2format"]]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:method"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "method", "Python method"]}, "titleterms": {"modul": 0, "ttool": [0, 1], "samgener": 0, "samtranscript": 0, "samgenom": 0, "assai": 0, "go_enrich": 0, "method": 0, "nascent": 0, "plot": 0, "profil": 0, "secondari": 0, "indic": 1, "tabl": 1, "script": 2, "main": 2, "sam2profilesgenom": 2, "py": 2, "sam2profil": 2, "support": 2, "csv2pickl": 2, "mergesalmon": 2}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 6, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 56}})