<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Modules &mdash; TTools 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Scripts" href="scripts.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> TTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.SAMgeneral">TTools.SAMgeneral module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.SAMtranscripts">TTools.SAMtranscripts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.SAMgenome">TTools.SAMgenome module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.assays">TTools.assays module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.go_enrichment">TTools.go_enrichment module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.methods">TTools.methods module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.nascent">TTools.nascent module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.plotting">TTools.plotting module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.profiles">TTools.profiles module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-TTools.secondary">TTools.secondary module</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Modules</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/TTools.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="modules">
<h1>Modules<a class="headerlink" href="#modules" title="Permalink to this heading"></a></h1>
<section id="module-TTools.SAMgeneral">
<span id="ttools-samgeneral-module"></span><h2>TTools.SAMgeneral module<a class="headerlink" href="#module-TTools.SAMgeneral" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.countDeletion">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countDeletion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#countDeletion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.countDeletion" title="Permalink to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list of mapped deletions. Each deletion is counted once, but expand parameter can merge some longer deletions (common expansion).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nucleotides to expand for each side, defaults to 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countDeletion</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([415])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">countDeletion</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">),</span><span class="n">expand</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([412, 413, 414, 415, 416, 417, 418])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.countMiddle">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countMiddle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#countMiddle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.countMiddle" title="Permalink to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list with the middle of mapped read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – number of nucleotides to expand for each side, defaults to 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countMiddle</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([412])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">countMiddle</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">),</span><span class="n">expand</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([409, 410, 411, 412, 413, 414])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.countRead">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">countRead</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#countRead"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.countRead" title="Permalink to this definition"></a></dt>
<dd><p>Takes tuple (position,CIGARstring) and returns list of mapped positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>i</strong> (<em>tuple</em>) – tuple (first position of the read ,CIGAR string)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of mapped positions</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countRead</span><span class="p">((</span><span class="mi">400</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">))</span>
<span class="go">array([400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412,</span>
<span class="go">   413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.groupCIGAR">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">groupCIGAR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cigar_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#groupCIGAR"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.groupCIGAR" title="Permalink to this definition"></a></dt>
<dd><p>Split CIGAR string to list of tuples</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cigar_string</strong> (<em>str</em>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples <code class="docutils literal notranslate"><span class="pre">[(</span> <span class="pre">),()]</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">groupCIGAR</span><span class="p">(</span><span class="s2">&quot;3S44M1S1H&quot;</span><span class="p">)</span>
<span class="go">[(&#39;3&#39;, &#39;S&#39;), (&#39;44&#39;, &#39;M&#39;), (&#39;1&#39;, &#39;S&#39;), (&#39;1&#39;, &#39;H&#39;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.noncoded2profile">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">noncoded2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_input=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#noncoded2profile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.noncoded2profile" title="Permalink to this definition"></a></dt>
<dd><p>Turns non-coded ends into profile</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_input</strong> (<em>DataFrame</em>) – output of parseNoncoded function</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – chromosome lengths</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with profiles</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.parseNoncoded">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">parseNoncoded</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minLen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#parseNoncoded"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.parseNoncoded" title="Permalink to this definition"></a></dt>
<dd><p>Parse dict with non-coded ends and returns structured DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>d</strong> (<em>dict</em>) – dictionary with list of tuples for each chromosme <code class="docutils literal notranslate"><span class="pre">{&quot;chrI&quot;</span> <span class="pre">:</span> <span class="pre">[],</span> <span class="pre">&quot;chrII&quot;</span> <span class="pre">:</span> <span class="pre">[]}</span></code>, defaults to dict()</p></li>
<li><p><strong>minLen</strong> (<em>int</em><em>, </em><em>optional</em>) – minimal length for non-coded end to keep, defaults to 3</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with parsed non-coded ends</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parseNoncoded</span><span class="p">({</span><span class="s2">&quot;chrI&quot;</span><span class="p">:[(</span><span class="mi">40</span><span class="p">,</span> <span class="s1">&#39;AAA&#39;</span><span class="p">),</span> <span class="p">(</span><span class="mi">35</span><span class="p">,</span> <span class="s1">&#39;AACAA&#39;</span><span class="p">)]})</span>
<span class="go">   index  AAA  AACAA   chr</span>
<span class="go">0     40  1.0    NaN  chrI</span>
<span class="go">1     35  NaN    1.0  chrI</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.selectPolyA">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">selectPolyA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#selectPolyA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.selectPolyA" title="Permalink to this definition"></a></dt>
<dd><p>Select only polyA non-coded ends containinig <code class="docutils literal notranslate"><span class="pre">&quot;AAA&quot;</span></code> and “A”-content above 75%</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> (<em>DataFrame</em>) – output of parseNoncoded function</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>modified DataFrame</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.stripCIGAR">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">stripCIGAR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_strip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'H'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#stripCIGAR"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.stripCIGAR" title="Permalink to this definition"></a></dt>
<dd><p>Removes H from output of groupCIGAR</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p></li>
<li><p><strong>to_strip</strong> (<em>str</em><em>, </em><em>optional</em>) – CIGAR mark to be stripped, defaults to “H”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>modified list of tuples</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.stripSubstitutions">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">stripSubstitutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#stripSubstitutions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.stripSubstitutions" title="Permalink to this definition"></a></dt>
<dd><p>Strip substutiotns on both ends</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.tostripCIGARfive">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">tostripCIGARfive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#tostripCIGARfive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.tostripCIGARfive" title="Permalink to this definition"></a></dt>
<dd><p>Calculates length of soft-clipped nucleotides at the 5’ end of the read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>number of substituted nucleotides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tostripCIGARfive</span><span class="p">([(</span><span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;44&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">)])</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgeneral.tostripCIGARthree">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgeneral.</span></span><span class="sig-name descname"><span class="pre">tostripCIGARthree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgeneral.html#tostripCIGARthree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgeneral.tostripCIGARthree" title="Permalink to this definition"></a></dt>
<dd><p>Nucleotides without alignment at the 3’ end of the read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>match</strong> (<em>list</em>) – output of groupCIGAR, defaults to []</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>number of substituted nucleotides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tostripCIGARthree</span><span class="p">([(</span><span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;44&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">)])</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-TTools.SAMtranscripts">
<span id="ttools-samtranscripts-module"></span><h2>TTools.SAMtranscripts module<a class="headerlink" href="#module-TTools.SAMtranscripts" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMtranscripts.reads2profile">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">reads2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMtranscripts.html#reads2profile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMtranscripts.reads2profile" title="Permalink to this definition"></a></dt>
<dd><p>Takes list of aligned reads and transform to profile for each transctipt. Tested only for PLUS strand.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMtranscripts.reads2profileDeletions">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">reads2profileDeletions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand=5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMtranscripts.html#reads2profileDeletions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMtranscripts.reads2profileDeletions" title="Permalink to this definition"></a></dt>
<dd><p>Tested only for PLUS strand.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMtranscripts.sam2profiles">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">sam2profiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneList=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deletions=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks=0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMtranscripts.html#sam2profiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMtranscripts.sam2profiles" title="Permalink to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles. Executed using wrapping script SAM2profiles.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – </p></li>
<li><p><strong>path</strong> (<em>str</em>) – </p></li>
<li><p><strong>geneList</strong> (<em>list</em>) – list of transcript to be selected</p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – element of filename to be removed, defaults to ‘’</p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – Details of transcripts</p></li>
<li><p><strong>deletions</strong> (<em>bool</em><em>, </em><em>optional</em>) – Generate profile of deletions, defaults to False</p></li>
<li><p><strong>expand</strong> (<em>int</em><em>, </em><em>optional</em>) – Expand deletions, defaults to 5</p></li>
<li><p><strong>pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – save output in pickle format, defaults to False</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Read SAM file in chunks, defaults to 0</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMtranscripts.transcript2profile">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">transcript2profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMtranscripts.html#transcript2profile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMtranscripts.transcript2profile" title="Permalink to this definition"></a></dt>
<dd><p>Takes list of tuples position,CIGARstring) and generates profile. Works with entire reads, for both strands.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMtranscripts.transcript2profileDeletions">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMtranscripts.</span></span><span class="sig-name descname"><span class="pre">transcript2profileDeletions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMtranscripts.html#transcript2profileDeletions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMtranscripts.transcript2profileDeletions" title="Permalink to this definition"></a></dt>
<dd><p>Takes list of tuples position,CIGARstring) and generates profile. Not tested for MINUS strand.</p>
</dd></dl>

</section>
<section id="module-TTools.SAMgenome">
<span id="ttools-samgenome-module"></span><h2>TTools.SAMgenome module<a class="headerlink" href="#module-TTools.SAMgenome" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.chromosome2profile3end">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">chromosome2profile3end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FWD'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#chromosome2profile3end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.chromosome2profile3end" title="Permalink to this definition"></a></dt>
<dd><p>Generates profile for the 3’ ends of reads and saves position of non-coded end</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>list</em>) – list of triple tuples (position, cigar_string, sequence), defaults to []</p></li>
<li><p><strong>length</strong> (<em>int</em>) – length of chromosome</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – <code class="docutils literal notranslate"><span class="pre">'FWD'</span></code> or <code class="docutils literal notranslate"><span class="pre">'REV'</span></code>, defaults to ‘FWD’</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>profile, noncoded</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.array, list of tuples</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="p">[(</span><span class="mi">10</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">,</span><span class="s2">&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;</span><span class="p">)],</span><span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="s2">&quot;FWD&quot;</span><span class="p">)</span>
<span class="go">(0     0.0</span>
<span class="go">1     0.0</span>
<span class="gp">...</span>
<span class="go">34    0.0</span>
<span class="go">35    1.0</span>
<span class="go">36    0.0</span>
<span class="gp">...</span>
<span class="go">50    0.0</span>
<span class="go">dtype: float64,</span>
<span class="go">[(35, &#39;CC&#39;)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome2profile3end</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="p">[(</span><span class="mi">40</span><span class="p">,</span><span class="s2">&quot;3S15M1D9M2S&quot;</span><span class="p">,</span><span class="s2">&quot;TTTGCGCAGTCGTGCGGGGCGCAGCGCCC&quot;</span><span class="p">)],</span><span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="s2">&quot;REV&quot;</span><span class="p">)</span>
<span class="go">(0     0.0</span>
<span class="go">1     0.0</span>
<span class="gp">...</span>
<span class="go">39    0.0</span>
<span class="go">40    1.0</span>
<span class="go">41    0.0</span>
<span class="gp">...</span>
<span class="go">50    0.0</span>
<span class="go">dtype: float64,</span>
<span class="go">[(40, &#39;AAA&#39;)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.parseHeader">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">parseHeader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#parseHeader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.parseHeader" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.reads2genome">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#reads2genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.reads2genome" title="Permalink to this definition"></a></dt>
<dd><p>Function used by sam2genome. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – </p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.reads2genome3end">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome3end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#reads2genome3end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.reads2genome3end" title="Permalink to this definition"></a></dt>
<dd><p>Function used by sam2genome3end. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – </p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
<li><p><strong>noncoded</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True then will parse and save non-coded ends, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log, noncoded_fwd, noncoded_rev</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list, DataFrame, DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.reads2genome5end">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">reads2genome5end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirPath=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_details=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#reads2genome5end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.reads2genome5end" title="Permalink to this definition"></a></dt>
<dd><p>Function used by sam2genome5end. Works for both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of experiment</p></li>
<li><p><strong>dirPath</strong> (<em>str</em>) – </p></li>
<li><p><strong>df_details</strong> (<em>DataFrame</em>) – lengths of chromosomes</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>output_df_fwd, output_df_rev, log</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>DataFrame, DataFrame, list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.sam2genome">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">sam2genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#sam2genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.sam2genome" title="Permalink to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles. Executed using wrapping script SAM2profilesGenomic.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – </p></li>
<li><p><strong>path</strong> (<em>str</em>) – </p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – element of filename to be removed, defaults to ‘’</p></li>
<li><p><strong>pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – save output in pickle format, defaults to False</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Read SAM file in chunks, defaults to 0</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.sam2genome3end">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">sam2genome3end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded_pA</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noncoded_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#sam2genome3end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.sam2genome3end" title="Permalink to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles for the 3’ end of reads. Executed using wrapping script SAM2profilesGenomic.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – </p></li>
<li><p><strong>path</strong> (<em>str</em>) – </p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – element of filename to be removed, defaults to ‘’</p></li>
<li><p><strong>pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – save output in pickle format, defaults to False</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Read SAM file in chunks, defaults to 0</p></li>
<li><p><strong>noncoded_pA</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save non-coded polyA ends, defaults to True</p></li>
<li><p><strong>noncoded_raw</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save all non-coded ends, defaults to False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.SAMgenome.sam2genome5end">
<span class="sig-prename descclassname"><span class="pre">TTools.SAMgenome.</span></span><span class="sig-name descname"><span class="pre">sam2genome5end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/SAMgenome.html#sam2genome5end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.SAMgenome.sam2genome5end" title="Permalink to this definition"></a></dt>
<dd><p>Function handling SAM files and generating profiles for the 3’ end of reads. Executed using wrapping script SAM2profilesGenomic.py.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – </p></li>
<li><p><strong>path</strong> (<em>str</em>) – </p></li>
<li><p><strong>toClear</strong> (<em>str</em><em>, </em><em>optional</em>) – element of filename to be removed, defaults to ‘’</p></li>
<li><p><strong>pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – save output in pickle format, defaults to False</p></li>
<li><p><strong>chunks</strong> (<em>int</em><em>, </em><em>optional</em>) – Read SAM file in chunks, defaults to 0</p></li>
<li><p><strong>noncoded_pA</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save non-coded polyA ends, defaults to True</p></li>
<li><p><strong>noncoded_raw</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save all non-coded ends, defaults to False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.assays">
<span id="ttools-assays-module"></span><h2>TTools.assays module<a class="headerlink" href="#module-TTools.assays" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.bulkInputIDT">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">bulkInputIDT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#bulkInputIDT"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.bulkInputIDT" title="Permalink to this definition"></a></dt>
<dd><p>Tranforms dataframe with colums “template” and “non-template” and prepares table to be used as bulk input
discards oligos longer than 200 nt.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – DataFrame()</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with sequences to order</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.extruded">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">extruded</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buried</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#extruded"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.extruded" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – str()</p></li>
<li><p><strong>buried</strong> – int() length of sequence buried within RNAP</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Extruded sequence, str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.findPrimer">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">findPrimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#findPrimer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.findPrimer" title="Permalink to this definition"></a></dt>
<dd><p>Find query sequence within given sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – str() containing sequence</p></li>
<li><p><strong>query</strong> – str() within searched sequences</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(start,stop)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.nonTemplateDNA">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">nonTemplateDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#nonTemplateDNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.nonTemplateDNA" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – str() sequence</p></li>
<li><p><strong>primer</strong> – str() sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DNA sequence of non-template strand, str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.sequenceConstrain">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">sequenceConstrain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#sequenceConstrain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.sequenceConstrain" title="Permalink to this definition"></a></dt>
<dd><p>Returns sequence constrained with the 5prime RNA primer</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> – str()</p></li>
<li><p><strong>stall</strong> – str() single letter</p></li>
<li><p><strong>RNAprimer</strong> – str()</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str() with sequence constrains for folding algorithm</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.stalled">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">stalled</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#stalled"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.stalled" title="Permalink to this definition"></a></dt>
<dd><p>Finds and returns stalled sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – str()</p></li>
<li><p><strong>stall</strong> – str() default “AAA”</p></li>
<li><p><strong>primer</strong> – str()</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>stalled sequence, str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.structureFile">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">structureFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#structureFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.structureFile" title="Permalink to this definition"></a></dt>
<dd><p>Saves structure file in current directory</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> – str()</p></li>
<li><p><strong>stall</strong> – str()</p></li>
<li><p><strong>RNAprimer</strong> – str()</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.templateDNA">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">templateDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#templateDNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.templateDNA" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – str() sequence of RNA</p></li>
<li><p><strong>overhang5end</strong> – str() sequence of the 5’end DNA ovethang</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DNA sequence of template strand, str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.testScaffold">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">testScaffold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNA_primer=''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#testScaffold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.testScaffold" title="Permalink to this definition"></a></dt>
<dd><p>Prints scaffold to test before ordering</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame()</p></li>
<li><p><strong>overhang5end</strong> – str() overhang to shift sequences</p></li>
<li><p><strong>RNA_primer</strong> – str() sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.assays.toOrder">
<span class="sig-prename descclassname"><span class="pre">TTools.assays.</span></span><span class="sig-name descname"><span class="pre">toOrder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buried='GUCUGUUUUGUGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stallFull='AAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">afterStall='TGATCGGTAC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang5end='TGA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNA_primer='AGGCCGAAA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bulkInput=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lengthMax=200</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/assays.html#toOrder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.assays.toOrder" title="Permalink to this definition"></a></dt>
<dd><p>Transfers folding function to DNA sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame()</p></li>
<li><p><strong>buried</strong> – str() sequence</p></li>
<li><p><strong>stallFull</strong> – str() sequence</p></li>
<li><p><strong>afterStall</strong> – str() sequence</p></li>
<li><p><strong>overhang5end</strong> – str() sequence</p></li>
<li><p><strong>RNA_primer</strong> – str() sequence</p></li>
<li><p><strong>bulkInput</strong> – boolean() default True</p></li>
<li><p><strong>test</strong> – boolean() default True</p></li>
<li><p><strong>lengthMax</strong> – int() default 200</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame of sequences to order</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.go_enrichment">
<span id="ttools-go-enrichment-module"></span><h2>TTools.go_enrichment module<a class="headerlink" href="#module-TTools.go_enrichment" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.go_enrichment.get_enrichment">
<span class="sig-prename descclassname"><span class="pre">TTools.go_enrichment.</span></span><span class="sig-name descname"><span class="pre">get_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_reference_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biological_process'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FISHER'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FDR'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/go_enrichment.html#get_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.go_enrichment.get_enrichment" title="Permalink to this definition"></a></dt>
<dd><p>Run a GO term enrichment test using PANTHER API</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_genes</strong> (<em>list</em>) – List of sequence identifiers of queried genes (e.g. transcript ids, gene ids)</p></li>
<li><p><strong>organism</strong> (<em>str</em>) – Taxid of query species (e.g. “9606” for H. sapiens)</p></li>
<li><p><strong>use_reference_set</strong> (<em>bool</em>) – Use a custom set of rerence (background) genes? Default False. If True, ref_genes and ref_species need to be specifed.</p></li>
<li><p><strong>ref_genes</strong> (<em>list</em>) – <em>optional</em> list of reference genes. Specifying None (default) will use the whole genome of species specified in organism. When passing a list, ref_organism taxid must also be provided.</p></li>
<li><p><strong>ref_species</strong> (<em>str</em>) – Taxid of reference species, required when ref_genes is not None</p></li>
<li><p><strong>go_dataset</strong> (<em>str</em>) – Which annotation dataset to query, “biological_process” or “molecular_function”</p></li>
<li><p><strong>test_type</strong> (<em>str</em>) – Which tatistical test to use. Available: “FISHER” (default), “BINOMIAL”</p></li>
<li><p><strong>correction</strong> (<em>str</em>) – Which multiple testing correction method to use. Available: “FDR” (default), “BONFERRONI”, “NONE”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unfiltered DataFrame of results.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame:</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.methods">
<span id="ttools-methods-module"></span><h2>TTools.methods module<a class="headerlink" href="#module-TTools.methods" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.addCluster">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">addCluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n=10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#addCluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.addCluster" title="Permalink to this definition"></a></dt>
<dd><p>Assigns n clusters to the data using KMeans algorithm
:param df: DataFrame
:param n: no. of clusters, int
:return:</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.bashCommand">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">bashCommand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bashCommand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#bashCommand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.bashCommand" title="Permalink to this definition"></a></dt>
<dd><p>Run command in bash using subprocess.call()</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.calGC">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">calGC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[],</span> <span class="pre">calFor=['G',</span> <span class="pre">'C']</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#calGC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.calGC" title="Permalink to this definition"></a></dt>
<dd><p>Returns GC content in a given string - uses [‘nucleotide’] column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dataset</strong> – DataFrame() with “nucleotide” column</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>fraction of GC content, float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.cleanNames">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">cleanNames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_tags=[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#cleanNames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.cleanNames" title="Permalink to this definition"></a></dt>
<dd><p>Cleans some problems with names if exist</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame() where names of columns are name of experiments</p></li>
<li><p><strong>additional_tags</strong> – list()</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame() with new names</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.define_experiments">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">define_experiments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">paths_in</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">whole_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_hittable_reads.txt'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#define_experiments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.define_experiments" title="Permalink to this definition"></a></dt>
<dd><p>Parse file names and extract experiment name from them</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>paths_in</strong> – str()</p></li>
<li><p><strong>whole_name</strong> – boolean() default False. As defaults script takes first ‘a_b_c’</p></li>
<li><p><strong>strip</strong> – str() to strip from filename.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list() of experiment names, list() of paths.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.expNameParser">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">expNameParser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'b_d_e_p'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#expNameParser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.expNameParser" title="Permalink to this definition"></a></dt>
<dd><p>Function handles experiment name; recognizes AB123456 as experiment date; BY4741 or HTP or given string as bait protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – </p></li>
<li><p><strong>additional_tags</strong> – list of tags</p></li>
<li><p><strong>output</strong> – default ‘root’ ; print other elements when ‘all’</p></li>
<li><p><strong>order</strong> – defoult ‘b_d_e_p’ b-bait; d-details, e-experiment, p-prefix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of reordered name</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.expStats">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">expStats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#expStats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.expStats" title="Permalink to this definition"></a></dt>
<dd><p>Returns DataFrame with ‘mean’, ‘median’, ‘min’, ‘max’ and quartiles if more than 2 experiments</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_df</strong> – DataFrame</p></li>
<li><p><strong>smooth</strong> – boolean, if True apply smoothing window, default=True</p></li>
<li><p><strong>window</strong> – int, smoothing window, default 10</p></li>
<li><p><strong>win_type</strong> – str type of smoothing window, default “blackman”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.filterExp">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">filterExp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datasets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">let_in</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">let_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['wont_find_this_string']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#filterExp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.filterExp" title="Permalink to this definition"></a></dt>
<dd><p>Returns object with filtered columns/keys.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>datasets</strong> – DataFrame() or dict() with exp name as a key</p></li>
<li><p><strong>let_in</strong> – list() with elements of name to filter in</p></li>
<li><p><strong>let_out</strong> – list() with elements of name to filter out</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame() or dict()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.groupCRACsamples">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">groupCRACsamples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='protein'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toDrop=[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#groupCRACsamples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.groupCRACsamples" title="Permalink to this definition"></a></dt>
<dd><p>Parse CRAC names and annotates them using on of following features
[‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’, ‘sample’,’sampleRep’]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>use</strong> – str, choose from [‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’, ‘sample’,’sampleRep’], default = ‘protein’</p></li>
<li><p><strong>toDrop</strong> – list of word in CRAC name that will qualify the sample to rejection, default = []</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with added column [‘group’]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.indexOrder">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">indexOrder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_tags=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output='root'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order='b_d_e_p'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#indexOrder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.indexOrder" title="Permalink to this definition"></a></dt>
<dd><p>Apply expNameParser to whole DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame() where names of columns are name of experiments</p></li>
<li><p><strong>additional_tags</strong> – list()</p></li>
<li><p><strong>output</strong> – </p></li>
<li><p><strong>order</strong> – str() default ‘b_d_e_p’ b-bait; d-details, e-experiment, p-prefix</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame() with new names</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.letterContent">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">letterContent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'A'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#letterContent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.letterContent" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.list_paths_in_current_dir">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">list_paths_in_current_dir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stdin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#list_paths_in_current_dir"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.list_paths_in_current_dir" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>suffix</strong> – str() lists paths in current directory ending with an indicated suffix only</p></li>
<li><p><strong>stdin</strong> – boolean() if True read from standard input instead current directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list() of paths</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.normalize">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocounts=0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#normalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.normalize" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>log2</strong> – boolean, default=False</p></li>
<li><p><strong>pseudocounts</strong> – float, default=0.1</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.parseCRACname">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">parseCRACname</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=&lt;class</span> <span class="pre">'pandas.core.series.Series'&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#parseCRACname"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.parseCRACname" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Parse CRAC name into [‘expID’, ‘expDate’, ‘protein’, ‘condition1’, ‘condition2’, ‘condition3’] using this order.</dt><dd><p>“_” is used to split the name</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>s1</strong> – Series,</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.quantileCategory">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">quantileCategory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q=4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#quantileCategory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.quantileCategory" title="Permalink to this definition"></a></dt>
<dd><p>Quantile-based discretization function based on pandas.qcut function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> – Series()</p></li>
<li><p><strong>q</strong> – int() number of quantiles: 10 for deciles, 5 for quantiles, 4 for quartiles, etc., default q=4</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.randomDNAall">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">randomDNAall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CGTA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#randomDNAall"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.randomDNAall" title="Permalink to this definition"></a></dt>
<dd><p>Generates all possible random sequences of a given length</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> – int()</p></li>
<li><p><strong>letters</strong> – str() with letters that will be used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list() of str()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.randomDNAsingle">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">randomDNAsingle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CGTA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#randomDNAsingle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.randomDNAsingle" title="Permalink to this definition"></a></dt>
<dd><p>Random generator of nucleotide sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> – int()</p></li>
<li><p><strong>letters</strong> – str() with letters that will be used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.readSalmon">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">readSalmon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NumReads'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#readSalmon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.readSalmon" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nameElem</strong> – str, elem to load</p></li>
<li><p><strong>path</strong> – str</p></li>
<li><p><strong>toLoad</strong> – str, additional param for filtering, by default equal to nameElem</p></li>
<li><p><strong>toClear</strong> – str</p></li>
<li><p><strong>toAdd</strong> – str</p></li>
<li><p><strong>df</strong> – pd.DataFrame</p></li>
<li><p><strong>overwrite</strong> – boolean, default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.read_HTSeq_output">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">read_HTSeq_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'classes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#read_HTSeq_output"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.read_HTSeq_output" title="Permalink to this definition"></a></dt>
<dd><p>Reads multiple HTSeq tab files to one DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – str, path to directory with files</p></li>
<li><p><strong>toClear</strong> – str, will be removed from file name</p></li>
<li><p><strong>toAdd</strong> – str, to be added to file name</p></li>
<li><p><strong>df</strong> – DataFrame, to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> – boolean, allows for overwriting during appending, default = False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.read_STARstats">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">read_STARstats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#read_STARstats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.read_STARstats" title="Permalink to this definition"></a></dt>
<dd><p>Reads multiple HTSeq tab files to one DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – str, path to directory with files</p></li>
<li><p><strong>toClear</strong> – str, will be removed from file name</p></li>
<li><p><strong>toAdd</strong> – str, to be added to file name</p></li>
<li><p><strong>df</strong> – DataFrame, to be appended; default=None</p></li>
<li><p><strong>overwrite</strong> – boolean, allows for overwriting during appending, default = False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.read_featureCount">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">read_featureCount</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#read_featureCount"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.read_featureCount" title="Permalink to this definition"></a></dt>
<dd><p>Read tab files with common first column
:param nameElem: str, present in all files
:param path: str, path to directory with files
:param toLoad: str, to be present in file name (optional)
:param toClear: str, will be removed from file name
:param toAdd: str, to be added to file name
:param df: DataFrame, to be appended; default=None
:param overwrite: boolean, allows for overwriting during appending, default = False
:return: DataFrame</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.read_list">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">read_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#read_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.read_list" title="Permalink to this definition"></a></dt>
<dd><p>Read list from file. Each row becomes item in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filepath</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.read_tabFile">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">read_tabFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nameElem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toLoad</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toClear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toAdd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#read_tabFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.read_tabFile" title="Permalink to this definition"></a></dt>
<dd><p>Read tab files with common first column
:param nameElem: str, present in all files
:param path: str, path to directory with files
:param toLoad: str, to be present in file name
:param toClear: str, will be removed from file name
:param toAdd: str, to be added to file name
:param df: DataFrame, to be appended; default=None
:param overwrite: boolean, allows for overwriting during appending, default = False
:return: DataFrame</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.reverse_complement">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.reverse_complement" title="Permalink to this definition"></a></dt>
<dd><p>Reverse complement</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.reverse_complement_DNA">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement_DNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#reverse_complement_DNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.reverse_complement_DNA" title="Permalink to this definition"></a></dt>
<dd><p>Reverse complement</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.reverse_complement_RNA">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">reverse_complement_RNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#reverse_complement_RNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.reverse_complement_RNA" title="Permalink to this definition"></a></dt>
<dd><p>Reverse complement</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.rollingGC">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">rollingGC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s=&lt;class</span> <span class="pre">'pandas.core.series.Series'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#rollingGC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.rollingGC" title="Permalink to this definition"></a></dt>
<dd><p>Calculates GC from sequence, uses ‘boxcar’ window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s</strong> – Series containing sequence</p></li>
<li><p><strong>window</strong> – window size for GC calculation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Series with GC calculated, center=False</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.runPCA">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">runPCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components=2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#runPCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.runPCA" title="Permalink to this definition"></a></dt>
<dd><p>Run PCA analysis and re-assigns column names and index names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame</p></li>
<li><p><strong>n_components</strong> – int, default 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple consisting of DataFrame with PCA results and a list of PC values</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.timestamp">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">timestamp</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.timestamp" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>timestamp as a str()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.methods.timestampRandomInt">
<span class="sig-prename descclassname"><span class="pre">TTools.methods.</span></span><span class="sig-name descname"><span class="pre">timestampRandomInt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/methods.html#timestampRandomInt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.methods.timestampRandomInt" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>timestamp and random number as a str()</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.nascent">
<span id="ttools-nascent-module"></span><h2>TTools.nascent module<a class="headerlink" href="#module-TTools.nascent" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="TTools.nascent.Fold">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">Fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tempDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Fold.RNAfold">
<span class="sig-name descname"><span class="pre">RNAfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold.RNAfold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold.RNAfold" title="Permalink to this definition"></a></dt>
<dd><p>Calculates dG using RNAfold (ViennaRNA)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – input data {list, Series, DataFrame with “name” column}</p></li>
<li><p><strong>saveData</strong> – boolean, default False</p></li>
<li><p><strong>temp</strong> – int, default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Fold.RNAinvert">
<span class="sig-name descname"><span class="pre">RNAinvert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quick</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold.RNAinvert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold.RNAinvert" title="Permalink to this definition"></a></dt>
<dd><p>Returns n sequences with with given structure</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> – str with secondary RNA structure</p></li>
<li><p><strong>saveData</strong> – boolean, default False</p></li>
<li><p><strong>temp</strong> – int, default None</p></li>
<li><p><strong>n</strong> – int number of output sequences, default 5</p></li>
<li><p><strong>RNAprimer</strong> – str sequence</p></li>
<li><p><strong>stall</strong> – str nucleotide</p></li>
<li><p><strong>quick</strong> – boolean if False uses -Fmp -f 0.01 params, default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Fold.RNAinvertStall">
<span class="sig-name descname"><span class="pre">RNAinvertStall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNAprimer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'A'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold.RNAinvertStall"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold.RNAinvertStall" title="Permalink to this definition"></a></dt>
<dd><p>Returns sequences without nt that is present in stall</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> – str with secondary RNA structure</p></li>
<li><p><strong>RNAprimer</strong> – str sequence</p></li>
<li><p><strong>stall</strong> – str sequence, default “A”</p></li>
<li><p><strong>n</strong> – int, default 200</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Fold.UNAfold">
<span class="sig-name descname"><span class="pre">UNAfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold.UNAfold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold.UNAfold" title="Permalink to this definition"></a></dt>
<dd><p>Calculates dG using UNAfold</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – input data {list, Series, DataFrame with “name” column}</p></li>
<li><p><strong>saveData</strong> – boolean, default False</p></li>
<li><p><strong>temp</strong> – int, default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Fold.bashFolding">
<span class="sig-name descname"><span class="pre">bashFolding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RNA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Fold.bashFolding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Fold.bashFolding" title="Permalink to this definition"></a></dt>
<dd><p>Runs RNA folding using bash</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>method</strong> – “RNA” for ViennaRNA or “UNA” for UNAfold”</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="TTools.nascent.Hybrid">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">Hybrid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tempDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Hybrid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Hybrid" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Hybrid.RNAhybrid">
<span class="sig-name descname"><span class="pre">RNAhybrid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveData</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Hybrid.RNAhybrid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Hybrid.RNAhybrid" title="Permalink to this definition"></a></dt>
<dd><p>Calculates dG using hybrid-min</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – input data {list, Series, DataFrame with “name” column}</p></li>
<li><p><strong>saveData</strong> – boolean, default False</p></li>
<li><p><strong>temp</strong> – int, default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="TTools.nascent.Hybrid.bashHybrid">
<span class="sig-name descname"><span class="pre">bashHybrid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#Hybrid.bashHybrid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.Hybrid.bashHybrid" title="Permalink to this definition"></a></dt>
<dd><p>Runs hybris-min using bash</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.analyseViennamarkGC">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">analyseViennamarkGC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#analyseViennamarkGC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.analyseViennamarkGC" title="Permalink to this definition"></a></dt>
<dd><p>Leaves only C and G in stem structures</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str with vienna format</p></li>
<li><p><strong>sequence</strong> – str sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.extendingWindow">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">extendingWindow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#extendingWindow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.extendingWindow" title="Permalink to this definition"></a></dt>
<dd><p>Returns DataFrame of sequences of all possible lengths between minimum (m) and length of input sequence -1</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – str</p></li>
<li><p><strong>name</strong> – str, default “name”</p></li>
<li><p><strong>strand</strong> – str {“plus”,”minus”,”both”}, default “plus” (not tested for others)</p></li>
<li><p><strong>temp</strong> – int, default 30</p></li>
<li><p><strong>m</strong> – int, default 7 - RNAfold does not return any values for length shorter than 7 even at 4 deg C</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame of sequences with names according to nascent.slidingWindow convention</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.foldNascentElem">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">foldNascentElem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#foldNascentElem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.foldNascentElem" title="Permalink to this definition"></a></dt>
<dd><p>Fold the very 3’ of nascent elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – DataFrame</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.handleInput">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">handleInput</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keepNames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#handleInput"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.handleInput" title="Permalink to this definition"></a></dt>
<dd><p>Input data with columns: “seq” or “sequence” and “name” (optional)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – {str, list, Series, DataFrame}</p>
</dd>
</dl>
<p>:param keepNames if True use given names, default True
:return: DataFrame where index become name of sequence to fold</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.join2d">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">join2d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use='format'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#join2d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.join2d" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>use</strong> – str, default “format”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.markVienna">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">markVienna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#markVienna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.markVienna" title="Permalink to this definition"></a></dt>
<dd><p>Apply analyseViennamarkGC</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> – DataFrame</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.merge2d">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">merge2d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#merge2d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.merge2d" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> – DataFrame</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.name2index">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">name2index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">object)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#name2index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.name2index" title="Permalink to this definition"></a></dt>
<dd><p>Extracts position from sequence name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>s1</strong> – Series with names from prepareNascent function</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Series with positions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.nascentElems">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentElems</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#nascentElems"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.nascentElems" title="Permalink to this definition"></a></dt>
<dd><p>Describe elements of secondary structure: stems, multistems.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str</p></li>
<li><p><strong>sequence</strong> – str</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.nascentElemsDataFrame">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentElemsDataFrame</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#nascentElemsDataFrame"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.nascentElemsDataFrame" title="Permalink to this definition"></a></dt>
<dd><p>Apply nascentElem to DataFrame of folded sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – DataFrame containing df[‘vienna’] and df[‘sequence’]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.nascentFolding">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">nascentFolding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#nascentFolding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.nascentFolding" title="Permalink to this definition"></a></dt>
<dd><p>Combines folding function: fold RNA, locate last nascent element and calculate dG of it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – str</p></li>
<li><p><strong>temp</strong> – int, default 30</p></li>
<li><p><strong>window</strong> – int, default 100</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.parseFoldingName">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">parseFoldingName</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#parseFoldingName"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.parseFoldingName" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.prepareNascent">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">prepareNascent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#prepareNascent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.prepareNascent" title="Permalink to this definition"></a></dt>
<dd><p>Divide long transcript into short sequences. Combines output of extendingWindow and slidingWindow.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – str</p></li>
<li><p><strong>name</strong> – str, default “name”</p></li>
<li><p><strong>strand</strong> – str {plus,minus,both}, default “plus” (not tested for others)</p></li>
<li><p><strong>temp</strong> – int, default 30</p></li>
<li><p><strong>window</strong> – int, default 100</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dataframe with sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.selectFoldedN">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">selectFoldedN</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pattern='(((((....)))))'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#selectFoldedN"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.selectFoldedN" title="Permalink to this definition"></a></dt>
<dd><p>Takes Fold().RNAFold() df as an input. Selects n rows with a given pattern
on the 5end and most different folding energy</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame</p></li>
<li><p><strong>n</strong> – int samples, default 5</p></li>
<li><p><strong>pattern</strong> – str vienna format, default “(((((….)))))”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.nascent.slidingWindow">
<span class="sig-prename descclassname"><span class="pre">TTools.nascent.</span></span><span class="sig-name descname"><span class="pre">slidingWindow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/nascent.html#slidingWindow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.nascent.slidingWindow" title="Permalink to this definition"></a></dt>
<dd><p>Slices sequence using sliding window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – str</p></li>
<li><p><strong>name</strong> – str default “name”</p></li>
<li><p><strong>strand</strong> – str {“plus”,”both”,”minus”} default “plus”</p></li>
<li><p><strong>temp</strong> – int default 30</p></li>
<li><p><strong>window</strong> – int default 80</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with sliding windows</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.plotting">
<span id="ttools-plotting-module"></span><h2>TTools.plotting module<a class="headerlink" href="#module-TTools.plotting" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.clusterClusterMap">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">clusterClusterMap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#clusterClusterMap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.clusterClusterMap" title="Permalink to this definition"></a></dt>
<dd><p>Clustermap for clusters
:param df: DataFrame
:return:</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats_chimeric">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_chimeric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats_chimeric"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats_chimeric" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats_mapping">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats_mapping" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats_mistmatches">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_mistmatches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats_mistmatches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats_mistmatches" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats_readLen">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_readLen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats_readLen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats_readLen" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.hplotSTARstats_reads">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">hplotSTARstats_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#hplotSTARstats_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.hplotSTARstats_reads" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotCumulativePeaks">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotCumulativePeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df2=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_pos=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color1='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color2='magenta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotCumulativePeaks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotCumulativePeaks" title="Permalink to this definition"></a></dt>
<dd><p>Plot single gene peaks metaplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> – str with path to csv file or DataFrame</p></li>
<li><p><strong>df2</strong> – DataFrame</p></li>
<li><p><strong>local_pos</strong> – list of features (peaks/troughs)</p></li>
<li><p><strong>dpi</strong> – int, default 150</p></li>
<li><p><strong>title</strong> – str</p></li>
<li><p><strong>start</strong> – int</p></li>
<li><p><strong>stop</strong> – int</p></li>
<li><p><strong>window</strong> – int, default 50</p></li>
<li><p><strong>figsize</strong> – tuple, default (4,3)</p></li>
<li><p><strong>color1</strong> – str, default “green”</p></li>
<li><p><strong>color2</strong> – str, default “magenta”</p></li>
<li><p><strong>lc</strong> – str, default “red”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotPCA">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotPCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PClimit=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">7)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PCval=[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotPCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotPCA" title="Permalink to this definition"></a></dt>
<dd><p>Plot PCA plot</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame</p></li>
<li><p><strong>names</strong> – list of names to annotate</p></li>
<li><p><strong>title</strong> – str</p></li>
<li><p><strong>PClimit</strong> – int number of PC to plot, default 1</p></li>
<li><p><strong>figsize</strong> – tuple, default (7,7)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotPCA</span><span class="p">(</span><span class="n">methods</span><span class="o">.</span><span class="n">runPCA</span><span class="p">(</span><span class="n">example_df</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats_chimeric">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_chimeric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats_chimeric"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats_chimeric" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats_mapping">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats_mapping" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats_mistmatches">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_mistmatches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats_mistmatches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats_mistmatches" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats_readLen">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_readLen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats_readLen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats_readLen" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plotSTARstats_reads">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plotSTARstats_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plotSTARstats_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plotSTARstats_reads" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>dpi</strong> – int, default=150</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plot_as_box_plot">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_as_box_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plot_as_box_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plot_as_box_plot" title="Permalink to this definition"></a></dt>
<dd><p>Plots figure similar to box plot: median, 2 and 3 quartiles and min-max range</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – Dataframe() containing following columns:<code class="docutils literal notranslate"><span class="pre">`['position']</span> <span class="pre">['mean']</span> <span class="pre">['median']</span> <span class="pre">['std']`</span></code>
optionally <code class="docutils literal notranslate"><span class="pre">`['nucleotide']</span> <span class="pre">['q1']</span> <span class="pre">['q3']</span> <span class="pre">['max']</span> <span class="pre">['min']`</span></code></p></li>
<li><p><strong>title</strong> – str</p></li>
<li><p><strong>start</strong> – int</p></li>
<li><p><strong>stop</strong> – int</p></li>
<li><p><strong>figsize</strong> – tuple, default (7,4)</p></li>
<li><p><strong>ylim</strong> – tuple OY axes lim. Default (None,0.01)</p></li>
<li><p><strong>dpi</strong> – int, default 150</p></li>
<li><p><strong>color</strong> – str, default “green”</p></li>
<li><p><strong>h_lines</strong> – list of horizontal lines</p></li>
<li><p><strong>lc</strong> – str color of horizontal lines, default “red”</p></li>
<li><p><strong>offset</strong> – int number to offset position if 5’ flank was used, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plot_diff">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_medians=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_ranges=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plot_diff"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plot_diff" title="Permalink to this definition"></a></dt>
<dd><p>Plot given dataset and reference, differences are marked</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> – str with path to csv file or DataFrame</p></li>
<li><p><strong>dataset</strong> – DataFrame containing following columns:<code class="docutils literal notranslate"><span class="pre">`['position']</span> <span class="pre">['mean']</span> <span class="pre">['median']</span> <span class="pre">['std']`</span></code>
optionally <code class="docutils literal notranslate"><span class="pre">`['nucleotide']</span> <span class="pre">['q1']</span> <span class="pre">['q3']</span> <span class="pre">['max']</span> <span class="pre">['min']`</span></code></p></li>
<li><p><strong>ranges</strong> – str “mm” : min-max or “qq” : q1-q3</p></li>
<li><p><strong>label</strong> – str</p></li>
<li><p><strong>start</strong> – int</p></li>
<li><p><strong>stop</strong> – int</p></li>
<li><p><strong>plot_medians</strong> – boolean if True plot medians, default True</p></li>
<li><p><strong>plot_ranges</strong> – boolean if True plot ranges, default True</p></li>
<li><p><strong>figsize</strong> – tuple, default (7,3)</p></li>
<li><p><strong>ylim</strong> – tuple OY axes lim, default (None,0.01)</p></li>
<li><p><strong>h_lines</strong> – list of horizontal lines</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plot_heatmap">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title='Heatmap</span> <span class="pre">of</span> <span class="pre">differences</span> <span class="pre">between</span> <span class="pre">dataset</span> <span class="pre">and</span> <span class="pre">reference</span> <span class="pre">plot</span> <span class="pre">for</span> <span class="pre">RDN37-1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plot_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plot_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Plot heat map of differences, from dataframe generated by compare1toRef(dataset, heatmap=True) function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>title</strong> – str</p></li>
<li><p><strong>vmin</strong> – </p></li>
<li><p><strong>vmax</strong> – </p></li>
<li><p><strong>figsize</strong> – tuple, default (20,10)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.plotting.plot_to_compare">
<span class="sig-prename descclassname"><span class="pre">TTools.plotting.</span></span><span class="sig-name descname"><span class="pre">plot_to_compare</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color1='green'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color2='black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_label=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim=(None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0.01)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_lines=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lc='red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset=300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/plotting.html#plot_to_compare"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.plotting.plot_to_compare" title="Permalink to this definition"></a></dt>
<dd><p>Figure to compare to plots similar to box plot: median, 2 and 3 quartiles and min-max range</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> – str with path to csv file or DataFrame</p></li>
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>color1</strong> – str, default “green”</p></li>
<li><p><strong>color2</strong> – str, default “black”</p></li>
<li><p><strong>ref_label</strong> – str</p></li>
<li><p><strong>label</strong> – str</p></li>
<li><p><strong>title</strong> – str</p></li>
<li><p><strong>start</strong> – int</p></li>
<li><p><strong>stop</strong> – int</p></li>
<li><p><strong>figsize</strong> – tuple, default (7,4)</p></li>
<li><p><strong>ylim</strong> – tuple OY axes lim. Default (None,0.01)</p></li>
<li><p><strong>h_lines</strong> – list of horizontal lines</p></li>
<li><p><strong>lc</strong> – str color of horizontal lines, default “red”</p></li>
<li><p><strong>dpi</strong> – int, default 150</p></li>
<li><p><strong>offset</strong> – int number to offset position if 5’ flank was used, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.profiles">
<span id="ttools-profiles-module"></span><h2>TTools.profiles module<a class="headerlink" href="#module-TTools.profiles" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.FoldingFromBigWig">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">FoldingFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwFWD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwREV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dG65nt&#64;30C'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#FoldingFromBigWig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.FoldingFromBigWig" title="Permalink to this definition"></a></dt>
<dd><p>Pulls folding information from BigWig folding data for a given gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> – str</p></li>
<li><p><strong>gtf</strong> – pyCRAC.GTF2 object with GTF and TAB files loaded</p></li>
<li><p><strong>bwFWD</strong> – dict of pyBigWig objects</p></li>
<li><p><strong>bwREV</strong> – dict of pyBigWig objects</p></li>
<li><p><strong>ranges</strong> – int flanks to be added for the gene, default 0</p></li>
<li><p><strong>offset</strong> – int to offset folding data, default 15</p></li>
<li><p><strong>fold</strong> – name of output column, default=”<a class="reference external" href="mailto:dG65nt&#37;&#52;&#48;30C">dG65nt<span>&#64;</span>30C</a>”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.calculateFDR">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">calculateFDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations=100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_FDR=0.05</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#calculateFDR"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.calculateFDR" title="Permalink to this definition"></a></dt>
<dd><p>Calculates False Discovery Rate (FDR) for a given dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – Series</p></li>
<li><p><strong>iterations</strong> – int, default 100</p></li>
<li><p><strong>target_FDR</strong> – float, detault 0.05</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.compare1toRef">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">compare1toRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heatmap=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relative=False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#compare1toRef"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.compare1toRef" title="Permalink to this definition"></a></dt>
<dd><p>Takes Series and compare this with reference DataFrame()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> – str with path to csv file or DataFrame</p></li>
<li><p><strong>dataset</strong> – Series</p></li>
<li><p><strong>ranges</strong> – mm : min-max or qq : q1-q3</p></li>
<li><p><strong>heatmap</strong> – boolean, heatmap=False: Dataframe with(reference_above_experiment minimum etc.): rae_min, rae_max, ear_min, ear_max;
heatmap=True: Series of differences to plot heatmap</p></li>
<li><p><strong>relative</strong> – boolean, only for heatmap, recalculates differences according to the peak size. Warning: negative values are in range -1 to 0 but positive are from 0 to values higher than 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dataframe (heatmap=False) or Series (heatmap=True)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.compareMoretoRef">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">compareMoretoRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges='mm'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#compareMoretoRef"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.compareMoretoRef" title="Permalink to this definition"></a></dt>
<dd><p>Takes Dataframe created by filter_df and compare this with reference DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> – str with path to csv file or DataFrame</p></li>
<li><p><strong>dataset</strong> – Series</p></li>
<li><p><strong>ranges</strong> – mm : min-max or qq : q1-q3</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.dictBigWig">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">dictBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strands</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#dictBigWig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.dictBigWig" title="Permalink to this definition"></a></dt>
<dd><p>Preloads BigWig files to memory using pyBigWig tools</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>files</strong> – list of files</p></li>
<li><p><strong>path</strong> – str</p></li>
<li><p><strong>strands</strong> – boolean, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict or dict, dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.findPeaks">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">findPeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=20</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#findPeaks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.findPeaks" title="Permalink to this definition"></a></dt>
<dd><p>Find local extrema using SciPy argrelextrema function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> – Series data to localize peaks</p></li>
<li><p><strong>window</strong> – int, To smooth data before peak-calling. default 1 (no smoothing)</p></li>
<li><p><strong>win_type</strong> – str type of smoothing window, default “blackman”</p></li>
<li><p><strong>order</strong> – int minimal spacing between peaks, argrelextrema order parameter, default  20</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of peaks</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.findTroughs">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">findTroughs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1=Series([]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype:</span> <span class="pre">float64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">win_type='blackman'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=20</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#findTroughs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.findTroughs" title="Permalink to this definition"></a></dt>
<dd><p>Find local minima using SciPy argrelextrema function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> – Series data to localize peaks</p></li>
<li><p><strong>window</strong> – int, To smooth data before trough-calling. default 1 (no smoothing)</p></li>
<li><p><strong>win_type</strong> – str type of smoothing window, default “blackman”</p></li>
<li><p><strong>order</strong> – int minimal spacing between min, argrelextrema order parameter, default 20</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of troughs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.geneFromBigWig">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">geneFromBigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwFWD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwREV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toStrip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#geneFromBigWig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.geneFromBigWig" title="Permalink to this definition"></a></dt>
<dd><p>Pulls genome coverage from BigWig data for a given gene. One BigWig file -&gt; one column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_name</strong> – str</p></li>
<li><p><strong>gtf</strong> – pyCRAC.GTF2 object with GTF and TAB files loaded</p></li>
<li><p><strong>bwFWD</strong> – dict of pyBigWig objects</p></li>
<li><p><strong>bwREV</strong> – dict of pyBigWig objects</p></li>
<li><p><strong>toStrip</strong> – str of name to be stripped</p></li>
<li><p><strong>ranges</strong> – int flanks to be added for the gene, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.ntotal">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">ntotal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#ntotal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.ntotal" title="Permalink to this definition"></a></dt>
<dd><p>Normalize data in DataFrame to fraction of total column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>drop</strong> – boolean, if True drop ‘position’ and ‘nucleotide’ columns, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.parseConcatFile">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">parseConcatFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reads'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RPM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#parseConcatFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.parseConcatFile" title="Permalink to this definition"></a></dt>
<dd><p>Parse concat file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – str with path of the concat file</p></li>
<li><p><strong>gtf</strong> – pyCRAC.GTF2 object with GTF and TAB files loaded</p></li>
<li><p><strong>use</strong> – str with name of column tu use [‘reads’, ‘substitutions’, ‘deletions’], default “reads”</p></li>
<li><p><strong>RPM</strong> – boolean, default False</p></li>
<li><p><strong>ranges</strong> – int flanks to be added for the gene, default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict of DataFrames; using gene name as a key</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.preprocess">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">preprocess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">input_df=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[],</span> <span class="pre">let_in=[''],</span> <span class="pre">let_out=['wont_find_this_string'],</span> <span class="pre">stats=False,</span> <span class="pre">smooth=True,</span> <span class="pre">window=10,</span> <span class="pre">win_type='blackman'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#preprocess"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.preprocess" title="Permalink to this definition"></a></dt>
<dd><p>Combines methods.filterExp and expStats. Returns DataFrame with choosen experiments, optionally apply smoothing and stats</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_df</strong> – DataFrame</p></li>
<li><p><strong>let_in</strong> – list of words that characterize experiment, default [‘’]</p></li>
<li><p><strong>let_out</strong> – list of words that disqualify experiments, default [‘wont_find_this_string’]</p></li>
<li><p><strong>stats</strong> – boolean, if True return stats for all experiments, default False</p></li>
<li><p><strong>smooth</strong> – boolean, if True apply smoothing window, default True</p></li>
<li><p><strong>window</strong> – int smoothing window, default 10</p></li>
<li><p><strong>win_type</strong> – str type of smoothing window, default “blackman”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with ‘mean’, ‘median’, ‘min’, ‘max’ and quartiles if more than 2 experiments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.pseudocounts">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">pseudocounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df=&lt;class</span> <span class="pre">'pandas.core.frame.DataFrame'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value=0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#pseudocounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.pseudocounts" title="Permalink to this definition"></a></dt>
<dd><p>Add pseudocounts to data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – DataFrame</p></li>
<li><p><strong>value</strong> – float, default 0.01</p></li>
<li><p><strong>drop</strong> – boolean, if True drop ‘position’ and ‘nucleotide’ columns, default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.save_csv">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">save_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_ref=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasets=Empty</span> <span class="pre">DataFrame</span> <span class="pre">Columns:</span> <span class="pre">[]</span> <span class="pre">Index:</span> <span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path=None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#save_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.save_csv" title="Permalink to this definition"></a></dt>
<dd><p>Saves Dataframe to csv</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_ref</strong> – DataFrame with <code class="docutils literal notranslate"><span class="pre">['position']</span></code> and <code class="docutils literal notranslate"><span class="pre">['nucleotide']</span></code> columns</p></li>
<li><p><strong>datasets</strong> – DataFrame containinig experimental data only</p></li>
<li><p><strong>path</strong> – str, Optional: path to save csv. Default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.profiles.stripBigWignames">
<span class="sig-prename descclassname"><span class="pre">TTools.profiles.</span></span><span class="sig-name descname"><span class="pre">stripBigWignames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/profiles.html#stripBigWignames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.profiles.stripBigWignames" title="Permalink to this definition"></a></dt>
<dd><p>Strip “_rev.bw” and “_fwd.bw” form file names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>files</strong> – list of filenames</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of unique names</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-TTools.secondary">
<span id="ttools-secondary-module"></span><h2>TTools.secondary module<a class="headerlink" href="#module-TTools.secondary" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.Lstem">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">Lstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#Lstem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.Lstem" title="Permalink to this definition"></a></dt>
<dd><p>Returns list of positions where “(” is found using coordinates {1,inf}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>vienna</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.Rstem">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">Rstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#Rstem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.Rstem" title="Permalink to this definition"></a></dt>
<dd><p>Returns list of positions where “)” is found using coordinates {1,inf}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>vienna</strong> – str</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.checkVienna">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">checkVienna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#checkVienna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.checkVienna" title="Permalink to this definition"></a></dt>
<dd><p>Validates integrity of vienna file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – str</p></li>
<li><p><strong>vienna</strong> – str</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if pass</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.loopStems">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">loopStems</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loopsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testPrint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#loopStems"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.loopStems" title="Permalink to this definition"></a></dt>
<dd><p>Returns postions of stem of single hairpins and multiloop stems. Use coordinates {1:inf}.
Warninig: tested with single multiloop stems only</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str</p></li>
<li><p><strong>loopsList</strong> – list (option)</p></li>
<li><p><strong>testPrint</strong> – boolean to default=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list, list (stems: list of tuples; multistems: list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.loops">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">loops</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#loops"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.loops" title="Permalink to this definition"></a></dt>
<dd><p>Returns first positions outside the loop i.e. “.((….)).” returns [(3,8)]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>vienna</strong> – vienna</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.substructures">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">substructures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#substructures"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.substructures" title="Permalink to this definition"></a></dt>
<dd><p>list sub-structures of the given structure</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str</p></li>
<li><p><strong>sequence</strong> – str</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.test">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stems</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multistems</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linkers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.test" title="Permalink to this definition"></a></dt>
<dd><p>Prints vienna with given features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str</p></li>
<li><p><strong>loops</strong> – list of tuples (option)</p></li>
<li><p><strong>stems</strong> – list of tuples (option)</p></li>
<li><p><strong>multistems</strong> – list (option)</p></li>
<li><p><strong>linkers</strong> – list (option)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="TTools.secondary.vienna2format">
<span class="sig-prename descclassname"><span class="pre">TTools.secondary.</span></span><span class="sig-name descname"><span class="pre">vienna2format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vienna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loopsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stemsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multistemsList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testPrint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/TTools/secondary.html#vienna2format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#TTools.secondary.vienna2format" title="Permalink to this definition"></a></dt>
<dd><p>Converts vienna format to letters: O - loop, S - stem, M - multiloop stem and L - linker</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vienna</strong> – str</p></li>
<li><p><strong>loopsList</strong> – list (optional)</p></li>
<li><p><strong>stemsList</strong> – list (optional)</p></li>
<li><p><strong>multistemsList</strong> – list (optional)</p></li>
<li><p><strong>testPrint</strong> – defauls=False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str in “format”</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="scripts.html" class="btn btn-neutral float-left" title="Scripts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Tomasz W. Turowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>