<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>TTools.go_enrichment &mdash; TTools 0.1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> TTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../TTools.html">TTools package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">TTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>TTools.go_enrichment</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for TTools.go_enrichment</h1><div class="highlight"><pre>
<span></span><span class="n">author</span><span class="o">=</span><span class="s2">&quot;Jan Miko≈Çajczyk&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="n">go_mappings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;biological_process&#39;</span><span class="p">:</span> <span class="s2">&quot;GO:0008150&quot;</span><span class="p">,</span>
    <span class="s1">&#39;molecular_function&#39;</span><span class="p">:</span> <span class="s2">&quot;GO:0003674&quot;</span>
<span class="p">}</span>
<span class="n">available_test_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;FISHER&quot;</span><span class="p">,</span> <span class="s2">&quot;BINOMIAL&quot;</span><span class="p">]</span>
<span class="n">available_corrections</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;FDR&quot;</span><span class="p">,</span> <span class="s2">&quot;BONFERRONI&quot;</span><span class="p">,</span> <span class="s2">&quot;NONE&quot;</span><span class="p">]</span>

<div class="viewcode-block" id="get_enrichment"><a class="viewcode-back" href="../../TTools.html#TTools.go_enrichment.get_enrichment">[docs]</a><span class="k">def</span> <span class="nf">get_enrichment</span><span class="p">(</span><span class="n">query_genes</span><span class="p">,</span> <span class="n">organism</span><span class="p">,</span> <span class="n">use_reference_set</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ref_organism</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">go_dataset</span><span class="o">=</span><span class="s1">&#39;biological_process&#39;</span><span class="p">,</span> <span class="n">test_type</span><span class="o">=</span><span class="s2">&quot;FISHER&quot;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="s2">&quot;FDR&quot;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Run a GO term enrichment test using PANTHER API</span>

<span class="sd">    :param query_genes: List of sequence identifiers of queried genes (e.g. transcript ids, gene ids)</span>
<span class="sd">    :type query_genes: list</span>
<span class="sd">    :param organism: Taxid of query species (e.g. &quot;9606&quot; for H. sapiens)</span>
<span class="sd">    :type organism: str</span>
<span class="sd">    :param use_reference_set: Use a custom set of rerence (background) genes? Default False. If True, ref_genes and ref_species need to be specifed.</span>
<span class="sd">    :type use_reference_set: bool</span>
<span class="sd">    :param ref_genes: *optional* list of reference genes. Specifying None (default) will use the whole genome of species specified in organism. When passing a list, ref_organism taxid must also be provided.</span>
<span class="sd">    :type ref_genes: list</span>
<span class="sd">    :param ref_species: Taxid of reference species, required when ref_genes is not None</span>
<span class="sd">    :type ref_species: str</span>
<span class="sd">    :param go_dataset: Which annotation dataset to query, &quot;biological_process&quot; or &quot;molecular_function&quot;</span>
<span class="sd">    :type go_dataset: str</span>
<span class="sd">    :param test_type: Which tatistical test to use. Available: &quot;FISHER&quot; (default), &quot;BINOMIAL&quot;</span>
<span class="sd">    :type test_type: str</span>
<span class="sd">    :param correction: Which multiple testing correction method to use. Available: &quot;FDR&quot; (default), &quot;BONFERRONI&quot;, &quot;NONE&quot;</span>
<span class="sd">    :type correction: str</span>
<span class="sd">    :returns: Unfiltered DataFrame of results.</span>
<span class="sd">    :rtype: pandas.DataFrame:</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1">## return empty df if empty list provided as input</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">query_genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">dummy_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s2">&quot;number_in_list&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;fold_enrichment&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;fdr&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;expected&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;number_in_reference&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;pValue&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;plus_minus&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;term.id&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;term.label&quot;</span><span class="p">:</span> <span class="p">[]</span>
        <span class="p">})</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Empty list provided, returning empty DataFrame.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dummy_df</span>
    <span class="c1">## check if passed parameters are correct</span>
    <span class="k">if</span> <span class="n">go_dataset</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">go_mappings</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Incorrect go_dataset value specified. Available values: &quot;biological_process&quot;, &quot;molecular_function&quot;&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">test_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">available_test_types</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Incorrect test_type value specified. Available values: &quot;FISHER&quot;, &quot;BINOMIAL&quot;&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">correction</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">available_corrections</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Incorrect correction value specified. Available values: &quot;FDR&quot;, &quot;BONFERRONI&quot;, &quot;NONE&quot;&#39;</span><span class="p">)</span>
    <span class="c1">## format provided lists into strings and construct query</span>
    <span class="n">seq_ids</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">query_genes</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">use_reference_set</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ref_genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Reference gene list needs to be provided when use_reference_set is True&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ref_organism</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Reference organism taxid needs to be provided when use_reference_set is True&quot;</span><span class="p">)</span>
        <span class="n">ref_ids</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ref_genes</span><span class="p">)</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;geneInputList&quot;</span><span class="p">:</span> <span class="n">seq_ids</span><span class="p">,</span>
            <span class="s2">&quot;organism&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="p">,</span>
            <span class="s2">&quot;refInputList&quot;</span><span class="p">:</span> <span class="n">ref_ids</span><span class="p">,</span>
            <span class="s2">&quot;refOrganism&quot;</span><span class="p">:</span> <span class="n">ref_organism</span><span class="p">,</span>
            <span class="s2">&quot;annotDataSet&quot;</span><span class="p">:</span> <span class="n">go_mappings</span><span class="p">[</span><span class="n">go_dataset</span><span class="p">],</span>
            <span class="s2">&quot;enrichmentTestType&quot;</span><span class="p">:</span> <span class="n">test_type</span><span class="p">,</span>
            <span class="s2">&quot;correction&quot;</span><span class="p">:</span> <span class="n">correction</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;geneInputList&quot;</span><span class="p">:</span> <span class="n">seq_ids</span><span class="p">,</span>
            <span class="s2">&quot;organism&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="p">,</span>
            <span class="s2">&quot;annotDataSet&quot;</span><span class="p">:</span> <span class="n">go_mappings</span><span class="p">[</span><span class="n">go_dataset</span><span class="p">],</span>
            <span class="s2">&quot;enrichmentTestType&quot;</span><span class="p">:</span> <span class="n">test_type</span><span class="p">,</span>
            <span class="s2">&quot;correction&quot;</span><span class="p">:</span> <span class="n">correction</span>
        <span class="p">}</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span><span class="s1">&#39;http://pantherdb.org/services/oai/pantherdb/enrich/overrep&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">query</span><span class="p">)</span>
    <span class="c1">## extract data to a dataframe</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">json_normalize</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()[</span><span class="s1">&#39;results&#39;</span><span class="p">][</span><span class="s1">&#39;result&#39;</span><span class="p">])</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Tomasz W. Turowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>