<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>TTools.profiles &mdash; TTools 0.1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> TTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../TTools.html">TTools package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">TTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>TTools.profiles</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for TTools.profiles</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">scipy.signal</span> <span class="kn">import</span> <span class="n">argrelextrema</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">warnings</span> <span class="kn">import</span> <span class="n">warn</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>

<span class="c1">### PROFILES (ProfileAnalyser) ###</span>
<span class="c1"># functions to deal with profiles of individual genes</span>

<span class="c1">#importing data</span>
<span class="c1">##concat file</span>
<div class="viewcode-block" id="parseConcatFile"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.parseConcatFile">[docs]</a><span class="k">def</span> <span class="nf">parseConcatFile</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">use</span><span class="o">=</span><span class="s1">&#39;reads&#39;</span><span class="p">,</span> <span class="n">RPM</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Parse concat file</span>

<span class="sd">    :param path: str with path of the concat file</span>
<span class="sd">    :param gtf: pyCRAC.GTF2 object with GTF and TAB files loaded</span>
<span class="sd">    :param use: str with name of column tu use [&#39;reads&#39;, &#39;substitutions&#39;, &#39;deletions&#39;], default &quot;reads&quot;</span>
<span class="sd">    :param RPM: boolean, default False</span>
<span class="sd">    :param ranges: int flanks to be added for the gene, default 0</span>
<span class="sd">    :return: dict of DataFrames; using gene name as a key</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># load concat</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="s1">&#39;reads&#39;</span><span class="p">,</span> <span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;deletions&#39;</span><span class="p">,</span> <span class="s1">&#39;experiment&#39;</span><span class="p">,</span> <span class="s1">&#39;reads_pM&#39;</span><span class="p">,</span>
               <span class="s1">&#39;substitutions_pM&#39;</span><span class="p">,</span> <span class="s1">&#39;deletions_pM&#39;</span><span class="p">]</span>
    <span class="n">df_rawConcat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span>

    <span class="c1"># use collumn</span>
    <span class="k">if</span> <span class="n">RPM</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="n">use</span> <span class="o">=</span> <span class="n">use</span> <span class="o">+</span> <span class="s2">&quot;_pM&quot;</span>

    <span class="c1"># output dict</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># genes</span>
    <span class="n">gene_list</span> <span class="o">=</span> <span class="n">df_rawConcat</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

    <span class="c1"># loop trough genes</span>
    <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
        <span class="n">geneLen</span> <span class="o">=</span> <span class="n">gtf</span><span class="o">.</span><span class="n">geneLength</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="c1"># preparing dataframe</span>
        <span class="n">df_output</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="n">df_output</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">ranges</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">geneLen</span> <span class="o">+</span> <span class="n">ranges</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="n">df_output</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gtf</span><span class="o">.</span><span class="n">genomicSequence</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="n">ranges</span><span class="p">)))</span>

        <span class="c1"># loop trough experiments</span>
        <span class="k">for</span> <span class="n">e</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_rawConcat</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;experiment&#39;</span><span class="p">):</span>
            <span class="n">df_output</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()[</span><span class="n">use</span><span class="p">]</span>

            <span class="c1"># check &#39;nucleotide&#39; collumn</span>
            <span class="k">if</span> <span class="n">df_output</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">!=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
                <span class="n">warn</span><span class="p">(</span><span class="s2">&quot;For &quot;</span> <span class="o">+</span> <span class="n">gene_name</span> <span class="o">+</span> <span class="s2">&quot; gene list of nucleotides does not match/&quot;</span><span class="p">,</span> <span class="ne">UserWarning</span><span class="p">)</span>

        <span class="n">output</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_output</span>
    <span class="k">return</span> <span class="n">output</span></div>

<span class="c1">##BigWig</span>
<div class="viewcode-block" id="stripBigWignames"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.stripBigWignames">[docs]</a><span class="k">def</span> <span class="nf">stripBigWignames</span><span class="p">(</span><span class="n">files</span><span class="o">=</span><span class="p">[]):</span>
    <span class="sd">&#39;&#39;&#39;Strip &quot;_rev.bw&quot; and &quot;_fwd.bw&quot; form file names</span>

<span class="sd">    :param files: list of filenames</span>
<span class="sd">    :return: list of unique names</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># returns uniq names for *fwd.bw and *rev.bw files</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">f</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_rev.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_fwd.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">files</span><span class="p">]))</span></div>


<div class="viewcode-block" id="dictBigWig"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.dictBigWig">[docs]</a><span class="k">def</span> <span class="nf">dictBigWig</span><span class="p">(</span><span class="n">files</span><span class="o">=</span><span class="p">[],</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">strands</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Preloads BigWig files to memory using pyBigWig tools</span>

<span class="sd">    :param files: list of files</span>
<span class="sd">    :param path: str</span>
<span class="sd">    :param strands: boolean, default True</span>
<span class="sd">    :return: dict or dict, dict</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">if</span> <span class="n">strands</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="c1"># output dict</span>
        <span class="n">output_dict_fwd</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="n">output_dict_rev</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

        <span class="n">strippedNames</span> <span class="o">=</span> <span class="n">stripBigWignames</span><span class="p">(</span><span class="n">files</span><span class="o">=</span><span class="n">files</span><span class="p">)</span>  <span class="c1"># list of common roots for strand specific data</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">strippedNames</span><span class="p">:</span>
            <span class="n">output_dict_fwd</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">f</span> <span class="o">+</span> <span class="s2">&quot;_fwd.bw&quot;</span><span class="p">)</span>
            <span class="n">output_dict_rev</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">f</span> <span class="o">+</span> <span class="s2">&quot;_rev.bw&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">output_dict_fwd</span><span class="p">,</span> <span class="n">output_dict_rev</span>

    <span class="k">if</span> <span class="n">strands</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="c1"># output dict</span>
        <span class="n">output_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">output_dict</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.bw&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">path</span> <span class="o">+</span> <span class="n">f</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">output_dict</span></div>

<div class="viewcode-block" id="geneFromBigWig"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.geneFromBigWig">[docs]</a><span class="k">def</span> <span class="nf">geneFromBigWig</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">bwFWD</span><span class="o">=</span><span class="p">{},</span> <span class="n">bwREV</span><span class="o">=</span><span class="p">{},</span> <span class="n">toStrip</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Pulls genome coverage from BigWig data for a given gene. One BigWig file -&gt; one column.</span>

<span class="sd">    :param gene_name: str</span>
<span class="sd">    :param gtf: pyCRAC.GTF2 object with GTF and TAB files loaded</span>
<span class="sd">    :param bwFWD: dict of pyBigWig objects</span>
<span class="sd">    :param bwREV: dict of pyBigWig objects</span>
<span class="sd">    :param toStrip: str of name to be stripped</span>
<span class="sd">    :param ranges: int flanks to be added for the gene, default 0</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># preparing output</span>
    <span class="n">df_t1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">df_t1</span><span class="p">[</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gtf</span><span class="o">.</span><span class="n">genomicSequence</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">ranges</span><span class="o">=</span><span class="n">ranges</span><span class="p">))</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>

    <span class="c1"># coordinates from GTF file</span>
    <span class="n">strand</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">coordinates</span> <span class="o">=</span> <span class="n">gtf</span><span class="o">.</span><span class="n">strand</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span> <span class="n">gtf</span><span class="o">.</span><span class="n">chromosome</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span> <span class="n">gtf</span><span class="o">.</span><span class="n">chromosomeCoordinates</span><span class="p">(</span>
        <span class="n">gene_name</span><span class="p">)</span>

    <span class="c1"># extracting data</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">bwFWD</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">bwFWD</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">bw</span><span class="o">.</span><span class="n">values</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">-</span> <span class="n">ranges</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">+</span> <span class="n">ranges</span><span class="p">))</span>
            <span class="n">df_t1</span><span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">toStrip</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">s1</span>

    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">bwREV</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">bwREV</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">bw</span><span class="o">.</span><span class="n">values</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">-</span> <span class="n">ranges</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">+</span> <span class="n">ranges</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">df_t1</span><span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">toStrip</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span> <span class="o">=</span> <span class="n">s1</span>

    <span class="k">return</span> <span class="n">df_t1</span></div>

<div class="viewcode-block" id="FoldingFromBigWig"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.FoldingFromBigWig">[docs]</a><span class="k">def</span> <span class="nf">FoldingFromBigWig</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">bwFWD</span><span class="o">=</span><span class="p">{},</span> <span class="n">bwREV</span><span class="o">=</span><span class="p">{},</span> <span class="n">ranges</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">offset</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span><span class="n">fold</span><span class="o">=</span><span class="s2">&quot;dG65nt@30C&quot;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Pulls folding information from BigWig folding data for a given gene.</span>

<span class="sd">    :param gene_name: str</span>
<span class="sd">    :param gtf: pyCRAC.GTF2 object with GTF and TAB files loaded</span>
<span class="sd">    :param bwFWD: dict of pyBigWig objects</span>
<span class="sd">    :param bwREV: dict of pyBigWig objects</span>
<span class="sd">    :param ranges: int flanks to be added for the gene, default 0</span>
<span class="sd">    :param offset: int to offset folding data, default 15</span>
<span class="sd">    :param fold: name of output column, default=&quot;dG65nt@30C&quot;</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># preparing output</span>
    <span class="n">df_t1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="c1"># coordinates from GTF file</span>
    <span class="n">strand</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">coordinates</span> <span class="o">=</span> <span class="n">gtf</span><span class="o">.</span><span class="n">strand</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span> <span class="n">gtf</span><span class="o">.</span><span class="n">chromosome</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span> <span class="n">gtf</span><span class="o">.</span><span class="n">chromosomeCoordinates</span><span class="p">(</span>
        <span class="n">gene_name</span><span class="p">)</span>

    <span class="c1"># extracting data</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">bwFWD</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">bwFWD</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">bw</span><span class="o">.</span><span class="n">values</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">-</span> <span class="n">ranges</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">+</span> <span class="n">ranges</span><span class="p">))</span>
            <span class="c1"># shifting to the extrustion point</span>
            <span class="n">df_t1</span><span class="p">[</span><span class="n">fold</span><span class="o">+</span><span class="s2">&quot;_add&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">offset</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;nt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">((</span><span class="n">offset</span> <span class="o">*</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">])</span> <span class="o">+</span> <span class="n">s1</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>

    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">bwREV</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">bwREV</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">s1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">bw</span><span class="o">.</span><span class="n">values</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">-</span> <span class="n">ranges</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span> <span class="o">+</span> <span class="n">ranges</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="c1"># shifting to the extrustion point</span>
            <span class="n">df_t1</span><span class="p">[</span><span class="n">fold</span><span class="o">+</span><span class="s2">&quot;_add&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">offset</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;nt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">((</span><span class="n">offset</span> <span class="o">*</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">])</span> <span class="o">+</span> <span class="n">s1</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">df_t1</span></div>

<span class="c1">##normalization</span>
<div class="viewcode-block" id="pseudocounts"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.pseudocounts">[docs]</a><span class="k">def</span> <span class="nf">pseudocounts</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Add pseudocounts to data</span>

<span class="sd">    :param df: DataFrame</span>
<span class="sd">    :param value: float, default 0.01</span>
<span class="sd">    :param drop: boolean, if True drop &#39;position&#39; and &#39;nucleotide&#39; columns, default True</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># drop additional columns</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="s1">&#39;position&#39;</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="n">s1_pos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;nucleotide&#39;</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="n">s2_nt</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># pseudocounts</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

    <span class="c1"># output</span>
    <span class="k">if</span> <span class="n">drop</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">df</span>
    <span class="k">elif</span> <span class="n">drop</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s1_pos</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s2_nt</span>
        <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="ntotal"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.ntotal">[docs]</a><span class="k">def</span> <span class="nf">ntotal</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Normalize data in DataFrame to fraction of total column</span>

<span class="sd">    :param df: DataFrame</span>
<span class="sd">    :param drop: boolean, if True drop &#39;position&#39; and &#39;nucleotide&#39; columns, default True</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># drop additional columns</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="s1">&#39;position&#39;</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="n">s1_pos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;nucleotide&#39;</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="n">s2_nt</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># normalization</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span> <span class="o">/</span> <span class="n">df</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="c1"># output</span>
    <span class="k">if</span> <span class="n">drop</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">df</span>
    <span class="k">elif</span> <span class="n">drop</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s1_pos</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s2_nt</span>
        <span class="k">return</span> <span class="n">df</span></div>

<span class="c1">################################################</span>
<span class="c1">#############        major TTools</span>

<div class="viewcode-block" id="preprocess"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.preprocess">[docs]</a><span class="k">def</span> <span class="nf">preprocess</span><span class="p">(</span><span class="n">input_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">let_in</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">],</span> <span class="n">let_out</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;wont_find_this_string&#39;</span><span class="p">],</span>
              <span class="n">stats</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">True</span> <span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Combines methods.filterExp and expStats. Returns DataFrame with choosen experiments, optionally apply smoothing and stats</span>

<span class="sd">    :param input_df: DataFrame</span>
<span class="sd">    :param let_in: list of words that characterize experiment, default [&#39;&#39;]</span>
<span class="sd">    :param let_out: list of words that disqualify experiments, default [&#39;wont_find_this_string&#39;]</span>
<span class="sd">    :param stats: boolean, if True return stats for all experiments, default False</span>
<span class="sd">    :param smooth: boolean, if True apply smoothing window, default True</span>
<span class="sd">    :param window: int smoothing window, default 10</span>
<span class="sd">    :param win_type: str type of smoothing window, default &quot;blackman&quot;</span>
<span class="sd">    :return: DataFrame with &#39;mean&#39;, &#39;median&#39;, &#39;min&#39;, &#39;max&#39; and quartiles if more than 2 experiments</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># filtering</span>
    <span class="n">selected</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">input_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span> <span class="k">if</span>
                <span class="nb">all</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">d</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">let_in</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">all</span><span class="p">(</span><span class="n">o</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">d</span> <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">let_out</span><span class="p">)]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Experiments: &quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">selected</span><span class="p">)</span>

    <span class="c1"># selecting data</span>
    <span class="n">working_df</span><span class="p">,</span> <span class="n">result_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">selected</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">smooth</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">working_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">input_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span><span class="o">.</span><span class="n">rolling</span><span class="p">(</span><span class="n">window</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="n">win_type</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">working_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">input_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>

    <span class="c1"># preparing output</span>
    <span class="k">if</span> <span class="n">stats</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">working_df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">function</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">]:</span> <span class="n">result_df</span><span class="p">[</span><span class="n">function</span><span class="p">]</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">working_df</span><span class="p">,</span> <span class="n">function</span><span class="p">)(</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># calculates using pandas function listed in []</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">selected</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>  <span class="c1"># calculating quartiles only in more than two experiments</span>
            <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;q1&#39;</span><span class="p">],</span> <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;q3&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">working_df</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">working_df</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result_df</span></div>

<div class="viewcode-block" id="calculateFDR"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.calculateFDR">[docs]</a><span class="k">def</span> <span class="nf">calculateFDR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(),</span> <span class="n">iterations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">target_FDR</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Calculates False Discovery Rate (FDR) for a given dataset.</span>

<span class="sd">    :param data: Series</span>
<span class="sd">    :param iterations: int, default 100</span>
<span class="sd">    :param target_FDR: float, detault 0.05</span>
<span class="sd">    :return: Series</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">normalized_data</span> <span class="o">=</span> <span class="n">data</span> <span class="o">/</span> <span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># normalize data</span>
    <span class="n">random_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">),</span> <span class="n">iterations</span><span class="p">))</span>  <span class="c1"># generating random datasets</span>
    <span class="n">normalized_random_df</span> <span class="o">=</span> <span class="n">random_df</span> <span class="o">/</span> <span class="n">random_df</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># normalize random df</span>

    <span class="c1"># getting Discovery Rate (DR)</span>
    <span class="n">DR_df</span> <span class="o">=</span> <span class="n">normalized_random_df</span><span class="o">.</span><span class="n">T</span> <span class="o">&gt;=</span> <span class="n">normalized_data</span>  <span class="c1"># random discovery rate</span>
    <span class="n">DR</span> <span class="o">=</span> <span class="n">DR_df</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">iterations</span>

    <span class="c1"># comparing DR to False Discovery Rate (FDR)</span>
    <span class="n">FDR</span> <span class="o">=</span> <span class="n">DR</span> <span class="o">&lt;=</span> <span class="n">target_FDR</span>
    <span class="k">return</span> <span class="n">data</span> <span class="o">*</span> <span class="n">FDR</span></div>

<div class="viewcode-block" id="findPeaks"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.findPeaks">[docs]</a><span class="k">def</span> <span class="nf">findPeaks</span><span class="p">(</span><span class="n">s1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(),</span> <span class="n">window</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Find local extrema using SciPy argrelextrema function</span>

<span class="sd">    :param s1: Series data to localize peaks</span>
<span class="sd">    :param window: int, To smooth data before peak-calling. default 1 (no smoothing)</span>
<span class="sd">    :param win_type: str type of smoothing window, default &quot;blackman&quot;</span>
<span class="sd">    :param order: int minimal spacing between peaks, argrelextrema order parameter, default  20</span>
<span class="sd">    :return: list of peaks</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="c1"># smoothing</span>
    <span class="k">if</span> <span class="n">window</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">s1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">rolling</span><span class="p">(</span><span class="n">window</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="n">win_type</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

    <span class="c1"># find peaks</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">argrelextrema</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">comparator</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">greater</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">output</span><span class="p">)</span></div>

<div class="viewcode-block" id="findTroughs"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.findTroughs">[docs]</a><span class="k">def</span> <span class="nf">findTroughs</span><span class="p">(</span><span class="n">s1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(),</span> <span class="n">window</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="s1">&#39;blackman&#39;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Find local minima using SciPy argrelextrema function</span>

<span class="sd">    :param s1: Series data to localize peaks</span>
<span class="sd">    :param window: int, To smooth data before trough-calling. default 1 (no smoothing)</span>
<span class="sd">    :param win_type: str type of smoothing window, default &quot;blackman&quot;</span>
<span class="sd">    :param order: int minimal spacing between min, argrelextrema order parameter, default 20</span>
<span class="sd">    :return: list of troughs</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="c1"># smoothing</span>
    <span class="k">if</span> <span class="n">window</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">s1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">rolling</span><span class="p">(</span><span class="n">window</span><span class="p">,</span> <span class="n">win_type</span><span class="o">=</span><span class="n">win_type</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

    <span class="c1"># find troughs</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">argrelextrema</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">comparator</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">less</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">output</span><span class="p">)</span></div>


<div class="viewcode-block" id="compare1toRef"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.compare1toRef">[docs]</a><span class="k">def</span> <span class="nf">compare1toRef</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(),</span> <span class="n">ranges</span><span class="o">=</span><span class="s1">&#39;mm&#39;</span><span class="p">,</span> <span class="n">heatmap</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Takes Series and compare this with reference DataFrame()</span>

<span class="sd">    :param ref: str with path to csv file or DataFrame</span>
<span class="sd">    :param dataset: Series</span>
<span class="sd">    :param ranges: mm : min-max or qq : q1-q3</span>
<span class="sd">    :param heatmap: boolean, heatmap=False: Dataframe with(reference_above_experiment minimum etc.): rae_min, rae_max, ear_min, ear_max;</span>
<span class="sd">            heatmap=True: Series of differences to plot heatmap</span>
<span class="sd">    :param relative: boolean, only for heatmap, recalculates differences according to the peak size. Warning: negative values are in range -1 to 0 but positive are from 0 to values higher than 1</span>
<span class="sd">    :return: Dataframe (heatmap=False) or Series (heatmap=True)</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">ranges_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;mm&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">],</span> <span class="s1">&#39;qq&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;q1&#39;</span><span class="p">,</span> <span class="s1">&#39;q3&#39;</span><span class="p">]}</span>

    <span class="c1"># preparing dataframe and reference</span>
    <span class="n">differences_df</span><span class="p">,</span> <span class="n">return_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="c1">#handling reference plot</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="n">ref</span>

    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>  <span class="c1"># choosing q1 or min</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span>  <span class="c1"># choosing q3 or max</span>

    <span class="c1">## finding differences (only positive value indicate difference)</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span>

    <span class="c1">## filling differences</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">,</span> <span class="s1">&#39;rae_max&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_min&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_max&#39;</span><span class="p">,</span> <span class="s1">&#39;diff_ear&#39;</span><span class="p">,</span> <span class="s1">&#39;diff_rae&#39;</span><span class="p">,</span> <span class="s1">&#39;diff&#39;</span><span class="p">]:</span> <span class="n">differences_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_max&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_min&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_max&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span>

    <span class="c1"># combining differences for heatmap</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_ear&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_max&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_min&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_rae&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_max&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_ear&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_ear&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_rae&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff_rae&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">heatmap</span> <span class="o">==</span> <span class="kc">True</span> <span class="ow">and</span> <span class="n">relative</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_median&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rel_diff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_median&#39;</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rel_diff&#39;</span><span class="p">]</span>  <span class="c1"># return Series</span>
    <span class="k">elif</span> <span class="n">heatmap</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;diff&#39;</span><span class="p">]</span>  <span class="c1"># return Series</span>
    <span class="k">if</span> <span class="n">heatmap</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">return_df</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">,</span> <span class="s1">&#39;rae_max&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_min&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_max&#39;</span><span class="p">]:</span> <span class="n">return_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">return_df</span>  <span class="c1"># return Dataframe</span></div>


<div class="viewcode-block" id="compareMoretoRef"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.compareMoretoRef">[docs]</a><span class="k">def</span> <span class="nf">compareMoretoRef</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">ranges</span><span class="o">=</span><span class="s1">&#39;mm&#39;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Takes Dataframe created by filter_df and compare this with reference DataFrame</span>

<span class="sd">    :param ref: str with path to csv file or DataFrame</span>
<span class="sd">    :param dataset: Series</span>
<span class="sd">    :param ranges: mm : min-max or qq : q1-q3</span>
<span class="sd">    :return: Dataframe</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">ranges_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;mm&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">],</span> <span class="s1">&#39;qq&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;q1&#39;</span><span class="p">,</span> <span class="s1">&#39;q3&#39;</span><span class="p">]}</span>
    <span class="c1"># preparing dataframe and reference</span>
    <span class="n">differences_df</span><span class="p">,</span> <span class="n">return_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="c1">#handling reference plot</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="n">ref</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># if only two experiments</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_min&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">]</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># if more than two exp</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_min&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>  <span class="c1"># choosing q1 or min</span>
        <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span>  <span class="c1"># choosing q3 or max</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>  <span class="c1"># choosing q1 or min</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">ranges_dict</span><span class="p">[</span><span class="n">ranges</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span>  <span class="c1"># choosing q3 or max</span>

    <span class="c1">## finding differences (only positive value indicate difference)</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_max&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_min&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span>

    <span class="c1">## filling differences</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">,</span> <span class="s1">&#39;rae_max&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_min&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_max&#39;</span><span class="p">]:</span> <span class="n">differences_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_max&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;rae_max&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_above_exp&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_min&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_min&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ref_max&#39;</span><span class="p">]</span>
    <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;ear_max&#39;</span><span class="p">][</span><span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_above_ref&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="s1">&#39;exp_min&#39;</span><span class="p">]</span>
    <span class="c1">#     return_df = dataset</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;rae_min&#39;</span><span class="p">,</span> <span class="s1">&#39;rae_max&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_min&#39;</span><span class="p">,</span> <span class="s1">&#39;ear_max&#39;</span><span class="p">]:</span> <span class="n">return_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">differences_df</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">return_df</span>  <span class="c1"># returns Dataframe</span></div>


<span class="c1">################################################</span>
<span class="c1">#############        plotting</span>

<span class="c1"># def plot_ChIP(df_sense=pd.DataFrame(), df_anti=pd.DataFrame(), title=None, start=None, stop=None, figsize=(15, 6),</span>
<span class="c1">#               ylim=(-0.001, 0.001), s_color=&#39;red&#39;, as_color=&#39;blue&#39;, h_lines=list(), lc=&#39;black&#39;, dpi=150,</span>
<span class="c1">#               csv_path=&#39;/home/tturowski/notebooks/RDN37_reference_collapsed.csv&#39;, color=&#39;green&#39;):</span>
<span class="c1">#     &#39;&#39;&#39;Function creates plot similar to box plot: median, 2 and 3 quartiles and min-max range</span>
<span class="c1">#     csv_path: str()</span>
<span class="c1">#         Path to CRAC or other reference file</span>
<span class="c1">#     title: str()</span>
<span class="c1">#     start: int()</span>
<span class="c1">#     stop: int()</span>
<span class="c1">#     figsize: tuple()</span>
<span class="c1">#     ylim: tuple()</span>
<span class="c1">#         OY axes lim - def (None,0.01)</span>
<span class="c1">#     color: str()</span>
<span class="c1">#         plot color</span>
<span class="c1">#     h_lines: list()</span>
<span class="c1">#         Optional. list() of horizontal lines</span>
<span class="c1">#     lc: str()</span>
<span class="c1">#         color of horizontal lines</span>
<span class="c1">#     Returns</span>
<span class="c1">#     -------</span>
<span class="c1">#     None</span>
<span class="c1">#     &#39;&#39;&#39;</span>
<span class="c1">#     reference = pd.read_csv(csv_path, index_col=0).drop(&#39;nucleotide&#39;, 1)</span>
<span class="c1">#     s2 = reference[start:stop]</span>
<span class="c1">#     # plotting reference dataset</span>
<span class="c1">#     fig, ax1 = plt.subplots(figsize=figsize, dpi=dpi)</span>
<span class="c1">#     plt.title(title)</span>
<span class="c1">#     ax1.set_xlabel(&#39;position&#39;)</span>
<span class="c1">#     ax1.set_ylabel(&#39;fraction of reads&#39;)</span>
<span class="c1">#     ax1.set_ylim(ylim)</span>
<span class="c1">#     # reference plot</span>
<span class="c1">#     ax1.plot(s2.index, s2[&#39;median&#39;], color=color)</span>
<span class="c1">#     if set([&#39;q1&#39;, &#39;q3&#39;]).issubset(list(s2.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, s2[&#39;q1&#39;], s2[&#39;q3&#39;], label=&#39;range (2nd-3rd quartile)&#39;, color=color, alpha=0.2)</span>
<span class="c1">#     if set([&#39;min&#39;, &#39;max&#39;]).issubset(list(s2.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, s2[&#39;min&#39;], s2[&#39;max&#39;], label=&#39;range (min-max)&#39;, color=color, alpha=0.07)</span>
<span class="c1">#</span>
<span class="c1">#     for i in [i for i in h_lines if i in range(start, stop)]: ax1.axvline(i, color=lc)</span>
<span class="c1">#     ax1.axhline(0, color=&#39;black&#39;, alpha=0.7, ls=&#39;dashed&#39;, lw=1)</span>
<span class="c1">#</span>
<span class="c1">#     # ChIP sense</span>
<span class="c1">#     c1 = df_sense[start:stop]</span>
<span class="c1">#     ax1.plot(s2.index, c1[&#39;median&#39;], color=s_color)</span>
<span class="c1">#     if set([&#39;q1&#39;, &#39;q3&#39;]).issubset(list(c1.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, c1[&#39;q1&#39;], c1[&#39;q3&#39;], label=&#39;range (2nd-3rd quartile)&#39;, color=s_color, alpha=0.2)</span>
<span class="c1">#     if set([&#39;min&#39;, &#39;max&#39;]).issubset(list(c1.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, c1[&#39;min&#39;], c1[&#39;max&#39;], label=&#39;range (min-max)&#39;, color=s_color, alpha=0.07)</span>
<span class="c1">#     # ChIP antisense</span>
<span class="c1">#     c2 = df_anti[start:stop] * -1</span>
<span class="c1">#     ax1.plot(s2.index, c2[&#39;median&#39;], color=as_color)</span>
<span class="c1">#     if set([&#39;q1&#39;, &#39;q3&#39;]).issubset(list(c2.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, c2[&#39;q1&#39;], c2[&#39;q3&#39;], label=&#39;range (2nd-3rd quartile)&#39;, color=as_color, alpha=0.2)</span>
<span class="c1">#     if set([&#39;min&#39;, &#39;max&#39;]).issubset(list(c2.columns.values)):</span>
<span class="c1">#         ax1.fill_between(s2.index, c2[&#39;min&#39;], c2[&#39;max&#39;], label=&#39;range (min-max)&#39;, color=as_color, alpha=0.07)</span>
<span class="c1">#</span>
<span class="c1">#     ax1.legend()</span>
<span class="c1">#     return None</span>


<span class="c1">################################################</span>
<span class="c1">#############        other</span>

<div class="viewcode-block" id="save_csv"><a class="viewcode-back" href="../../TTools.html#TTools.profiles.save_csv">[docs]</a><span class="k">def</span> <span class="nf">save_csv</span><span class="p">(</span><span class="n">data_ref</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">datasets</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(),</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;Saves Dataframe to csv</span>

<span class="sd">    :param data_ref: DataFrame with ``[&#39;position&#39;]`` and ``[&#39;nucleotide&#39;]`` columns</span>
<span class="sd">    :param datasets: DataFrame containinig experimental data only</span>
<span class="sd">    :param path: str, Optional: path to save csv. Default None</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">reference</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="c1"># if &#39;position&#39;</span>

    <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">],</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_ref</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">],</span> <span class="n">data_ref</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span>
    <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">],</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">datasets</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;q1&#39;</span><span class="p">],</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;q3&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">datasets</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">],</span> <span class="n">reference</span><span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">datasets</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">path</span><span class="p">:</span>
        <span class="n">reference</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>  <span class="c1">## reference plot</span>
    <span class="k">return</span> <span class="n">reference</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Tomasz W. Turowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>